diff --git a/m4/eval.c b/m4/eval.c
index 9ee73fc..c88ecfb 100644
--- a/m4/eval.c
+++ b/m4/eval.c
@@ -871,13 +871,19 @@ dosubstr(const char *argv[], int argc)
 {
 	const char *ap, *fc, *k;
 	int nc;
+	int offset;
+	size_t len;
 
 	ap = argv[2];		       /* target string */
 #ifdef EXPR
-	fc = ap + expr(argv[3]);       /* first char */
+	offset = expr(argv[3]);	       /* first char */
 #else
-	fc = ap + atoi(argv[3]);       /* first char */
+	offset = atoi(argv[3]);	       /* first char */
 #endif
+	len = strlen(ap);
+	if (offset < 0 || (size_t)offset >= len)
+		return;
+	fc = ap + offset;
 	nc = strlen(fc);
 	if (argc >= 5)
 #ifdef EXPR
@@ -885,7 +891,7 @@ dosubstr(const char *argv[], int argc)
 #else
 		nc = min(nc, atoi(argv[4]));
 #endif
-	if (fc >= ap && fc < ap + strlen(ap))
+	if (fc >= ap && fc < ap + len)
 		for (k = fc + nc - 1; k >= fc; k--)
 			pushback(*k);
 }
