diff --git a/httpd/server_http.c b/httpd/server_http.c
index b01e018..3a57a5b 100644
--- a/httpd/server_http.c
+++ b/httpd/server_http.c
@@ -145,7 +145,7 @@ server_http_authenticate(struct server_config *srv_conf, struct client *clt)
 		goto done;
 
 	if (b64_pton(strchr(ba->kv_value, ' ') + 1, (uint8_t *)decoded,
-	    sizeof(decoded)) <= 0)
+	    sizeof(decoded) - 1) <= 0)
 		goto done;
 
 	if ((clt_pass = strchr(decoded, ':')) == NULL)
