diff --git a/usr.bin/ssh/channels.c b/usr.bin/ssh/channels.c
index 712e816..834269d 100644
--- a/usr.bin/ssh/channels.c
+++ b/usr.bin/ssh/channels.c
@@ -4208,9 +4208,11 @@ remote_open_match(struct permission *allowed_open, struct Forward *fwd)
 	int ret;
 	char *lhost;
 
-	/* XXX add ACLs for streamlocal */
-	if (fwd->listen_path != NULL)
-		return 1;
+	if (fwd->listen_path != NULL) {
+		if (allowed_open->listen_path == NULL)
+			return 0;
+		return strcmp(allowed_open->listen_path, fwd->listen_path) == 0;
+	}
 
 	if (fwd->listen_host == NULL || allowed_open->listen_host == NULL)
 		return 0;
