diff --git a/usr.bin/rdist/common.c b/usr.bin/rdist/common.c
index 8243dfe..721feaf 100644
--- a/usr.bin/rdist/common.c
+++ b/usr.bin/rdist/common.c
@@ -522,44 +522,48 @@ response(void)
 {
 	static u_char resp[BUFSIZ];
 	u_char *s;
-	int n;
+	int n, notices = 0;
 
 	debugmsg(DM_CALL, "response() start\n");
 
-	n = remline(s = resp, sizeof(resp), 0);
+	for (;;) {
+		n = remline(s = resp, sizeof(resp), 0);
 
-	n--;
-	switch (*s++) {
+		n--;
+		switch (*s++) {
         case C_ACK:
-		debugmsg(DM_PROTO, "received ACK\n");
-		return(0);
-	case C_LOGMSG:
-		if (n > 0) {
-			message(MT_CHANGE, "%s", s);
-			return(1);
+			debugmsg(DM_PROTO, "received ACK\n");
+			return(0);
+		case C_LOGMSG:
+			if (n > 0) {
+				message(MT_CHANGE, "%s", s);
+				return(1);
+			}
+			debugmsg(DM_PROTO, "received EMPTY logmsg\n");
+			return(0);
+		case C_NOTEMSG:
+			if (++notices > BUFSIZ)
+				return(-1);
+			if (s)
+				message(MT_NOTICE, "%s", s);
+			continue;
+
+		default:
+			s--;
+			n++;
+			/* fall into... */
+
+		case C_ERRMSG:	/* Normal error message */
+			if (s)
+				message(MT_NERROR, "%s", s);
+			return(-1);
+
+		case C_FERRMSG:	/* Fatal error message */
+			if (s)
+				message(MT_FERROR, "%s", s);
+			finish();
+			return(-1);
 		}
-		debugmsg(DM_PROTO, "received EMPTY logmsg\n");
-		return(0);
-	case C_NOTEMSG:
-		if (s)
-			message(MT_NOTICE, "%s", s);
-		return(response());
-
-	default:
-		s--;
-		n++;
-		/* fall into... */
-
-	case C_ERRMSG:	/* Normal error message */
-		if (s)
-			message(MT_NERROR, "%s", s);
-		return(-1);
-
-	case C_FERRMSG:	/* Fatal error message */
-		if (s)
-			message(MT_FERROR, "%s", s);
-		finish();
-		return(-1);
 	}
 	/*NOTREACHED*/
 }
