diff --git a/usr.sbin/relayd/relay_http.c b/usr.sbin/relayd/relay_http.c
index d6960e8..c458d59 100644
--- a/usr.sbin/relayd/relay_http.c
+++ b/usr.sbin/relayd/relay_http.c
@@ -1323,6 +1323,9 @@ relay_writeheader_http(struct ctl_relay_event *dst, struct ctl_relay_event
 	struct http_descriptor	*desc = (struct http_descriptor *)cre->desc;
 
 	RB_FOREACH(hdr, kvtree, &desc->http_headers) {
+		if (desc->http_chunked &&
+		    strcasecmp(hdr->kv_key, "Content-Length") == 0)
+			continue;
 		if (relay_writeheader_kv(dst, hdr) == -1)
 			return (-1);
 		TAILQ_FOREACH(kv, &hdr->kv_children, kv_entry) {
