diff --git a/usr.bin/ssh/mux.c b/usr.bin/ssh/mux.c
index db4ac80..00d7bdc 100644
--- a/usr.bin/ssh/mux.c
+++ b/usr.bin/ssh/mux.c
@@ -284,6 +284,7 @@ mux_master_process_hello(struct ssh *ssh, u_int rid,
 		if ((r = sshbuf_get_cstring(m, &name, NULL)) != 0 ||
 		    (r = sshbuf_get_string_direct(m, NULL, &value_len)) != 0) {
 			error_fr(r, "parse extension");
+			free(name);
 			return -1;
 		}
 		if (strcmp(name, "info") == 0) {
