diff --git a/libelf/gelf_rel.c b/libelf/gelf_rel.c
index 8058630..9af6433 100644
--- a/libelf/gelf_rel.c
+++ b/libelf/gelf_rel.c
@@ -134,7 +134,8 @@ gelf_update_rel(Elf_Data *ed, int ndx, GElf_Rel *dr)
 
 	assert(ndx >= 0);
 
-	if (msz * (size_t) ndx >= d->d_data.d_size) {
+	if (d->d_data.d_size < msz ||
+	    (size_t) ndx > (d->d_data.d_size - msz) / msz) {
 		LIBELF_SET_ERROR(ARGUMENT, 0);
 		return (0);
 	}
