diff --git a/sbin/iked/ikev2.c b/sbin/iked/ikev2.c
index e235542..f0d667d 100644
--- a/sbin/iked/ikev2.c
+++ b/sbin/iked/ikev2.c
@@ -7023,6 +7023,8 @@ ikev2_print_id(struct iked_id *id, char *idstr, size_t idstrlen)
 
 	switch (id->id_type) {
 	case IKEV2_ID_IPV4:
+		if ((size_t)len != sizeof(s4.sin_addr.s_addr))
+			return (-1);
 		s4.sin_family = AF_INET;
 		s4.sin_len = sizeof(s4);
 		memcpy(&s4.sin_addr.s_addr, ptr, len);
@@ -7042,6 +7044,8 @@ ikev2_print_id(struct iked_id *id, char *idstr, size_t idstrlen)
 		free(str);
 		break;
 	case IKEV2_ID_IPV6:
+		if ((size_t)len != sizeof(s6.sin6_addr))
+			return (-1);
 		s6.sin6_family = AF_INET6;
 		s6.sin6_len = sizeof(s6);
 		memcpy(&s6.sin6_addr, ptr, len);
