diff --git a/sbin/iked/ocsp.c b/sbin/iked/ocsp.c
index 91c44fc..7ad1a9b 100644
--- a/sbin/iked/ocsp.c
+++ b/sbin/iked/ocsp.c
@@ -545,13 +545,13 @@ ocsp_parse_response(struct iked_ocsp *ocsp, OCSP_RESPONSE *resp)
 
 	status = OCSP_check_nonce(ocsp->ocsp_req, bs);
 	if (status <= 0) {
-		if (status == -1)
+		if (status == -1) {
 			log_warnx("%s: no nonce in response",
 			    SPI_SH(&ocsp->ocsp_sh, __func__));
-		else {
+			errstr = "no nonce in response";
+		} else
 			errstr = "nonce verify error";
-			goto done;
-		}
+		goto done;
 	}
 
 	store = X509_STORE_new();
