diff --git a/sbin/isakmpd/policy.c b/sbin/isakmpd/policy.c
index d76f390..6fd50b5 100644
--- a/sbin/isakmpd/policy.c
+++ b/sbin/isakmpd/policy.c
@@ -2031,7 +2031,7 @@ keynote_cert_validate(void *scert)
 		return 0;
 
 	for (i = 0; i < num; i++) {
-		if (kn_verify_assertion(scert, strlen((char *)scert))
+		if (kn_verify_assertion(foo[i], strlen(foo[i]))
 		    != SIGRESULT_TRUE) {
 			for (; i < num; i++)
 				free(foo[i]);
