diff --git a/usr.bin/openssl/s_client.c b/usr.bin/openssl/s_client.c
index 2b05fac..1bd9e2f 100644
--- a/usr.bin/openssl/s_client.c
+++ b/usr.bin/openssl/s_client.c
@@ -1315,6 +1315,8 @@ s_client_main(int argc, char **argv)
 		BIO_printf(sbio,
 		    "<starttls xmlns='urn:ietf:params:xml:ns:xmpp-tls'/>");
 		seen = BIO_read(sbio, sbuf, BUFSIZZ);
+		if (seen < 0 || seen >= BUFSIZZ)
+			goto shut;
 		sbuf[seen] = 0;
 		if (!strstr(sbuf, "<proceed"))
 			goto shut;
