diff --git a/usr.bin/rdistd/server.c b/usr.bin/rdistd/server.c
index 7a0f98c..d4ea25b 100644
--- a/usr.bin/rdistd/server.c
+++ b/usr.bin/rdistd/server.c
@@ -1435,7 +1435,7 @@ recvit(char *cmd, int type)
 		 owner, group, file, catname, (type == S_IFDIR) ? 1 : 0);
 
 	if (type == S_IFDIR) {
-		if ((size_t) catname >= sizeof(sptarget)) {
+		if ((size_t) catname >= sizeof(sptarget) / sizeof(sptarget[0])) {
 			error("%s: too many directory levels", target);
 			return;
 		}
