diff --git a/usr.bin/rdist/client.c b/usr.bin/rdist/client.c
index ef4a574..8b9999f 100644
--- a/usr.bin/rdist/client.c
+++ b/usr.bin/rdist/client.c
@@ -521,6 +521,14 @@ rmchk(opt_t opts)
 				error("rmchk: cannot decode file");
 				return(-1);
 			}
+			if (targ[0] == '/' || strcmp(targ, "..") == 0 ||
+			    strncmp(targ, "../", 3) == 0 ||
+			    strstr(targ, "/../") != NULL ||
+			    (strlen(targ) >= 3 &&
+			    strcmp(targ + strlen(targ) - 3, "/..") == 0)) {
+				error("rmchk: invalid filename");
+				return(-1);
+			}
 			(void) snprintf(ptarget,
 					sizeof(target) - (ptarget - target),
 					"%s%s",
