diff --git a/usr.sbin/user/user.c b/usr.sbin/user/user.c
index b1bea38..ed0022c 100644
--- a/usr.sbin/user/user.c
+++ b/usr.sbin/user/user.c
@@ -1450,7 +1450,6 @@ moduser(char *login_name, char *newlogin, user_t *up)
 	char		acctlock_str[] = "-";
 	char		pwlock_str[] = "*";
 	char		pw_len[PasswordLength + 1];
-	char		shell_len[MaxShellNameLen];
 	char		*shell_last_char;
 	size_t		colonc, loginc;
 	size_t		cc;
@@ -1549,14 +1548,15 @@ moduser(char *login_name, char *newlogin, user_t *up)
 		if (up->u_flags & F_ACCTLOCK) {
 			/* lock the account */
 			if (*shell_last_char != *acctlock_str) {
-				shell_tmp = malloc(strlen(pwp->pw_shell) + sizeof(acctlock_str));
+				shell_buf = strlen(pwp->pw_shell) + sizeof(acctlock_str);
+				shell_tmp = malloc(shell_buf);
 				if (shell_tmp == NULL) {
 					close(ptmpfd);
 					pw_abort();
 					errx(EXIT_FAILURE, "account lock: cannot allocate memory");
 				}
-				strlcpy(shell_tmp, pwp->pw_shell, sizeof(shell_len));
-				strlcat(shell_tmp, acctlock_str, sizeof(shell_len));
+				strlcpy(shell_tmp, pwp->pw_shell, shell_buf);
+				strlcat(shell_tmp, acctlock_str, shell_buf);
 				pwp->pw_shell = shell_tmp;
 			} else {
 				locked++;
