diff --git a/lib/libagentx/ax.c b/lib/libagentx/ax.c
index f4d6eb7..1ac9e7b 100644
--- a/lib/libagentx/ax.c
+++ b/lib/libagentx/ax.c
@@ -1288,6 +1288,8 @@ ax_pdutoostring(struct ax_pdu_header *header,
 	buf += 4;
 	if (ostring->aos_slen > rawlen)
 		goto fail;
+	if (ostring->aos_slen + (4 - (ostring->aos_slen % 4)) % 4 > rawlen)
+		goto fail;
 	if ((ostring->aos_string = malloc(ostring->aos_slen + 1)) == NULL)
 		return -1;
 	memcpy(ostring->aos_string, buf, ostring->aos_slen);
