diff --git a/sbin/restore/tape.c b/sbin/restore/tape.c
index fc7f280..79b447c 100644
--- a/sbin/restore/tape.c
+++ b/sbin/restore/tape.c
@@ -263,6 +263,8 @@ setup(void)
 	getfile(xtrmap, xtrmapskip);
 	if (spcl.c_type != TS_BITS)
 		errx(1, "Cannot find file dump list");
+	if (spcl.c_count > howmany(maxino, NBBY) / TP_BSIZE)
+		errx(1, "File dump list is too large");
 	map = calloc(1, howmany(maxino, NBBY));
 	if (map == NULL)
 		panic("no memory for file dump list\n");
