diff --git a/usr.bin/nm/nm.c b/usr.bin/nm/nm.c
index 41b2540..49119cf 100644
--- a/usr.bin/nm/nm.c
+++ b/usr.bin/nm/nm.c
@@ -293,6 +293,7 @@ process_file(int count, const char *fname)
 }
 
 char *nametab;
+u_long nametablen;
 
 /*
  *
@@ -308,6 +309,8 @@ mmbr_name(struct ar_hdr *arh, char **name, int baselen, int *namelen, FILE *fp)
 		int len;
 
 		i = atol(&arh->ar_name[1]);
+		if (i < 0 || (u_long)i >= nametablen)
+			return (1);
 		len = strlen(&nametab[i]) + 1;
 		if (len > *namelen) {
 			p -= (long)*name;
@@ -500,6 +503,7 @@ show_archive(int count, const char *fname, FILE *fp)
 
 	rval = 0;
 	nametab = NULL;
+	nametablen = 0;
 	symtaboff = 0;
 	symtablen = 0;
 
@@ -550,6 +554,7 @@ show_archive(int count, const char *fname, FILE *fp)
 				break;
 			}
 
+			nametablen = mmbrlen;
 			for (p = nametab, i = mmbrlen; i--; p++)
 				if (*p == '\n')
 					*p = '\0';
@@ -618,6 +623,7 @@ skip:		if (fseeko(fp, last_ar_off + even(mmbrlen), SEEK_SET)) {
 	}
 	free(nametab);
 	nametab = NULL;
+	nametablen = 0;
 	free(name);
 	return(rval);
 }
