diff --git a/usr.sbin/installboot/i386_installboot.c b/usr.sbin/installboot/i386_installboot.c
index d1830fb..4909bc3 100644
--- a/usr.sbin/installboot/i386_installboot.c
+++ b/usr.sbin/installboot/i386_installboot.c
@@ -667,6 +667,8 @@ findgptefisys(int devfd, struct disklabel *dl, int *gpartp,
 	ghpartsize = letoh32(gh.gh_part_size);
 	ghpartspersec = dl->d_secsize / ghpartsize;
 	ghpartnum = letoh32(gh.gh_part_num);
+	if (ghpartnum > NGPTPARTITIONS)
+		return (-1);
 	if ((secbuf = malloc(dl->d_secsize)) == NULL)
 		err(1, NULL);
 	for (i = 0; i < (ghpartnum + ghpartspersec - 1) / ghpartspersec; i++) {
