diff --git a/usr.sbin/npppd/npppd/npppd_iface.c b/usr.sbin/npppd/npppd/npppd_iface.c
index 103520f..af1638f 100644
--- a/usr.sbin/npppd/npppd/npppd_iface.c
+++ b/usr.sbin/npppd/npppd/npppd_iface.c
@@ -418,6 +418,7 @@ npppd_iface_network_input_delegate(struct radish *radish, void *args0)
 			/* output via MPPE if MPPE started */
 			mppe_pkt_output(&ppp->mppe, PPP_PROTO_IP, args->pktp,
 			    args->lpktp);
+			return 0;
 		} else if (MPPE_IS_REQUIRED(ppp)) {
 			/* in case MPPE not started but MPPE is mandatory, */
 			/* it is not necessary to log because of multicast. */
