diff --git a/lib/libsndio/aucat.c b/lib/libsndio/aucat.c
index ee1a460..51d0159 100644
--- a/lib/libsndio/aucat.c
+++ b/lib/libsndio/aucat.c
@@ -69,6 +69,11 @@ _aucat_rmsg(struct aucat *hdl, int *eof)
 		hdl->rtodo -= n;
 	}
 	if (ntohl(hdl->rmsg.cmd) == AMSG_DATA) {
+		if (ntohl(hdl->rmsg.u.data.size) > AMSG_DATAMAX) {
+			DPRINTF("_aucat_rmsg: data too large\n");
+			*eof = 1;
+			return 0;
+		}
 		hdl->rtodo = ntohl(hdl->rmsg.u.data.size);
 		hdl->rstate = RSTATE_DATA;
 	} else {
