diff --git a/usr.sbin/ldpd/ldpd.c b/usr.sbin/ldpd/ldpd.c
index 4032c02..49b143f 100644
--- a/usr.sbin/ldpd/ldpd.c
+++ b/usr.sbin/ldpd/ldpd.c
@@ -1259,7 +1259,7 @@ merge_auths(struct ldpd_conf *conf, struct ldpd_conf *xconf)
 	/* find deleted auths */
 	LIST_FOREACH_SAFE(auth, &conf->auth_list, entry, nauth) {
 		xauth = auth_find(xconf, auth);
-		if (xauth == NULL)
+		if (xauth != NULL)
 			continue;
 
 		LIST_REMOVE(auth, entry);
