diff --git a/usr.sbin/ocspcheck/ocspcheck.c b/usr.sbin/ocspcheck/ocspcheck.c
index ae4b836..5ffdcb1 100644
--- a/usr.sbin/ocspcheck/ocspcheck.c
+++ b/usr.sbin/ocspcheck/ocspcheck.c
@@ -481,6 +481,10 @@ validate_response(char *buf, size_t size, ocsp_request *request,
 			warnx("Certificate revoked at: %s", ctime(&rev_t));
 		goto err;
 	}
+	if (cert_status != V_OCSP_CERTSTATUS_GOOD) {
+		warnx("Invalid OCSP reply: certificate status is not good");
+		goto err;
+	}
 	if ((this_t = parse_ocsp_time(thisupd)) == -1) {
 		warnx("unable to parse this update time in OCSP reply");
 		goto err;
