diff --git a/lib/libfuse/fuse_ops.c b/lib/libfuse/fuse_ops.c
index 9527191..bfe8163 100644
--- a/lib/libfuse/fuse_ops.c
+++ b/lib/libfuse/fuse_ops.c
@@ -444,7 +444,7 @@ ifuse_ops_lookup(fuse_req_t req, fuse_ino_t parent, const char *name)
 	struct fuse_entry_param entry;
 	struct fuse_vnode *vn;
 	char *realname;
-	int err;
+	int err, newvn = 0;
 
 	vn = get_vn_by_name_and_parent(f, name, parent);
 	if (vn == NULL) {
@@ -454,6 +454,7 @@ ifuse_ops_lookup(fuse_req_t req, fuse_ino_t parent, const char *name)
 			goto out;
 		}
 		set_vn(f, vn); /*XXX*/
+		newvn = 1;
 	} else if (vn->ino != FUSE_ROOT_INO)
 		ref_vn(vn);
 
@@ -471,8 +472,11 @@ ifuse_ops_lookup(fuse_req_t req, fuse_ino_t parent, const char *name)
 out:
 	if (!err)
 		fuse_reply_entry(req, &entry);
-	else
+	else {
+		if (newvn)
+			unref_vn(f, vn);
 		fuse_reply_err(req, -err);
+	}
 }
 
 static void
