diff --git a/usr.bin/ssh/auth-rhosts.c b/usr.bin/ssh/auth-rhosts.c
index 3717e93..ab36a20 100644
--- a/usr.bin/ssh/auth-rhosts.c
+++ b/usr.bin/ssh/auth-rhosts.c
@@ -46,7 +46,8 @@ extern ServerOptions options;
 /*
  * This function processes an rhosts-style file (.rhosts, .shosts, or
  * /etc/hosts.equiv).  This returns true if authentication can be granted
- * based on the file, and returns zero otherwise.
+ * based on the file, -1 if access is denied by a negative entry, and zero
+ * otherwise.
  */
 
 static int
@@ -167,7 +168,7 @@ check_rhosts_file(const char *filename, const char *hostname,
 		if (negated) {
 			auth_debug_add("Matched negative entry in %.100s.",
 			    filename);
-			return 0;
+			return -1;
 		}
 		/* Accept authentication. */
 		return 1;
@@ -235,14 +236,20 @@ auth_rhosts2(struct passwd *pw, const char *client_user, const char *hostname,
 	if (pw->pw_uid == 0)
 		debug3_f("root user, ignoring system hosts files");
 	else {
-		if (check_rhosts_file(_PATH_RHOSTS_EQUIV, hostname, ipaddr,
-		    client_user, pw->pw_name)) {
+		r = check_rhosts_file(_PATH_RHOSTS_EQUIV, hostname, ipaddr,
+		    client_user, pw->pw_name);
+		if (r < 0)
+			return 0;
+		if (r > 0) {
 			auth_debug_add("Accepted for %.100s [%.100s] by "
 			    "/etc/hosts.equiv.", hostname, ipaddr);
 			return 1;
 		}
-		if (check_rhosts_file(_PATH_SSH_HOSTS_EQUIV, hostname, ipaddr,
-		    client_user, pw->pw_name)) {
+		r = check_rhosts_file(_PATH_SSH_HOSTS_EQUIV, hostname, ipaddr,
+		    client_user, pw->pw_name);
+		if (r < 0)
+			return 0;
+		if (r > 0) {
 			auth_debug_add("Accepted for %.100s [%.100s] by "
 			    "%.100s.", hostname, ipaddr, _PATH_SSH_HOSTS_EQUIV);
 			return 1;
@@ -312,8 +319,9 @@ auth_rhosts2(struct passwd *pw, const char *client_user, const char *hostname,
 			continue;
 		}
 		/* Check if authentication is permitted by the file. */
-		if (check_rhosts_file(path, hostname, ipaddr,
-		    client_user, pw->pw_name)) {
+		r = check_rhosts_file(path, hostname, ipaddr,
+		    client_user, pw->pw_name);
+		if (r > 0) {
 			auth_debug_add("Accepted by %.100s.",
 			    rhosts_files[rhosts_file_index]);
 			/* Restore the privileged uid. */
