diff --git a/usr.sbin/tftp-proxy/tftp-proxy.c b/usr.sbin/tftp-proxy/tftp-proxy.c
index 1d494bb..b55710f 100644
--- a/usr.sbin/tftp-proxy/tftp-proxy.c
+++ b/usr.sbin/tftp-proxy/tftp-proxy.c
@@ -58,6 +58,7 @@
 #define DEFTRANSWAIT	2
 #define NTOP_BUFS	4
 #define PKTSIZE		SEGSIZE+4
+#define MAX_REQUESTS	256
 
 const char *opcode(int);
 const char *sock_ntop(struct sockaddr *);
@@ -192,6 +193,7 @@ struct proxy_child {
 	struct event push_ev;
 	struct event pop_ev;
 	struct evbuffer *buf;
+	size_t nrequests;
 };
 
 struct proxy_child *child = NULL;
@@ -691,6 +693,11 @@ proxy_recv(int fd, short events, void *arg)
 	struct proxy_request *r;
 	struct tftphdr *tp;
 
+	if (child->nrequests >= MAX_REQUESTS) {
+		recv(fd, safety, sizeof(safety), 0);
+		return;
+	}
+
 	r = calloc(1, sizeof(*r));
 	if (r == NULL) {
 		recv(fd, safety, sizeof(safety), 0);
@@ -756,6 +763,7 @@ proxy_recv(int fd, short events, void *arg)
 	}
 
 	TAILQ_INSERT_TAIL(&child->fdrequests, r, entry);
+	child->nrequests++;
 	evbuffer_add(child->buf, &r->addrs, sizeof(r->addrs));
 	event_add(&child->push_ev, NULL);
 
@@ -903,6 +911,7 @@ unprivproc_timeout(int fd, short events, void *arg)
 	struct proxy_request *r = arg;
 
 	TAILQ_REMOVE(&child->tmrequests, r, entry);
+	child->nrequests--;
 
 	/* delete our rdr rule and clean up */
 	prepare_commit(r->id);
