diff --git a/sbin/isakmpd/message.c b/sbin/isakmpd/message.c
index ee0df3a..5e6959f 100644
--- a/sbin/isakmpd/message.c
+++ b/sbin/isakmpd/message.c
@@ -911,6 +911,12 @@ message_validate_notify(struct message *msg, struct payload *p)
 		message_free(msg);
 		return -1;
 	}
+	if (GET_ISAKMP_GEN_LENGTH(p->p) < ISAKMP_NOTIFY_SPI_OFF +
+	    GET_ISAKMP_NOTIFY_SPI_SZ(p->p)) {
+		log_print("message_validate_notify: payload too short");
+		message_drop(msg, ISAKMP_NOTIFY_PAYLOAD_MALFORMED, 0, 1, 1);
+		return -1;
+	}
 
 	/* Validate the SPI. XXX Just ISAKMP for now.  */
 	if (proto == ISAKMP_PROTO_ISAKMP &&
