diff --git a/usr.bin/less/decode.c b/usr.bin/less/decode.c
index 4846e0c..0498f0c 100644
--- a/usr.bin/less/decode.c
+++ b/usr.bin/less/decode.c
@@ -543,15 +543,33 @@ gint(char **sp)
 static int
 old_lesskey(char *buf, int len)
 {
+	char *p;
+	char *end;
+	int a;
+
 	/*
 	 * Old-style lesskey file.
-	 * The file must end with either
-	 *	..,cmd,0,action
-	 * or	...,cmd,0,action|A_EXTRA,string,0
-	 * So the last byte or the second to last byte must be zero.
+	 * Each command must have an action byte after its terminating null.
+	 * If the action has A_EXTRA, it must also have a terminating null.
 	 */
-	if (buf[len-1] != '\0' && buf[len-2] != '\0')
-		return (-1);
+	end = buf + len;
+	for (p = buf; p < end; ) {
+		while (p < end && *p != '\0')
+			p++;
+		if (p == end)
+			return (-1);
+		p++;
+		if (p == end)
+			return (-1);
+		a = *p++ & 0377;
+		if (a & A_EXTRA) {
+			while (p < end && *p != '\0')
+				p++;
+			if (p == end)
+				return (-1);
+			p++;
+		}
+	}
 	add_fcmd_table(buf, len);
 	return (0);
 }
