diff --git a/usr.sbin/mopd/common/print.c b/usr.sbin/mopd/common/print.c
index 13df7d8..fbcb2a2 100644
--- a/usr.sbin/mopd/common/print.c
+++ b/usr.sbin/mopd/common/print.c
@@ -440,6 +440,8 @@ mopPrintInfo(FILE *fd, u_char *pkt, int *idx, u_short moplen, u_char mopcode,
 
 	while (*idx < (moplen + 2)) {
 		ilen = mopGetChar(pkt, idx);
+		if (ilen > (moplen + 2) - *idx)
+			return;
 		switch (itype) {
 		case 0:
 			tmpc  = mopGetChar(pkt, idx);
