diff --git a/usr.bin/mg/extend.c b/usr.bin/mg/extend.c
index e4a539d..982f052 100644
--- a/usr.bin/mg/extend.c
+++ b/usr.bin/mg/extend.c
@@ -832,9 +832,13 @@ excline(char *line, int llen, int lnum)
 					}
 					argp++;
 				}
-				if (bind == BINDARG)
+				if (bind == BINDARG) {
+					if (key.k_count >= MAXKEY) {
+						status = FALSE;
+						goto cleanup;
+					}
 					key.k_chars[key.k_count++] = c;
-				else
+				} else
 					lp->l_text[lp->l_used++] = c;
 			}
 			if (*line)
