diff --git a/usr.bin/ssh/ssh-keyscan.c b/usr.bin/ssh/ssh-keyscan.c
index 8b169db..bf6bf47 100644
--- a/usr.bin/ssh/ssh-keyscan.c
+++ b/usr.bin/ssh/ssh-keyscan.c
@@ -433,6 +433,7 @@ congreet(int s)
 	char buf[256], *cp;
 	char remote_version[sizeof buf];
 	size_t bufsiz;
+	struct timespec now;
 	con *c = &fdcon[s];
 
 	/* send client banner */
@@ -459,6 +460,11 @@ congreet(int s)
 	 * in multiple iterations of the outer loop).
 	 */
 	for (;;) {
+		monotime_ts(&now);
+		if (timespeccmp(&now, &c->c_ts, >=)) {
+			conrecycle(s);
+			return;
+		}
 		memset(buf, '\0', sizeof(buf));
 		bufsiz = sizeof(buf);
 		cp = buf;
