diff --git a/usr.sbin/lldpd/lldpd.c b/usr.sbin/lldpd/lldpd.c
index cbfa1d6..d45a731 100644
--- a/usr.sbin/lldpd/lldpd.c
+++ b/usr.sbin/lldpd/lldpd.c
@@ -63,6 +63,8 @@
 int rdaemon(int);
 
 #define LLDPD_USER		"_lldpd"
+#define LLDPD_MAX_MSAPS		4096
+#define LLDPD_MAX_IF_MSAPS	256
 
 #define CMSG_FOREACH(_cmsg, _msgp) \
 	for ((_cmsg) = CMSG_FIRSTHDR((_msgp)); \
@@ -594,6 +596,8 @@ ensock_recv(int s, short events, void *arg)
 	int ok;
 	unsigned int idlen;
 	unsigned int ttl;
+	unsigned int if_msaps = 0;
+	unsigned int msaps = 0;
 	struct timeval age;
 	int update = 0;
 
@@ -727,6 +731,7 @@ ensock_recv(int s, short events, void *arg)
 	}
 
 	TAILQ_FOREACH(msap, &ifp->if_msaps, msap_entry) {
+		if_msaps++;
 		if (msap->msap_id_len == idlen &&
 		    memcmp(msap->msap_pdu, buf, idlen) == 0)
 			break;
@@ -738,6 +743,17 @@ ensock_recv(int s, short events, void *arg)
 			return;
 		}
 
+		TAILQ_FOREACH(msap, &lldpd->msaps, msap_aentry)
+			msaps++;
+
+		if (if_msaps >= LLDPD_MAX_IF_MSAPS ||
+		    msaps >= LLDPD_MAX_MSAPS) {
+			ldebug("%s: too many msaps", ifp->if_key.if_name);
+			agent_counter_inc(lldpd, ifp,
+			    statsFramesDiscardedTotal);
+			return;
+		}
+
 		msap = malloc(sizeof(*msap));
 		if (msap == NULL) {
 			lwarn("%s: msap alloc", ifp->if_key.if_name);
