diff --git a/sbin/unwind/libunbound/util/data/dname.c b/sbin/unwind/libunbound/util/data/dname.c
index 5370aa6..c5c19e1 100644
--- a/sbin/unwind/libunbound/util/data/dname.c
+++ b/sbin/unwind/libunbound/util/data/dname.c
@@ -240,6 +240,8 @@ dname_pkt_compare(sldns_buffer* pkt, uint8_t* d1, uint8_t* d2)
 	while( len1 != 0 || len2 != 0 ) {
 		/* resolve ptrs */
 		if(LABEL_IS_PTR(len1)) {
+			if(d1 >= sldns_buffer_end(pkt))
+				return -1;
 			if((size_t)PTR_OFFSET(len1, *d1)
 				>= sldns_buffer_limit(pkt))
 				return -1;
@@ -250,6 +252,8 @@ dname_pkt_compare(sldns_buffer* pkt, uint8_t* d1, uint8_t* d2)
 			continue;
 		}
 		if(LABEL_IS_PTR(len2)) {
+			if(d2 >= sldns_buffer_end(pkt))
+				return 1;
 			if((size_t)PTR_OFFSET(len2, *d2)
 				>= sldns_buffer_limit(pkt))
 				return 1;
