diff --git a/usr.bin/infocmp/infocmp.c b/usr.bin/infocmp/infocmp.c
index e23b729..2de91d3 100644
--- a/usr.bin/infocmp/infocmp.c
+++ b/usr.bin/infocmp/infocmp.c
@@ -1087,7 +1087,7 @@ file_comparison(int argc, char *argv[])
 			   argv[0],
 			   qp->ncrosslinks,
 			   argv[1]);
-	    for (i = 0; i < qp->ncrosslinks; i++)
+	    for (i = 0; i < qp->ncrosslinks && i < MAX_CROSSLINKS; i++)
 		(void) fprintf(stderr,
 			       "\t%s\n",
 			       _nc_first_name((qp->crosslinks[i])->tterm.term_names));
@@ -1102,7 +1102,7 @@ file_comparison(int argc, char *argv[])
 			   argv[1],
 			   rp->ncrosslinks,
 			   argv[0]);
-	    for (i = 0; i < rp->ncrosslinks; i++)
+	    for (i = 0; i < rp->ncrosslinks && i < MAX_CROSSLINKS; i++)
 		(void) fprintf(stderr,
 			       "\t%s\n",
 			       _nc_first_name((rp->crosslinks[i])->tterm.term_names));
