diff --git a/usr.bin/ldap/aldap.c b/usr.bin/ldap/aldap.c
index aee14a6..817c4a5 100644
--- a/usr.bin/ldap/aldap.c
+++ b/usr.bin/ldap/aldap.c
@@ -418,9 +418,12 @@ aldap_parse(struct aldap *ldap)
 		if (m->msg->be_sub) {
 			for (ep = m->msg->be_sub; ep != NULL; ep = ep->be_next) {
 				ober_scanf_elements(ep, "t", &class, &type);
-				if (class == 2 && type == 0)
+				if (class == 2 && type == 0) {
+					if (ep->be_sub == NULL || ep->be_sub->be_sub == NULL)
+						goto parsefail;
 					m->page = aldap_parse_page_control(ep->be_sub->be_sub,
 					    ep->be_sub->be_sub->be_len);
+				}
 			}
 		} else
 			m->page = NULL;
