diff --git a/usr.bin/ldap/aldap.c b/usr.bin/ldap/aldap.c
index aee14a6..5e469e4 100644
--- a/usr.bin/ldap/aldap.c
+++ b/usr.bin/ldap/aldap.c
@@ -471,6 +471,13 @@ aldap_parse_page_control(struct ber_element *control, size_t len)
 		return NULL;
 	}
 
+	if (elm == NULL || elm->be_sub == NULL || elm->be_sub->be_next == NULL) {
+		if (elm != NULL)
+			ober_free_elements(elm);
+		ober_free(&b);
+		free(page);
+		return NULL;
+	}
 	ober_scanf_elements(elm->be_sub, "is", &page->size, &s);
 	page->cookie_len = elm->be_sub->be_next->be_len;
