diff --git a/usr.sbin/traceroute/worker.c b/usr.sbin/traceroute/worker.c
index 0865f4d..1ee5c84 100644
--- a/usr.sbin/traceroute/worker.c
+++ b/usr.sbin/traceroute/worker.c
@@ -657,7 +657,7 @@ packet_ok6(struct tr_conf *conf, struct msghdr *mhdr, int cc, int *seq)
 		    icp->icmp6_code);
 		p = (u_int8_t *)(icp + 1);
 #define WIDTH	16
-		for (i = 0; i < cc; i++) {
+		for (i = 0; i < cc - sizeof(*icp); i++) {
 			if (i % WIDTH == 0)
 				printf("%04x:", i);
 			if (i % 4 == 0)
@@ -666,7 +666,7 @@ packet_ok6(struct tr_conf *conf, struct msghdr *mhdr, int cc, int *seq)
 			if (i % WIDTH == WIDTH - 1)
 				printf("\n");
 		}
-		if (cc % WIDTH != 0)
+		if ((cc - sizeof(*icp)) % WIDTH != 0)
 			printf("\n");
 	}
 	return(0);
