diff --git a/sbin/restore/dirs.c b/sbin/restore/dirs.c
index baebb2c..a35b00b 100644
--- a/sbin/restore/dirs.c
+++ b/sbin/restore/dirs.c
@@ -494,7 +494,8 @@ rst_readdir(RST_DIR *dirp)
 			return (NULL);
 		}
 		dirp->dd_loc += dp->d_reclen;
-		if (dp->d_ino == 0 && strcmp(dp->d_name, "/") == 0)
+		if (dp->d_ino == 0 && dp->d_namlen == 1 &&
+		    dp->d_name[0] == '/')
 			return (NULL);
 		if (dp->d_ino >= maxino) {
 			Dprintf(stderr, "corrupted directory: bad inum %llu\n",
