diff --git a/usr.sbin/hostapd/apme.c b/usr.sbin/hostapd/apme.c
index d478eac..d8df3a7 100644
--- a/usr.sbin/hostapd/apme.c
+++ b/usr.sbin/hostapd/apme.c
@@ -348,11 +348,11 @@ hostapd_apme_frame(struct hostapd_apme *apme, u_int8_t *buf, u_int len)
 
 	if ((offset = hostapd_apme_offset(apme, buf, len)) < 0)
 		return;
-	wh = (struct ieee80211_frame *)(buf + offset);
 
 	/* Ignore short frames or fragments */
-	if (len < sizeof(struct ieee80211_frame))
+	if (len - offset < sizeof(struct ieee80211_frame))
 		return;
+	wh = (struct ieee80211_frame *)(buf + offset);
 
 	/* Handle received frames */
 	if ((hostapd_handle_input(apme, buf, len) ==
