diff --git a/sbin/iked/dh.c b/sbin/iked/dh.c
index 0dd32ce..c7e7e1c 100644
--- a/sbin/iked/dh.c
+++ b/sbin/iked/dh.c
@@ -754,9 +754,12 @@ ec25519_create_exchange(struct dh_group *group, uint8_t *buf)
 int
 ec25519_create_shared(struct dh_group *group, uint8_t *shared, uint8_t *public)
 {
+	static const uint8_t	 zero[CURVE25519_SIZE];
 	struct curve25519_key	*curve25519 = group->curve25519;
 
 	crypto_scalarmult_curve25519(shared, curve25519->secret, public);
+	if (timingsafe_bcmp(shared, zero, CURVE25519_SIZE) == 0)
+		return (-1);
 	return (0);
 }
