diff --git a/sbin/unwind/libunbound/sldns/parse.c b/sbin/unwind/libunbound/sldns/parse.c
index 8ea0846..40987ab 100644
--- a/sbin/unwind/libunbound/sldns/parse.c
+++ b/sbin/unwind/libunbound/sldns/parse.c
@@ -513,7 +513,7 @@ sldns_bget_keyword_data(sldns_buffer *b, const char *keyword, const char *k_del,
                free(fkeyword);
                /* whee, the match! */
                /* retrieve it's data */
-               i = sldns_bget_token(b, data, d_del, 0);
+               i = sldns_bget_token(b, data, d_del, data_limit);
                return i;
        } else {
                free(fkeyword);
