diff --git a/usr.bin/gprof/elf.c b/usr.bin/gprof/elf.c
index 3233608..91ebc05 100644
--- a/usr.bin/gprof/elf.c
+++ b/usr.bin/gprof/elf.c
@@ -92,6 +92,7 @@ getnfile(const char *filename, char ***defaultEs)
     sh_strtab = &shdrs[sh_symtab->sh_link];
 
     if (sh_symtab->sh_offset >= s.st_size ||
+	sh_symtab->sh_size > s.st_size - sh_symtab->sh_offset ||
 	sh_symtab->sh_entsize == 0 ||
 	sh_strtab->sh_offset >= s.st_size)
 	errx(1, "%s: bad symbol table", filename);
