diff --git a/usr.sbin/snmpd/smi.c b/usr.sbin/snmpd/smi.c
index 0385b32..5c915a0 100644
--- a/usr.sbin/snmpd/smi.c
+++ b/usr.sbin/snmpd/smi.c
@@ -486,6 +486,8 @@ smi_print_element(struct ber_element *root)
 		case SNMP_T_IPADDR:
 			if (ober_get_string(root, &buf) == -1)
 				goto fail;
+			if (root->be_len != sizeof(struct in_addr))
+				goto fail;
 			if (asprintf(&str, "%s",
 			    inet_ntoa(*(struct in_addr *)buf)) == -1)
 					goto fail;
