diff --git a/sbin/isakmpd/ike_phase_1.c b/sbin/isakmpd/ike_phase_1.c
index d0638c4..699e6ac 100644
--- a/sbin/isakmpd/ike_phase_1.c
+++ b/sbin/isakmpd/ike_phase_1.c
@@ -1070,6 +1070,13 @@ ike_phase_1_recv_ID(struct message *msg)
 		}
 
 		/* Compare expected/desired and received remote ID */
+		if (GET_ISAKMP_GEN_LENGTH(payload->p) <
+		    ISAKMP_ID_DATA_OFF + sz) {
+			free(rid);
+			log_print("ike_phase_1_recv_ID: "
+			    "received remote ID shorter than expected %s", p);
+			return -1;
+		}
 		if (memcmp(rid, payload->p + ISAKMP_ID_DATA_OFF, sz) != 0) {
 			free(rid);
 			log_print("ike_phase_1_recv_ID: "
