diff --git a/lib/libkeynote/keynote-keygen.c b/lib/libkeynote/keynote-keygen.c
index edf013e..3087c55 100644
--- a/lib/libkeynote/keynote-keygen.c
+++ b/lib/libkeynote/keynote-keygen.c
@@ -99,7 +99,7 @@ keynote_keygen(int argc, char *argv[])
 {
     int begin = KEY_PRINT_OFFSET, prlen = KEY_PRINT_LENGTH;
     char *foo, *privalgname, seed[SEED_LEN];
-    int alg, enc, ienc, len = 0, counter;
+    int alg, enc, ienc, len = 0, counter, fd;
     struct keynote_deckey dc;
     unsigned long h;
     DSA *dsa;
@@ -321,10 +321,23 @@ keynote_keygen(int argc, char *argv[])
 	}
 	else
 	{
-	    fp = fopen(argv[4], "w");
+	    fd = open(argv[4], O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
+	    if (fd == -1)
+	    {
+		perror(argv[4]);
+		exit(1);
+	    }
+	    if (fchmod(fd, S_IRUSR | S_IWUSR) == -1)
+	    {
+		perror(argv[4]);
+		close(fd);
+		exit(1);
+	    }
+	    fp = fdopen(fd, "w");
 	    if (fp == NULL)
 	    {
 		perror(argv[4]);
+		close(fd);
 		exit(1);
 	    }
 	}
