diff --git a/sbin/isakmpd/dnssec.c b/sbin/isakmpd/dnssec.c
index d4857d3..8871b1f 100644
--- a/sbin/isakmpd/dnssec.c
+++ b/sbin/isakmpd/dnssec.c
@@ -192,6 +192,9 @@ dns_get_key(int type, struct message *msg, int *keylen)
 	 * first.
 	 */
 	for (i = 0; i < rr->rri_nrdatas && key_rr.datalen == 0; i++) {
+		if (rr->rri_rdatas[i].rdi_length < 4)
+			continue;
+
 		key_rr.flags = ntohs((u_int16_t) * rr->rri_rdatas[i].rdi_data);
 		key_rr.protocol = *(rr->rri_rdatas[i].rdi_data + 2);
 		key_rr.algorithm = *(rr->rri_rdatas[i].rdi_data + 3);
