diff --git a/usr.bin/indent/lexi.c b/usr.bin/indent/lexi.c
index 19b1105..f62ad58 100644
--- a/usr.bin/indent/lexi.c
+++ b/usr.bin/indent/lexi.c
@@ -139,7 +139,8 @@ lexi(void)
     }
 
     /* Scan an alphanumeric token */
-    if (chartype[(int)*buf_ptr] == alphanum ||
+    if (((unsigned char)*buf_ptr < sizeof(chartype) &&
+	chartype[(unsigned char)*buf_ptr] == alphanum) ||
 	(buf_ptr[0] == '.' && isdigit((unsigned char)buf_ptr[1]))) {
 	/*
 	 * we have a character or number
@@ -210,7 +211,8 @@ lexi(void)
 	    }
 	}
 	else
-	    while (chartype[(int)*buf_ptr] == alphanum) {	/* copy it over */
+	    while ((unsigned char)*buf_ptr < sizeof(chartype) &&
+		chartype[(unsigned char)*buf_ptr] == alphanum) {	/* copy it over */
 		CHECK_SIZE_TOKEN;
 		*e_token++ = *buf_ptr++;
 		if (buf_ptr >= buf_end)
