diff --git a/usr.sbin/ldomctl/mdesc.c b/usr.sbin/ldomctl/mdesc.c
index 3cae1f7..a111c34 100644
--- a/usr.sbin/ldomctl/mdesc.c
+++ b/usr.sbin/ldomctl/mdesc.c
@@ -476,6 +476,10 @@ md_ingest(void *buf, size_t size)
 		case MD_PROP_DATA:
 			if (node == NULL)
 				errx(1, "Corrupt MD");
+			if (betoh32(mde->d.y.data_offset) > data_blk_size ||
+			    betoh32(mde->d.y.data_len) > data_blk_size -
+			    betoh32(mde->d.y.data_offset))
+				errx(1, "Corrupt MD");
 			str = name_blk + betoh32(mde->name_offset);
 			data = data_blk + betoh32(mde->d.y.data_offset);
 			md_add_prop_data(md, node, str, data,
