diff --git a/usr.sbin/ldomctl/mdstore.c b/usr.sbin/ldomctl/mdstore.c
index 3592fca..ad284d6 100644
--- a/usr.sbin/ldomctl/mdstore.c
+++ b/usr.sbin/ldomctl/mdstore.c
@@ -18,6 +18,7 @@
 
 #include <assert.h>
 #include <err.h>
+#include <stddef.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -210,8 +211,14 @@ mdstore_rx_data(struct ldc_conn *lc, uint64_t svc_handle, void *data,
 {
 	struct mdstore_list_resp *mr = data;
 	struct mdstore_set *set;
+	char *end;
+	size_t payload_len;
 	int idx;
 
+	if (len < offsetof(struct mdstore_list_resp, result) +
+	    sizeof(mr->result))
+		goto out;
+
 	if (mr->result != MDST_SUCCESS) {
 		switch (mr->result) {
 		case MDST_SET_EXISTS_ERR:
@@ -228,13 +235,21 @@ mdstore_rx_data(struct ldc_conn *lc, uint64_t svc_handle, void *data,
 
 	switch (mdstore_command) {
 	case MDSET_LIST_REQUEST:
-		for (idx = 0, len = 0; len < mr->payload_len - 24; idx++) {
+		if (len < offsetof(struct mdstore_list_resp, sets) ||
+		    mr->payload_len < 24)
+			break;
+		payload_len = min(len - offsetof(struct mdstore_list_resp, sets),
+		    mr->payload_len - 24);
+		for (idx = 0, len = 0; len < payload_len; idx++) {
+			end = memchr(&mr->sets[len], '\0', payload_len - len);
+			if (end == NULL)
+				break;
 			set = xmalloc(sizeof(*set));
 			set->name = xstrdup(&mr->sets[len]);
 			set->booted_set = (idx == mr->booted_set);
 			set->boot_set = (idx == mr->boot_set);
 			TAILQ_INSERT_TAIL(&mdstore_sets, set, link);
-			len += strlen(&mr->sets[len]) + 1;
+			len += end - &mr->sets[len] + 1;
 			if (mdstore_major >= 2)
 				len += sizeof(uint64_t); /* skip timestamp */
 			if (mdstore_major >= 3)
@@ -243,6 +258,7 @@ mdstore_rx_data(struct ldc_conn *lc, uint64_t svc_handle, void *data,
 		break;
 	}
 
+out:
 	mdstore_command = 0;
 }
