diff --git a/usr.sbin/ldomd/ds.c b/usr.sbin/ldomd/ds.c
index 0090a60..e9f1203 100644
--- a/usr.sbin/ldomd/ds.c
+++ b/usr.sbin/ldomd/ds.c
@@ -592,6 +592,11 @@ ds_receive_msg(struct ldc_conn *lc, void *buf, size_t len)
 			return;
 		}
 
+		if ((lp.env & LDC_LEN_MASK) > len) {
+			ldc_reset(lc);
+			return;
+		}
+
 		bcopy(&lp.data, p, (lp.env & LDC_LEN_MASK));
 		p += (lp.env & LDC_LEN_MASK);
 		len -= (lp.env & LDC_LEN_MASK);
