diff --git a/usr.sbin/vmd/vioqcow2.c b/usr.sbin/vmd/vioqcow2.c
index 917cba2..290b771 100644
--- a/usr.sbin/vmd/vioqcow2.c
+++ b/usr.sbin/vmd/vioqcow2.c
@@ -589,7 +589,7 @@ copy_cluster(struct qcdisk *disk, struct qcdisk *base, off_t dst, off_t src)
 		fatal("out of memory");
 	src &= ~(disk->clustersz - 1);
 	dst &= ~(disk->clustersz - 1);
-	if (pread(base->fd, scratch, disk->clustersz, src) == -1)
+	if (pread(base->fd, scratch, disk->clustersz, src) != disk->clustersz)
 		fatal("%s: could not read cluster", __func__);
 	if (pwrite(disk->fd, scratch, disk->clustersz, dst) == -1)
 		fatal("%s: could not write cluster", __func__);
