diff --git a/lib/libkeynote/signature.c b/lib/libkeynote/signature.c
index d4123ca..a9b9086 100644
--- a/lib/libkeynote/signature.c
+++ b/lib/libkeynote/signature.c
@@ -518,6 +518,12 @@ kn_decode_key(struct keynote_deckey *dc, char *key, int keytype)
 
 	/* RSA-specific */
 	dc->dec_key = EVP_PKEY_get0_RSA(pPublicKey);
+	if (dc->dec_key == NULL) {
+	    free(ptr);
+	    X509_free(px509Cert);
+	    keynote_errno = ERROR_SYNTAX;
+	    return -1;
+	}
 	RSA_up_ref(dc->dec_key);
 
 	free(ptr);
