diff --git a/usr.sbin/npppd/npppd/npppd_ctl.c b/usr.sbin/npppd/npppd/npppd_ctl.c
index e67e568..53e0f2c 100644
--- a/usr.sbin/npppd/npppd/npppd_ctl.c
+++ b/usr.sbin/npppd/npppd/npppd_ctl.c
@@ -268,6 +268,7 @@ npppd_who_init(struct npppd_who *_this, npppd_ppp *ppp)
 	npppd_auth_base *realm = ppp->realm;
 	npppd_iface     *iface = ppp_iface(ppp);
 
+	memset(_this, 0, sizeof(*_this));
 	strlcpy(_this->username, ppp->username, sizeof(_this->username));
 	_this->time = ppp->start_time;
 	clock_gettime(CLOCK_MONOTONIC, &curr_time);
