diff --git a/usr.sbin/npppd/npppd/npppd_ctl.c b/usr.sbin/npppd/npppd/npppd_ctl.c
index e67e568..8c53b81 100644
--- a/usr.sbin/npppd/npppd/npppd_ctl.c
+++ b/usr.sbin/npppd/npppd/npppd_ctl.c
@@ -151,6 +151,7 @@ npppd_ctl_add_stopped_ppp(struct npppd_ctl *_this, npppd_ppp *ppp)
 		log_warn("malloc() failed in %s()", __func__);
 		return (-1);
 	}
+	memset(&stopped->ppp_who, 0, sizeof(stopped->ppp_who));
 	npppd_who_init(&stopped->ppp_who, ppp);
 	TAILQ_INSERT_TAIL(&_this->stopped_ppps, stopped, entry);
