diff --git a/usr.sbin/ospfd/lsupdate.c b/usr.sbin/ospfd/lsupdate.c
index 7a9872a..e34288f 100644
--- a/usr.sbin/ospfd/lsupdate.c
+++ b/usr.sbin/ospfd/lsupdate.c
@@ -269,7 +269,7 @@ recv_ls_update(struct nbr *nbr, char *buf, u_int16_t len)
 				return;
 			}
 			memcpy(&lsa, buf, sizeof(lsa));
-			if (len < ntohs(lsa.len)) {
+			if (ntohs(lsa.len) < sizeof(lsa) || len < ntohs(lsa.len)) {
 				log_warnx("recv_ls_update: bad packet size, "
 				    "neighbor ID %s (%s)", inet_ntoa(nbr->id),
 				    nbr->iface->name);
