diff --git a/usr.sbin/ypserv/common/yplib_host.c b/usr.sbin/ypserv/common/yplib_host.c
index 4a42107..6cac9a7 100644
--- a/usr.sbin/ypserv/common/yplib_host.c
+++ b/usr.sbin/ypserv/common/yplib_host.c
@@ -147,8 +147,13 @@ yp_match_host(CLIENT *client, char *indomain, char *inmap, const char *inkey,
 	if ( !(r = ypprot_err(yprv.stat)) ) {
 		*outvallen = yprv.val.valdat_len;
 		*outval = malloc(*outvallen+1);
-		memcpy(*outval, yprv.val.valdat_val, *outvallen);
-		(*outval)[*outvallen] = '\0';
+		if (*outval == NULL) {
+			*outvallen = 0;
+			r = YPERR_RESRC;
+		} else {
+			memcpy(*outval, yprv.val.valdat_val, *outvallen);
+			(*outval)[*outvallen] = '\0';
+		}
 	}
 	xdr_free(xdr_ypresp_val, (char *)&yprv);
 	return r;
