diff --git a/usr.bin/ctfconv/elf.c b/usr.bin/ctfconv/elf.c
index 043939a..ecbc188 100644
--- a/usr.bin/ctfconv/elf.c
+++ b/usr.bin/ctfconv/elf.c
@@ -195,7 +195,8 @@ elf_getsection(char *p, size_t filesize, const char *sname, const char *shstab,
 			continue;
 
 		sh = (Elf_Shdr *)(p + shoff);
-		if ((sh->sh_link >= eh->e_shnum) || (sh->sh_name >= shstabsz))
+		if ((sh->sh_link >= eh->e_shnum) || (sh->sh_name >= shstabsz) ||
+		    (snlen > shstabsz - sh->sh_name))
 			continue;
 
 		if (sh->sh_offset > filesize)
