diff --git a/usr.bin/locate/locate/fastfind.c b/usr.bin/locate/locate/fastfind.c
index aec9078..778dddc 100644
--- a/usr.bin/locate/locate/fastfind.c
+++ b/usr.bin/locate/locate/fastfind.c
@@ -197,6 +197,7 @@ fastfind_mmap
 				if (c == cc)
 #endif /* FF_ICASE */
 					foundchar = p;
+				sane_count(p - path + 1);
 				*p++ = c;
 			} else {
 				/* bigrams are parity-marked */
@@ -210,6 +211,7 @@ fastfind_mmap
 #endif /* FF_ICASE */
 					foundchar = p + 1;
 
+				sane_count(p - path + 2);
 				*p++ = bigram1[c];
 				*p++ = bigram2[c];
 			}
