diff --git a/usr.bin/rsync/ids.c b/usr.bin/rsync/ids.c
index 1eef94b..97c00e9 100644
--- a/usr.bin/rsync/ids.c
+++ b/usr.bin/rsync/ids.c
@@ -299,7 +299,10 @@ idents_recv(struct sess *sess,
 		} else if (sz == 0)
 			WARNX("zero-length name in identifier list");
 
-		assert(id < INT32_MAX);
+		if (id >= INT32_MAX) {
+			ERRX("identifier id too large: %u", id);
+			return 0;
+		}
 		(*ids)[*idsz].id = id;
 		(*ids)[*idsz].name = calloc(sz + 1, 1);
 		if ((*ids)[*idsz].name == NULL) {
