diff --git a/usr.bin/signify/zsig.c b/usr.bin/signify/zsig.c
index 507517e..5ef3ae9 100644
--- a/usr.bin/signify/zsig.c
+++ b/usr.bin/signify/zsig.c
@@ -47,6 +47,7 @@ struct gzheader {
 #define FCOMMENT_FLAG 16
 
 #define GZHEADERLENGTH 10
+#define GZHEADERMAX 65536LU
 #define MYBUFSIZE 65536LU
 
 
@@ -67,7 +68,11 @@ readgz_header(struct gzheader *h, int fd)
 
 	while (1) {
 		if (len == sz) {
+			if (sz >= GZHEADERMAX)
+				errx(1, "gzheader too long");
 			sz *= 2;
+			if (sz > GZHEADERMAX)
+				sz = GZHEADERMAX;
 			buf = realloc(buf, sz);
 			if (!buf)
 				err(1, "realloc");
