diff --git a/usr.sbin/ldapd/syntax.c b/usr.sbin/ldapd/syntax.c
index de0eb8e..a22edaa 100644
--- a/usr.sbin/ldapd/syntax.c
+++ b/usr.sbin/ldapd/syntax.c
@@ -265,10 +265,12 @@ syntax_is_time(struct schema *schema, char *value, size_t len, int gen)
 {
 	int	 n;
 	char	*p = value;
+	char	*end = value + len;
 
 #define CHECK_RANGE(min, max) \
 	do {						\
-		if (!isdigit((unsigned char)p[0]) ||	\
+		if (end - p < 2 ||			\
+		    !isdigit((unsigned char)p[0]) ||	\
 		    !isdigit((unsigned char)p[1]))	\
 			return 0;			\
 		n = (p[0] - '0') * 10 + (p[1] - '0');	\
@@ -285,27 +287,27 @@ syntax_is_time(struct schema *schema, char *value, size_t len, int gen)
 	/* FIXME: should check number of days in month */
 	CHECK_RANGE(0, 23);			/* hour */
 
-	if (!gen || isdigit((unsigned char)*p)) {
+	if (!gen || (p < end && isdigit((unsigned char)*p))) {
 		CHECK_RANGE(0, 59);		/* minute */
-		if (isdigit((unsigned char)*p))
+		if (p < end && isdigit((unsigned char)*p))
 			CHECK_RANGE(0, 59+gen);	/* second or leap-second */
-		if (!gen && *p == '\0')
+		if (!gen && p == end)
 			return 1;
 	}
 						/* fraction */
-	if (!gen && ((*p == ',' || *p == '.') &&
-	    !isdigit((unsigned char)*++p)))
+	if (!gen && p < end && ((*p == ',' || *p == '.') &&
+	    (++p == end || !isdigit((unsigned char)*p))))
 		return 0;
 
-	if (*p == '-' || *p == '+') {
+	if (p < end && (*p == '-' || *p == '+')) {
 		++p;
 		CHECK_RANGE(0, 23);		/* hour */
-		if (!gen || isdigit((unsigned char)*p))
+		if (!gen || (p < end && isdigit((unsigned char)*p)))
 			CHECK_RANGE(0, 59);	/* minute */
-	} else if (*p++ != 'Z')
+	} else if (p == end || *p++ != 'Z')
 		return 0;
 
-	return *p == '\0';
+	return p == end;
 }
 
 static int
