diff --git a/lib/common/http1client.c b/lib/common/http1client.c
index 3d404a4a2..e1630797d 100644
--- a/lib/common/http1client.c
+++ b/lib/common/http1client.c
@@ -28,6 +28,7 @@
 #include <sys/un.h>
 #include "picohttpparser.h"
 #include "h2o/httpclient.h"
+#include "h2o/string_.h"
 #include "h2o/token.h"
 
 #if !H2O_USE_LIBUV && defined(__linux__)
@@ -736,13 +737,21 @@ static void on_send_timeout(h2o_timer_t *entry)
     on_error(client, h2o_httpclient_error_io_timeout);
 }
 
+static h2o_iovec_t escape_lf_in_request_target(h2o_mem_pool_t *pool, h2o_iovec_t input)
+{
+    if (memchr(input.base, '\r', input.len) == NULL && memchr(input.base, '\n', input.len) == NULL)
+        return input;
+    return h2o_uri_escape(pool, input.base, input.len, "/:?@!$&'()*+,;=%[]");
+}
+
 static h2o_iovec_t build_request(struct st_h2o_http1client_t *client, h2o_iovec_t method, const h2o_url_t *url,
                                  h2o_iovec_t connection, const h2o_header_t *headers, size_t num_headers)
 {
-    h2o_iovec_t buf;
+    h2o_iovec_t buf, path = escape_lf_in_request_target(client->super.pool, url->path),
+                     authority = escape_lf_in_request_target(client->super.pool, url->authority);
     size_t offset = 0;
 
-    buf.len = method.len + url->path.len + url->authority.len + 512;
+    buf.len = method.len + path.len + authority.len + 512;
     buf.base = h2o_mem_alloc_pool(client->super.pool, char, buf.len);
 
 #define RESERVE(sz)                                                                                                                \
@@ -779,16 +788,16 @@ static h2o_iovec_t build_request(struct st_h2o_http1client_t *client, h2o_iovec_
     if (client->super.upgrade_to == h2o_httpclient_upgrade_to_connect) {
         if (h2o_memis(method.base, method.len, H2O_STRLIT("CONNECT-UDP"))) {
             APPEND_STRLIT("masque://");
-            APPEND(url->authority.base, url->authority.len);
+            APPEND(authority.base, authority.len);
             APPEND_STRLIT("/");
         } else {
-            APPEND(url->authority.base, url->authority.len);
+            APPEND(authority.base, authority.len);
         }
     } else {
-        APPEND(url->path.base, url->path.len);
+        APPEND(path.base, path.len);
     }
     APPEND_STRLIT(" HTTP/1.1\r\nhost: ");
-    APPEND(url->authority.base, url->authority.len);
+    APPEND(authority.base, authority.len);
     buf.base[offset++] = '\r';
     buf.base[offset++] = '\n';
     assert(offset <= buf.len);
