diff --git a/lib/handler/configurator/fastcgi.c b/lib/handler/configurator/fastcgi.c
index 15331e731..a2d31fc84 100644
--- a/lib/handler/configurator/fastcgi.c
+++ b/lib/handler/configurator/fastcgi.c
@@ -314,7 +314,7 @@ static int on_config_spawn(h2o_configurator_command_t *cmd, h2o_configurator_con
     ret = 0;
 Exit:
     if (dirname[0] != '\0')
-        unlink(dirname);
+        rmdir(dirname);
     free(kill_on_close_cmd_path);
     free(setuidgid_cmd_path);
     return ret;
