diff --git a/lib/handler/access_log.c b/lib/handler/access_log.c
index e5ce75c2b..85952f828 100644
--- a/lib/handler/access_log.c
+++ b/lib/handler/access_log.c
@@ -45,19 +45,33 @@ static void log_access(h2o_logger_t *_self, h2o_req_t *req)
 {
     struct st_h2o_access_logger_t *self = (struct st_h2o_access_logger_t *)_self;
     h2o_access_log_filehandle_t *fh = self->fh;
-    char *logline, buf[4096];
+    char *logline, *logline_start, buf[4096];
     size_t len;
 
     /* stringify */
     len = sizeof(buf);
     logline = h2o_log_request(fh->logconf, req, &len, buf);
+    logline_start = logline;
 
     /* emit */
-    write(fh->fd, logline, len);
+    while (len != 0) {
+        ssize_t wrote = write(fh->fd, logline, len);
+        if (wrote > 0) {
+            logline += wrote;
+            len -= wrote;
+            continue;
+        }
+        if (wrote == -1 && errno == EINTR)
+            continue;
+        if (wrote == 0)
+            errno = EIO;
+        h2o_error_printf("failed to write access log:%s\n", strerror(errno));
+        break;
+    }
 
     /* free memory */
-    if (logline != buf)
-        free(logline);
+    if (logline_start != buf)
+        free(logline_start);
 }
 
 static void on_dispose_handle(void *_fh)
