diff --git a/lib/http1.c b/lib/http1.c
index c6e9c1e7b..39cc82967 100644
--- a/lib/http1.c
+++ b/lib/http1.c
@@ -555,6 +555,14 @@ static int contains_crlf_only(const char *s, size_t len)
     return 1;
 }
 
+static int contains_crlf(const char *s, size_t len)
+{
+    for (; len != 0; ++s, --len)
+        if (*s == '\r' || *s == '\n')
+            return 1;
+    return 0;
+}
+
 static void send_bad_request(struct st_h2o_http1_conn_t *conn, const char *body)
 {
     h2o_socket_read_stop(conn->sock);
@@ -1141,12 +1149,20 @@ static void finalostream_send_informational(h2o_ostream_t *_self, h2o_req_t *req
     struct st_h2o_http1_conn_t *conn = (struct st_h2o_http1_conn_t *)req->conn;
     assert(_self == &conn->_ostr_final.super);
 
+    if (contains_crlf(req->res.reason, strlen(req->res.reason)))
+        return;
+
     size_t len = sizeof("HTTP/1.1  \r\n\r\n") + 3 + strlen(req->res.reason) - 1;
     h2o_iovec_t buf = h2o_iovec_init(NULL, len);
 
     int i;
-    for (i = 0; i != req->res.headers.size; ++i)
-        buf.len += req->res.headers.entries[i].name->len + req->res.headers.entries[i].value.len + 4;
+    for (i = 0; i != req->res.headers.size; ++i) {
+        const h2o_header_t *header = req->res.headers.entries + i;
+        const h2o_iovec_t *name = header->name;
+        if (contains_crlf(name->base, name->len) || contains_crlf(header->value.base, header->value.len))
+            return;
+        buf.len += name->len + header->value.len + 4;
+    }
 
     buf.base = h2o_mem_alloc_pool(&req->pool, char, buf.len);
     char *dst = buf.base;
