diff --git a/lib/common/memcached.c b/lib/common/memcached.c
index 8dbddabf4..3a0731eab 100644
--- a/lib/common/memcached.c
+++ b/lib/common/memcached.c
@@ -296,15 +296,9 @@ static void reader_main(h2o_memcached_context_t *ctx)
     pthread_mutex_unlock(&conn.ctx->mutex);
     pthread_join(writer_thread, NULL);
 
-    /* decrement num_threads_connected, and discard all the pending requests if no connections are alive */
+    /* decrement num_threads_connected */
     pthread_mutex_lock(&conn.ctx->mutex);
-    if (--conn.ctx->num_threads_connected == 0) {
-        while (!h2o_linklist_is_empty(&conn.ctx->pending)) {
-            h2o_memcached_req_t *req = H2O_STRUCT_FROM_MEMBER(h2o_memcached_req_t, pending, conn.ctx->pending.next);
-            h2o_linklist_unlink(&req->pending);
-            discard_req(req);
-        }
-    }
+    --conn.ctx->num_threads_connected;
     pthread_mutex_unlock(&conn.ctx->mutex);
 
     /* close the connection */
@@ -324,12 +318,8 @@ static void dispatch(h2o_memcached_context_t *ctx, h2o_memcached_req_t *req)
 {
     pthread_mutex_lock(&ctx->mutex);
 
-    if (ctx->num_threads_connected != 0) {
-        h2o_linklist_insert(&ctx->pending, &req->pending);
-        pthread_cond_signal(&ctx->cond);
-    } else {
-        discard_req(req);
-    }
+    h2o_linklist_insert(&ctx->pending, &req->pending);
+    pthread_cond_signal(&ctx->cond);
 
     pthread_mutex_unlock(&ctx->mutex);
 }
