diff --git a/lib/common/socket/uv-binding.c.h b/lib/common/socket/uv-binding.c.h
index e53741f81..142fad98b 100644
--- a/lib/common/socket/uv-binding.c.h
+++ b/lib/common/socket/uv-binding.c.h
@@ -350,6 +350,7 @@ h2o_socket_t *do_import(h2o_loop_t *loop, h2o_socket_export_t *info)
     if (sock == NULL)
         return NULL;
     if (uv_tcp_open((uv_tcp_t *)sock->handle, info->fd) != 0) {
+        close(info->fd);
         h2o_socket_close(&sock->super);
         return NULL;
     }
