diff --git a/lib/common/socket/uv-binding.c.h b/lib/common/socket/uv-binding.c.h
index e53741f81..0a009c11d 100644
--- a/lib/common/socket/uv-binding.c.h
+++ b/lib/common/socket/uv-binding.c.h
@@ -39,6 +39,7 @@ struct st_h2o_uv_socket_t {
 };
 
 static void do_ssl_write(struct st_h2o_uv_socket_t *sock, int is_first_call, h2o_iovec_t *initial_bufs, size_t initial_bufcnt);
+static void on_ssl_write_complete(uv_write_t *wreq, int status);
 
 static void alloc_inbuf(h2o_buffer_t **buf, uv_buf_t *_vec)
 {
@@ -208,12 +209,24 @@ static void on_call_write_error(h2o_timer_t *timer)
     on_do_write_complete(&sock->stream._wreq, 1);
 }
 
+static void on_call_ssl_write_error(h2o_timer_t *timer)
+{
+    struct st_h2o_uv_socket_t *sock = H2O_STRUCT_FROM_MEMBER(struct st_h2o_uv_socket_t, write_cb_timer, timer);
+    on_ssl_write_complete(&sock->stream._wreq, 1);
+}
+
 static void call_write_complete_delayed(struct st_h2o_uv_socket_t *sock, int status)
 {
     sock->write_cb_timer.cb = status == 0 ? on_call_write_success : on_call_write_error;
     h2o_timer_link(sock->handle->loop, 0, &sock->write_cb_timer);
 }
 
+static void call_ssl_write_complete_delayed(struct st_h2o_uv_socket_t *sock, int status)
+{
+    sock->write_cb_timer.cb = status == 0 ? on_call_write_success : on_call_ssl_write_error;
+    h2o_timer_link(sock->handle->loop, 0, &sock->write_cb_timer);
+}
+
 void report_early_write_error(h2o_socket_t *_sock)
 {
     struct st_h2o_uv_socket_t *sock = (struct st_h2o_uv_socket_t *)_sock;
@@ -284,7 +297,8 @@ void do_ssl_write(struct st_h2o_uv_socket_t *sock, int is_first_call, h2o_iovec_
 
     /* Send pending TLS records. */
     uv_buf_t uvbuf = {(char *)sock->super.ssl->output.buf.base, sock->super.ssl->output.buf.off};
-    uv_write(&sock->stream._wreq, (uv_stream_t *)sock->handle, &uvbuf, 1, on_ssl_write_complete);
+    if (uv_write(&sock->stream._wreq, (uv_stream_t *)sock->handle, &uvbuf, 1, on_ssl_write_complete) != 0)
+        call_ssl_write_complete_delayed(sock, 1);
 }
 
 void do_write(h2o_socket_t *_sock, h2o_iovec_t *bufs, size_t bufcnt)
@@ -294,7 +308,8 @@ void do_write(h2o_socket_t *_sock, h2o_iovec_t *bufs, size_t bufcnt)
 
     if (sock->super.ssl == NULL) {
         if (bufcnt > 0) {
-            uv_write(&sock->stream._wreq, (uv_stream_t *)sock->handle, (uv_buf_t *)bufs, (int)bufcnt, on_do_write_complete);
+            if (uv_write(&sock->stream._wreq, (uv_stream_t *)sock->handle, (uv_buf_t *)bufs, (int)bufcnt, on_do_write_complete) != 0)
+                call_write_complete_delayed(sock, 1);
         } else {
             call_write_complete_delayed(sock, 0);
         }
