diff --git a/lib/common/url.c b/lib/common/url.c
index 8b0c8258b..3dab7d022 100644
--- a/lib/common/url.c
+++ b/lib/common/url.c
@@ -225,7 +225,9 @@ const char *h2o_url_parse_hostport(const char *s, size_t len, h2o_iovec_t *host,
     /* parse port */
     if (token_start != end && *token_start == ':') {
         uint32_t p = 0;
-        for (++token_start; token_start != end; ++token_start) {
+        if (++token_start == end || !('0' <= *token_start && *token_start <= '9'))
+            return NULL;
+        for (; token_start != end; ++token_start) {
             if ('0' <= *token_start && *token_start <= '9') {
                 p = p * 10 + *token_start - '0';
                 if (p >= 65535)
