diff --git a/lib/handler/mruby/redis.c b/lib/handler/mruby/redis.c
index e3ee0fa55..8130bafb0 100644
--- a/lib/handler/mruby/redis.c
+++ b/lib/handler/mruby/redis.c
@@ -233,6 +233,8 @@ static void on_redis_command(redisReply *_reply, void *_ctx, const char *errstr)
 
     mrb_gc_arena_restore(mrb, gc_arena);
     mrb_gc_unregister(mrb, ctx->refs.command);
+    DATA_PTR(mrb_obj_ptr(ctx->refs.command)) = NULL;
+    on_gc_dispose_command(mrb, ctx);
 }
 
 static mrb_value call_method(mrb_state *mrb, mrb_value self)
