diff --git a/modules/caddyhttp/caddyauth/basicauth.go b/modules/caddyhttp/caddyauth/basicauth.go
index 4152d790..f6b221b2 100644
--- a/modules/caddyhttp/caddyauth/basicauth.go
+++ b/modules/caddyhttp/caddyauth/basicauth.go
@@ -15,8 +15,8 @@
 package caddyauth
 
 import (
+	"crypto/sha256"
 	"encoding/base64"
-	"encoding/hex"
 	"encoding/json"
 	"fmt"
 	weakrand "math/rand/v2"
@@ -172,8 +172,9 @@ func (hba HTTPBasicAuth) correctPassword(account Account, plaintextPassword []by
 		return compare()
 	}
 
-	// compute a cache key that is unique for these input parameters
-	cacheKey := hex.EncodeToString(append(account.password, plaintextPassword...))
+	// compute a fixed-size cache key that is unique for these input parameters
+	cacheKeyHash := sha256.Sum256(append(account.password, plaintextPassword...))
+	cacheKey := string(cacheKeyHash[:])
 
 	// fast track: if the result of the input is already cached, use it
 	hba.HashCache.mu.RLock()
