diff --git a/modules/caddytls/storageloader.go b/modules/caddytls/storageloader.go
index c9487e89..b41c8969 100644
--- a/modules/caddytls/storageloader.go
+++ b/modules/caddytls/storageloader.go
@@ -93,7 +93,7 @@ func (sl StorageLoader) LoadCertificates() ([]Certificate, error) {
 		case "pem":
 			// if the start of the key file looks like an encrypted private key,
 			// reject it with a helpful error message
-			if strings.Contains(string(keyData[:40]), "ENCRYPTED") {
+			if strings.Contains(string(keyData), "ENCRYPTED") {
 				return nil, fmt.Errorf("encrypted private keys are not supported; please decrypt the key first")
 			}

