diff --git a/std/experimental/allocator/building_blocks/region.d b/std/experimental/allocator/building_blocks/region.d
index 736b1858e..6d490dcd0 100644
--- a/std/experimental/allocator/building_blocks/region.d
+++ b/std/experimental/allocator/building_blocks/region.d
@@ -1097,7 +1097,10 @@ version (Posix) struct SbrkRegion(uint minAlign = platformAlignment)
         pthread_mutex_lock(cast(pthread_mutex_t*) &sbrkMutex) == 0 || assert(0);
         scope(exit) pthread_mutex_unlock(cast(pthread_mutex_t*) &sbrkMutex) == 0
             || assert(0);
-        return !_brkInitial || brk(_brkInitial) == 0;
+        if (!_brkInitial) return true;
+        if (brk(_brkInitial) != 0) return false;
+        _brkCurrent = _brkInitial;
+        return true;
     }
 
     /// Standard allocator API.
