diff --git a/src/cdc/amqp/spec.zig b/src/cdc/amqp/spec.zig
index f0927ff9a..183a408e8 100644
--- a/src/cdc/amqp/spec.zig
+++ b/src/cdc/amqp/spec.zig
@@ -793,7 +793,7 @@ pub const ClientMethod = union(ClientMethod.Tag) {
 
     pub fn decode(header: MethodHeader, decoder: *Decoder) Decoder.Error!ClientMethod {
         @setEvalBranchQuota(10_000);
-        const tag: Tag = @enumFromInt(@as(u32, @bitCast(header)));
+        const tag = std.meta.intToEnum(Tag, @as(u32, @bitCast(header))) catch return error.Unexpected;
         const value: ClientMethod = switch (tag) {
             inline else => |tag_comptime| value: {
                 const Method = std.meta.TagPayload(ClientMethod, tag_comptime);
