diff --git a/crypto/asn1/a_bitstr.cc b/crypto/asn1/a_bitstr.cc
index b7394c071..e346349ea 100644
--- a/crypto/asn1/a_bitstr.cc
+++ b/crypto/asn1/a_bitstr.cc
@@ -212,6 +212,10 @@ int ASN1_BIT_STRING_set_bit(ASN1_BIT_STRING *a, int n, int value) {
   int w, v, iv;
   unsigned char *c;
 
+  if (a == nullptr || n < 0) {
+    return 0;
+  }
+
   w = n / 8;
   v = 1 << (7 - (n & 0x07));
   iv = ~v;
@@ -219,10 +223,6 @@ int ASN1_BIT_STRING_set_bit(ASN1_BIT_STRING *a, int n, int value) {
     v = 0;
   }
 
-  if (a == nullptr) {
-    return 0;
-  }
-
   a->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);  // clear, set on write
 
   if ((a->length < (w + 1)) || (a->data == nullptr)) {
@@ -253,9 +253,13 @@ int ASN1_BIT_STRING_set_bit(ASN1_BIT_STRING *a, int n, int value) {
 int ASN1_BIT_STRING_get_bit(const ASN1_BIT_STRING *a, int n) {
   int w, v;
 
+  if (a == nullptr || n < 0) {
+    return 0;
+  }
+
   w = n / 8;
   v = 1 << (7 - (n & 0x07));
-  if ((a == nullptr) || (a->length < (w + 1)) || (a->data == nullptr)) {
+  if ((a->length < (w + 1)) || (a->data == nullptr)) {
     return 0;
   }
   return ((a->data[w] & v) != 0);
