diff --git a/crypto/cipher/e_rc2.cc b/crypto/cipher/e_rc2.cc
index 91c5f52d6..98a868e9a 100644
--- a/crypto/cipher/e_rc2.cc
+++ b/crypto/cipher/e_rc2.cc
@@ -356,6 +356,10 @@ static int rc2_cbc_cipher_update(EVP_CIPHER_CTX *ctx, uint8_t *out,
   EVP_RC2_KEY *key = (EVP_RC2_KEY *)ctx->cipher_data;
   static const size_t kChunkSize = 0x10000;
 
+  if (!ctx->encrypt && len % ctx->cipher->block_size != 0) {
+    return 0;
+  }
+
   while (len >= kChunkSize) {
     RC2_cbc_encrypt(in, out, kChunkSize, &key->ks, ctx->iv, ctx->encrypt);
     len -= kChunkSize;
