diff --git a/crypto/evp/evp.cc b/crypto/evp/evp.cc
index 7581c3946..ee1c55e96 100644
--- a/crypto/evp/evp.cc
+++ b/crypto/evp/evp.cc
@@ -288,7 +288,8 @@ int EVP_PKEY_get_raw_private_key(const EVP_PKEY *pkey, uint8_t *out,
                                  size_t *out_len) {
   auto *impl = FromOpaque(pkey);
 
-  if (impl->ameth->get_priv_raw == nullptr) {
+  if (impl == nullptr || impl->ameth == nullptr ||
+      impl->ameth->get_priv_raw == nullptr) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
     return 0;
   }
@@ -300,7 +301,8 @@ int EVP_PKEY_get_private_seed(const EVP_PKEY *pkey, uint8_t *out,
                               size_t *out_len) {
   auto *impl = FromOpaque(pkey);
 
-  if (impl->ameth->get_priv_seed == nullptr) {
+  if (impl == nullptr || impl->ameth == nullptr ||
+      impl->ameth->get_priv_seed == nullptr) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
     return 0;
   }
@@ -312,7 +314,8 @@ int EVP_PKEY_get_raw_public_key(const EVP_PKEY *pkey, uint8_t *out,
                                 size_t *out_len) {
   auto *impl = FromOpaque(pkey);
 
-  if (impl->ameth->get_pub_raw == nullptr) {
+  if (impl == nullptr || impl->ameth == nullptr ||
+      impl->ameth->get_pub_raw == nullptr) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
     return 0;
   }
@@ -385,7 +388,8 @@ int EVP_PKEY_set1_tls_encodedpoint(EVP_PKEY *pkey, const uint8_t *in,
                                    size_t len) {
   auto *impl = FromOpaque(pkey);
 
-  if (impl->ameth->set1_tls_encodedpoint == nullptr) {
+  if (impl == nullptr || impl->ameth == nullptr ||
+      impl->ameth->set1_tls_encodedpoint == nullptr) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
     return 0;
   }
@@ -396,7 +400,8 @@ int EVP_PKEY_set1_tls_encodedpoint(EVP_PKEY *pkey, const uint8_t *in,
 size_t EVP_PKEY_get1_tls_encodedpoint(const EVP_PKEY *pkey, uint8_t **out_ptr) {
   auto *impl = FromOpaque(pkey);
 
-  if (impl->ameth->get1_tls_encodedpoint == nullptr) {
+  if (impl == nullptr || impl->ameth == nullptr ||
+      impl->ameth->get1_tls_encodedpoint == nullptr) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
     return 0;
   }
