diff --git a/crypto/x509/t_crl.cc b/crypto/x509/t_crl.cc
index 5ef99ab38..1b8952cad 100644
--- a/crypto/x509/t_crl.cc
+++ b/crypto/x509/t_crl.cc
@@ -99,6 +99,7 @@ int X509_CRL_print(BIO *out, const X509_CRL *x) {
         BIO_printf(out, "\n") <= 0 ||
         !X509V3_extensions_print(out, "CRL entry extensions",
                                  X509_REVOKED_get0_extensions(r), 0, 8)) {
+      return 0;
     }
   }
