diff --git a/lib/fusion.c b/lib/fusion.c
index 62a93e6..6d741ce 100644
--- a/lib/fusion.c
+++ b/lib/fusion.c
@@ -1202,6 +1202,8 @@ static int aesgcm_setup(ptls_aead_context_t *_ctx, int is_enc, const void *key,
     ctx->super.do_decrypt = aead_do_decrypt;
 
     ctx->aesgcm = new_aesgcm(key, key_size, 1500 /* assume ordinary packet size */, 0 /* no support for aesni256 yet */);
+    if (ctx->aesgcm == NULL)
+        return PTLS_ERROR_NO_MEMORY;
 
     return 0;
 }
@@ -2138,6 +2140,8 @@ static int non_temporal_setup(ptls_aead_context_t *_ctx, int is_enc, const void
         new_aesgcm(key, key_size,
                    7 * (ptls_fusion_can_aesni256 ? 32 : 16), // 6 blocks at once, plus len(A) | len(C) that we might append
                    aesni256);
+    if (ctx->aesgcm == NULL)
+        return PTLS_ERROR_NO_MEMORY;
 
     return 0;
 }
