diff --git a/lib/hpke.c b/lib/hpke.c
index b81f19c..6617a7e 100644
--- a/lib/hpke.c
+++ b/lib/hpke.c
@@ -212,6 +212,8 @@ static int key_schedule(ptls_hpke_kem_t *kem, ptls_hpke_cipher_suite_t *cipher,
         goto Exit;
 
     *ctx = ptls_aead_new_direct(cipher->aead, is_enc, key, base_nonce);
+    if (*ctx == NULL)
+        ret = PTLS_ERROR_NO_MEMORY;
 
 Exit:
     ptls_buffer_dispose(&key_schedule_context);
