diff --git a/bn/bn_exp.c b/bn/bn_exp.c
index 6a5c1c8..959f3e9 100644
--- a/bn/bn_exp.c
+++ b/bn/bn_exp.c
@@ -1122,6 +1122,9 @@ BN_mod_exp_internal(BIGNUM *r, const BIGNUM *a, const BIGNUM *p, const BIGNUM *m
 			ret = BN_mod_exp_mont_word(r, A,p, m,ctx, NULL);
 		} else
 			ret = BN_mod_exp_mont_ct(r, a,p, m,ctx, NULL);
+	} else if (ct) {
+		BNerror(BN_R_CALLED_WITH_EVEN_MODULUS);
+		return (0);
 	} else	{
 		ret = BN_mod_exp_reciprocal(r, a,p, m, ctx);
 	}
