diff --git a/x509/x509_verify.c b/x509/x509_verify.c
index fc3fbc1..306b000 100644
--- a/x509/x509_verify.c
+++ b/x509/x509_verify.c
@@ -911,7 +911,7 @@ x509_verify_cert_extensions(struct x509_verify_ctx *ctx, X509 *cert, int need_ca
 		ctx->error = X509_V_ERR_INVALID_CA;
 		return 0;
 	}
-	if (ctx->purpose > 0 && X509_check_purpose(cert, ctx->purpose, need_ca)) {
+	if (ctx->purpose > 0 && X509_check_purpose(cert, ctx->purpose, need_ca) <= 0) {
 		ctx->error = X509_V_ERR_INVALID_PURPOSE;
 		return 0;
 	}
