diff --git a/rust/src/lib.rs b/rust/src/lib.rs
index 929d535..6307648 100644
--- a/rust/src/lib.rs
+++ b/rust/src/lib.rs
@@ -146,15 +146,19 @@ pub fn sign_url(
     } else {
         format!("&{url_data}")
     };
+    let fragment = parsed
+        .fragment()
+        .map(|fragment| format!("#{fragment}"))
+        .unwrap_or_default();
 
     if is_directory {
         Ok(format!(
-            "{base}/bcdn_token={token}{tail}&expires={expires}{}",
+            "{base}/bcdn_token={token}{tail}&expires={expires}{}{fragment}",
             parsed.path()
         ))
     } else {
         Ok(format!(
-            "{base}{}?token={token}{tail}&expires={expires}",
+            "{base}{}?token={token}{tail}&expires={expires}{fragment}",
             parsed.path()
         ))
     }
