diff --git a/lib/wasix/src/syscalls/wasix/proc_fork.rs b/lib/wasix/src/syscalls/wasix/proc_fork.rs
index 0b282bec72c..ad815533677 100644
--- a/lib/wasix/src/syscalls/wasix/proc_fork.rs
+++ b/lib/wasix/src/syscalls/wasix/proc_fork.rs
@@ -221,20 +221,40 @@ pub fn proc_fork<M: MemorySize>(
                 run::<M>(ctx, store, child_handle, None);
             };
 
-            tasks_outer
-                .task_wasm(
-                    TaskWasm::new(Box::new(run), child_env, module, false, false)
-                        .with_globals(snapshot)
-                        .with_memory(spawn_type),
-                )
-                .map_err(|err| {
+            match tasks_outer.task_wasm(
+                TaskWasm::new(Box::new(run), child_env, module, false, false)
+                    .with_globals(snapshot)
+                    .with_memory(spawn_type),
+            ) {
+                Ok(()) => {}
+                Err(err) => {
                     warn!(
                         "failed to fork as the process could not be spawned - {}",
                         err
                     );
-                    err
-                })
-                .ok();
+                    let spawn_err = Errno::from(err);
+                    {
+                        let mut inner = ctx.data().process.lock();
+                        inner.children.retain(|child| child.pid() != child_pid);
+                    }
+                    return match rewind::<M, _>(
+                        ctx,
+                        Some(memory_stack),
+                        rewind_stack,
+                        store_data,
+                        ForkResult {
+                            pid: 0,
+                            ret: spawn_err,
+                        },
+                    ) {
+                        Errno::Success => OnCalledAction::InvokeAgain,
+                        err => {
+                            warn!("failed - could not rewind the stack - errno={}", err);
+                            OnCalledAction::Trap(Box::new(WasiError::Exit(err.into())))
+                        }
+                    };
+                }
+            }
         };
 
         // Rewind the stack and carry on
