diff --git a/lib/wasix/src/syscalls/wasi/fd_write.rs b/lib/wasix/src/syscalls/wasi/fd_write.rs
index 68e50b52290..d81813c0a8e 100644
--- a/lib/wasix/src/syscalls/wasi/fd_write.rs
+++ b/lib/wasix/src/syscalls/wasi/fd_write.rs
@@ -418,9 +418,10 @@ pub(crate) fn fd_write_internal<M: MemorySize>(
                                 let buf_len: usize = wasi_try_ok_ok!(
                                     iovs.buf_len.try_into().map_err(|_| Errno::Inval)
                                 );
-                                let will_be_written = buf_len;
-
                                 let val_cnt = buf_len / std::mem::size_of::<u64>();
+                                let will_be_written =
+                                    val_cnt * std::mem::size_of::<u64>();
+
                                 let val_cnt: M::Offset =
                                     wasi_try_ok_ok!(val_cnt.try_into().map_err(|_| Errno::Inval));
