diff --git a/lib/wasix/src/syscalls/wasi/fd_write.rs b/lib/wasix/src/syscalls/wasi/fd_write.rs
index 68e50b52290..e475dd54609 100644
--- a/lib/wasix/src/syscalls/wasi/fd_write.rs
+++ b/lib/wasix/src/syscalls/wasi/fd_write.rs
@@ -529,8 +529,7 @@ pub(crate) fn fd_write_internal<M: MemorySize>(
                 fd_entry.inner.offset.load(Ordering::Acquire)
             };
 
-            // we set the size but we don't return any errors if it fails as
-            // pipes and sockets will not do anything with this
+            // we set the size but we don't return any errors if it fails
             let (mut memory, _, inodes) =
                 unsafe { env.get_memory_and_wasi_state_and_inodes(&ctx, 0) };
             if is_file {
@@ -546,9 +545,6 @@ pub(crate) fn fd_write_internal<M: MemorySize>(
                     // and get the max value between it and the current size.
                     stat.st_size = stat.st_size.max(offset + bytes_written as u64);
                 }
-            } else {
-                // Cast is valid because we don't support 128 bit systems...
-                fd_entry.inode.stat.write().unwrap().st_size += bytes_written as u64;
             }
         }
         bytes_written
