diff --git a/lib/wasix/src/runtime/resolver/utils.rs b/lib/wasix/src/runtime/resolver/utils.rs
index 1916b5113e6..18d000130d7 100644
--- a/lib/wasix/src/runtime/resolver/utils.rs
+++ b/lib/wasix/src/runtime/resolver/utils.rs
@@ -10,23 +10,36 @@ use crate::http::{HttpResponse, USER_AGENT};
 pub(crate) fn url_from_file_path(path: impl AsRef<Path>) -> Option<Url> {
     let path = path.as_ref();
 
-    if !path.is_absolute() {
-        return None;
+    #[cfg(any(unix, windows, target_os = "redox", target_os = "wasi", target_os = "hermit"))]
+    {
+        return Url::from_file_path(path).ok();
     }
 
-    let mut buffer = String::new();
+    #[cfg(not(any(unix, windows, target_os = "redox", target_os = "wasi", target_os = "hermit")))]
+    {
+        if !path.is_absolute() {
+            return None;
+        }
 
-    for component in path {
-        if !buffer.ends_with('/') {
+        let mut buffer = String::from("file://");
+
+        for component in path.components().skip(1) {
+            use url::percent_encoding::percent_encode;
+            use url::parser::SPECIAL_PATH_SEGMENT;
+
+            buffer.push('/');
+            buffer.extend(percent_encode(
+                component.as_os_str().to_string_lossy().as_bytes(),
+                SPECIAL_PATH_SEGMENT,
+            ));
+        }
+
+        if buffer == "file://" {
             buffer.push('/');
         }
 
-        buffer.push_str(component.to_str()?);
+        buffer.parse().ok()
     }
-
-    buffer.insert_str(0, "file://");
-
-    buffer.parse().ok()
 }
 
 pub(crate) fn webc_headers() -> HeaderMap {
@@ -95,6 +108,7 @@ mod tests {
             "/",
             "/path",
             "/path/to/file.txt",
+            "/path/with spaces and #fragments?.txt",
             "./path/to/file.txt",
             ".",
             "",
