diff --git a/lib/wasix/src/state/func_env.rs b/lib/wasix/src/state/func_env.rs
index 914edc2be0e..adaf1401b23 100644
--- a/lib/wasix/src/state/func_env.rs
+++ b/lib/wasix/src/state/func_env.rs
@@ -218,13 +218,13 @@ impl WasiFunctionEnv {
                         match stack_low.get(store) {
                             wasmer::Value::I32(a) => a as u64,
                             wasmer::Value::I64(a) => a as u64,
-                            _ => 0,
+                            _ => return Err(ExportError::IncompatibleType),
                         }
                     } else if let Some(data_end) = data_end {
                         let data_end = match data_end.get(store) {
                             wasmer::Value::I32(a) => a as u64,
                             wasmer::Value::I64(a) => a as u64,
-                            _ => 0,
+                            _ => return Err(ExportError::IncompatibleType),
                         };
                         // It's possible for the data section to be above the stack, we check for that here and
                         // if it is, we'll assume the stack starts at address 0
