diff --git a/lib/wasix/src/bin_factory/exec.rs b/lib/wasix/src/bin_factory/exec.rs
index 4e60ee04ff3..3893491ef79 100644
--- a/lib/wasix/src/bin_factory/exec.rs
+++ b/lib/wasix/src/bin_factory/exec.rs
@@ -352,7 +352,12 @@ fn call_module(
                 if let Err(err) = unsafe {
                     tasks.resume_wasm_after_poller(Box::new(respawn), ctx, store, deep.trigger)
                 } {
-                    debug!("failed to go into deep sleep - {}", err);
+                    let code = Errno::from(err).into();
+                    debug!("failed to go into deep sleep");
+                    ctx.data(&store).blocking_on_exit(Some(code));
+                    unsafe { run_recycle(recycle, ctx, store) };
+                    handle.thread.set_status_finished(Ok(code));
+                    return;
                 }
                 return;
             }
