diff --git a/lib/wasix/src/runtime/task_manager/mod.rs b/lib/wasix/src/runtime/task_manager/mod.rs
index 2a6457ebc4f..85566f9cc53 100644
--- a/lib/wasix/src/runtime/task_manager/mod.rs
+++ b/lib/wasix/src/runtime/task_manager/mod.rs
@@ -9,7 +9,7 @@ use std::{pin::Pin, time::Duration};
 use bytes::Bytes;
 use derive_more::Debug;
 use futures::future::BoxFuture;
-use futures::{Future, TryFutureExt};
+use futures::{Future, FutureExt};
 use wasmer::{
     AsStoreMut, AsStoreRef, FunctionEnv, Memory, MemoryType, Module, Store, StoreMut, StoreRef,
 };
@@ -493,14 +493,27 @@ where
         O: Send + 'static,
         F: FnOnce() -> O + Send + 'static,
     {
-        let (sender, receiver) = ::tokio::sync::oneshot::channel();
+        let (sender, receiver) =
+            ::tokio::sync::oneshot::channel::<Result<O, Box<dyn std::error::Error + Send + Sync>>>();
 
-        self.task_dedicated(Box::new(move || {
+        let receiver = match self.task_dedicated(Box::new(move || {
             let result = f();
-            let _ = sender.send(result);
-        }))
-        .unwrap();
+            let _ = sender.send(Ok(result));
+        })) {
+            Ok(()) => receiver,
+            Err(err) => {
+                let (sender, receiver) = ::tokio::sync::oneshot::channel::<Result<
+                    O,
+                    Box<dyn std::error::Error + Send + Sync>,
+                >>();
+                let _ = sender.send(Err(Box::new(err)));
+                receiver
+            }
+        };
 
-        Box::new(receiver.map_err(|e| Box::new(e).into()))
+        Box::new(receiver.map(|result| match result {
+            Ok(result) => result.map_err(|e| e as Box<dyn std::error::Error>),
+            Err(e) => Err(Box::new(e) as Box<dyn std::error::Error>),
+        }))
     }
 }
