diff --git a/lib/wasix/src/syscalls/journal/do_checkpoint_from_outside.rs b/lib/wasix/src/syscalls/journal/do_checkpoint_from_outside.rs
index 36603d3904d..e69e33a84fc 100644
--- a/lib/wasix/src/syscalls/journal/do_checkpoint_from_outside.rs
+++ b/lib/wasix/src/syscalls/journal/do_checkpoint_from_outside.rs
@@ -11,10 +11,9 @@ pub(crate) fn do_checkpoint_from_outside(
     // Initiate the checksum (if one already exists we must wait for it to end
     // before we start the next checksum)
 
-    // TODO: Disabled as this blocks the async runtime
-    //while !matches!(guard.checkpoint, WasiProcessCheckpoint::Execute) {
-    //    guard = process.1.wait(guard).unwrap();
-    //}
+    if !matches!(guard.checkpoint, WasiProcessCheckpoint::Execute) {
+        return;
+    }
 
     guard.checkpoint = checkpoint;
     for waker in guard.wakers.drain(..) {
