diff --git a/lib/wasix/src/syscalls/wasi/path_rename.rs b/lib/wasix/src/syscalls/wasi/path_rename.rs
index 132e552d6ae..2d105d6af86 100644
--- a/lib/wasix/src/syscalls/wasi/path_rename.rs
+++ b/lib/wasix/src/syscalls/wasi/path_rename.rs
@@ -204,7 +204,12 @@ pub fn path_rename_internal(
                     })?
                 };
                 if let Err(e) = res {
-                    return Ok(e);
+                    drop(guard);
+                    let mut guard = source_parent_inode.write();
+                    if let Kind::Dir { entries, .. } = guard.deref_mut() {
+                        entries.insert(source_entry_name, source_entry);
+                        return Ok(e);
+                    }
                 }
                 {
                     let source_dir_path = path.clone();
