diff --git a/lib/wasix/src/syscalls/wasi/path_rename.rs b/lib/wasix/src/syscalls/wasi/path_rename.rs
index 132e552d6ae..1ba9e0604ba 100644
--- a/lib/wasix/src/syscalls/wasi/path_rename.rs
+++ b/lib/wasix/src/syscalls/wasi/path_rename.rs
@@ -267,22 +267,19 @@ pub fn path_rename_internal(
 
     let source_size = source_entry.stat.read().unwrap().st_size;
 
-    if need_create {
-        let mut guard = target_parent_inode.write();
-        if let Kind::Dir { entries, .. } = guard.deref_mut() {
-            let result = entries.insert(target_entry_name.clone(), source_entry);
+    let target_inode = source_entry.clone();
+    let mut guard = target_parent_inode.write();
+    if let Kind::Dir { entries, .. } = guard.deref_mut() {
+        let replaced = entries.insert(target_entry_name.clone(), source_entry);
+        if need_create {
             assert!(
-                result.is_none(),
+                replaced.is_none(),
                 "fatal error: race condition on filesystem detected or internal logic error"
             );
         }
     }
+    drop(guard);
 
-    // The target entry is created, one way or the other
-    let target_inode = state
-        .fs
-        .get_inode_at_path(inodes, target_fd, target_path, true)
-        .expect("Expected target inode to exist, and it's too late to safely fail");
     *target_inode.name.write().unwrap() = target_entry_name.into();
     target_inode.stat.write().unwrap().st_size = source_size;
