diff --git a/lib/wasix/src/syscalls/wasix/proc_join.rs b/lib/wasix/src/syscalls/wasix/proc_join.rs
index 87c12ea03e5..7081a9e1264 100644
--- a/lib/wasix/src/syscalls/wasix/proc_join.rs
+++ b/lib/wasix/src/syscalls/wasix/proc_join.rs
@@ -151,16 +151,16 @@ pub(super) fn proc_join_internal<M: MemorySize + 'static>(
 
     // Waiting for a process that is an explicit child will join it
     // meaning it will no longer be a sub-process of the main process
-    let mut process = {
-        let mut inner = ctx.data().process.lock();
+    let (mut process, is_child) = {
+        let inner = ctx.data().process.lock();
         let process = inner
             .children
             .iter()
             .filter(|c| c.pid == pid)
             .map(Clone::clone)
             .next();
-        inner.children.retain(|c| c.pid != pid);
-        process
+        let is_child = process.is_some();
+        (process, is_child)
     };
 
     // Otherwise it could be the case that we are waiting for a process
@@ -181,6 +181,10 @@ pub(super) fn proc_join_internal<M: MemorySize + 'static>(
 
         if flags.contains(JoinFlags::NON_BLOCKING) {
             if let Some(status) = process.try_join() {
+                if is_child {
+                    let mut inner = ctx.data().process.lock();
+                    inner.children.retain(|c| c.pid != pid);
+                }
                 let exit_code = status.unwrap_or_else(|_| Errno::Child.into());
                 ret_result(ctx, JoinStatusResult::ExitNormal(pid, exit_code))
             } else {
@@ -188,9 +192,13 @@ pub(super) fn proc_join_internal<M: MemorySize + 'static>(
             }
         } else {
             // Wait for the process to finish
-            let process2 = process.clone();
+            let parent = is_child.then(|| ctx.data().process.clone());
             let res = __asyncify_with_deep_sleep::<M, _, _>(ctx, async move {
                 let exit_code = process.join().await.unwrap_or_else(|_| Errno::Child.into());
+                if let Some(parent) = parent {
+                    let mut inner = parent.lock();
+                    inner.children.retain(|c| c.pid != pid);
+                }
                 tracing::trace!(%exit_code, "triggered child join");
                 JoinStatusResult::ExitNormal(pid, exit_code)
             })?;
