diff --git a/lib/wasix/src/syscalls/wasi/path_unlink_file.rs b/lib/wasix/src/syscalls/wasi/path_unlink_file.rs
index 36b433ab664..2ff06fdf143 100644
--- a/lib/wasix/src/syscalls/wasi/path_unlink_file.rs
+++ b/lib/wasix/src/syscalls/wasi/path_unlink_file.rs
@@ -73,15 +73,53 @@ pub(crate) fn path_unlink_file_internal(
         }
     };
 
-    let removed_inode = {
+    let st_nlink = {
+        let guard = inode.stat.read().unwrap();
+        debug_assert!(guard.st_nlink > 0);
+        guard.st_nlink
+    };
+    if st_nlink == 1 {
+        let mut guard = inode.read();
+        match guard.deref() {
+            Kind::File { handle, path, .. } => {
+                if let Some(h) = handle {
+                    let mut h = h.write().unwrap();
+                    wasi_try_ok!(h.unlink().map_err(fs_error_into_wasi_err));
+                } else {
+                    // File is closed
+                    // problem with the abstraction, we can't call unlink because there's no handle
+                    // drop mutable borrow on `path`
+                    let path = path.clone();
+                    drop(guard);
+                    wasi_try_ok!(state.fs_remove_file(path));
+                }
+            }
+            Kind::Dir { .. } | Kind::Root { .. } => return Ok(Errno::Isdir),
+            Kind::Symlink { .. } => {
+                match state.fs_remove_file(host_adjusted_path.as_path()) {
+                    Ok(()) => {}
+                    Err(Errno::Noent)
+                        if state
+                            .fs
+                            .ephemeral_symlink_at(host_adjusted_path.as_path())
+                            .is_some() => {}
+                    Err(err) => return Ok(err),
+                }
+                state
+                    .fs
+                    .unregister_ephemeral_symlink(host_adjusted_path.as_path());
+            }
+            _ => unimplemented!("wasi::path_unlink_file for Buffer"),
+        }
+    }
+
+    {
         let mut guard = parent_inode.write();
         match guard.deref_mut() {
             Kind::Dir { entries, .. } => {
                 let removed_inode = wasi_try_ok!(entries.remove(&childs_name).ok_or(Errno::Inval));
                 // TODO: make this a debug assert in the future
                 assert!(inode.ino() == removed_inode.ino());
-                debug_assert!(inode.stat.read().unwrap().st_nlink > 0);
-                removed_inode
             }
             Kind::Root { .. } => return Ok(Errno::Access),
             _ => unreachable!(
@@ -90,46 +128,9 @@ pub(crate) fn path_unlink_file_internal(
         }
     };
 
-    let st_nlink = {
-        let mut guard = removed_inode.stat.write().unwrap();
+    {
+        let mut guard = inode.stat.write().unwrap();
         guard.st_nlink -= 1;
-        guard.st_nlink
-    };
-    if st_nlink == 0 {
-        {
-            let mut guard = removed_inode.read();
-            match guard.deref() {
-                Kind::File { handle, path, .. } => {
-                    if let Some(h) = handle {
-                        let mut h = h.write().unwrap();
-                        wasi_try_ok!(h.unlink().map_err(fs_error_into_wasi_err));
-                    } else {
-                        // File is closed
-                        // problem with the abstraction, we can't call unlink because there's no handle
-                        // drop mutable borrow on `path`
-                        let path = path.clone();
-                        drop(guard);
-                        wasi_try_ok!(state.fs_remove_file(path));
-                    }
-                }
-                Kind::Dir { .. } | Kind::Root { .. } => return Ok(Errno::Isdir),
-                Kind::Symlink { .. } => {
-                    match state.fs_remove_file(host_adjusted_path.as_path()) {
-                        Ok(()) => {}
-                        Err(Errno::Noent)
-                            if state
-                                .fs
-                                .ephemeral_symlink_at(host_adjusted_path.as_path())
-                                .is_some() => {}
-                        Err(err) => return Ok(err),
-                    }
-                    state
-                        .fs
-                        .unregister_ephemeral_symlink(host_adjusted_path.as_path());
-                }
-                _ => unimplemented!("wasi::path_unlink_file for Buffer"),
-            }
-        }
     }
 
     Ok(Errno::Success)
