diff --git a/lib/wasix/src/syscalls/wasix/sock_pair.rs b/lib/wasix/src/syscalls/wasix/sock_pair.rs
index 3f087738c2d..978fdac2960 100644
--- a/lib/wasix/src/syscalls/wasix/sock_pair.rs
+++ b/lib/wasix/src/syscalls/wasix/sock_pair.rs
@@ -44,6 +44,10 @@ pub fn sock_pair<M: MemorySize>(
     WasiEnv::do_pending_operations(&mut ctx)?;
 
     // only certain combinations are supported
+    match af {
+        Addressfamily::Inet4 | Addressfamily::Inet6 => {}
+        _ => return Ok(Errno::Notsup),
+    }
     match pt {
         SockProto::Tcp if ty != Socktype::Stream => {
             return Ok(Errno::Notsup);
@@ -51,7 +55,8 @@ pub fn sock_pair<M: MemorySize>(
         SockProto::Udp if ty != Socktype::Dgram => {
             return Ok(Errno::Notsup);
         }
-        _ => {}
+        SockProto::Tcp | SockProto::Udp => {}
+        _ => return Ok(Errno::Notsup),
     }
 
     // FIXME: currently, socket properties are ignored outright, since they
