diff --git a/lib/wasix/src/syscalls/wasix/futex_wake.rs b/lib/wasix/src/syscalls/wasix/futex_wake.rs
index ae2020df991..a45e9dab79c 100644
--- a/lib/wasix/src/syscalls/wasix/futex_wake.rs
+++ b/lib/wasix/src/syscalls/wasix/futex_wake.rs
@@ -23,24 +23,25 @@ pub fn futex_wake<M: MemorySize>(
     let pointer: u64 = futex_ptr.offset().into();
     Span::current().record("futex_idx", pointer);
 
-    let mut woken = false;
     let woken = {
         let mut guard = state.futexs.lock().unwrap();
         if let Some(futex) = guard.futexes.get_mut(&pointer) {
-            let first = futex.wakers.keys().copied().next();
-            if let Some(id) = first
+            let woken = if let Some(id) = futex.wakers.keys().copied().next()
                 && let Some(Some(w)) = futex.wakers.remove(&id)
             {
                 w.wake();
-            }
+                true
+            } else {
+                false
+            };
             if futex.wakers.is_empty() {
                 guard.futexes.remove(&pointer);
             }
-            tracing::trace!("wake(hit) on {pointer}");
-            true
+            tracing::trace!("wake({}) on {pointer}", if woken { "hit" } else { "miss" });
+            woken
         } else {
             tracing::trace!("wake(miss) on {pointer}");
-            true
+            false
         }
     };
     Span::current().record("woken", woken);
