diff --git a/lib/wasix/src/syscalls/wasi/path_filestat_set_times.rs b/lib/wasix/src/syscalls/wasi/path_filestat_set_times.rs
index e2e35eb5bde..f803ac11a10 100644
--- a/lib/wasix/src/syscalls/wasi/path_filestat_set_times.rs
+++ b/lib/wasix/src/syscalls/wasi/path_filestat_set_times.rs
@@ -80,7 +80,6 @@ pub(crate) fn path_filestat_set_times_internal(
     let env = ctx.data();
     let (memory, mut state, inodes) = unsafe { env.get_memory_and_wasi_state_and_inodes(&ctx, 0) };
     let fd_entry = state.fs.get_fd(fd)?;
-    let fd_inode = fd_entry.inode;
     if !fd_entry
         .inner
         .rights
@@ -109,7 +108,7 @@ pub(crate) fn path_filestat_set_times_internal(
         } else {
             get_current_time_in_nanos()?
         };
-        fd_inode.stat.write().unwrap().st_atim = time_to_set;
+        file_inode.stat.write().unwrap().st_atim = time_to_set;
     }
     if fst_flags.contains(Fstflags::SET_MTIM) || fst_flags.contains(Fstflags::SET_MTIM_NOW) {
         let time_to_set = if fst_flags.contains(Fstflags::SET_MTIM) {
@@ -117,7 +116,7 @@ pub(crate) fn path_filestat_set_times_internal(
         } else {
             get_current_time_in_nanos()?
         };
-        fd_inode.stat.write().unwrap().st_mtim = time_to_set;
+        file_inode.stat.write().unwrap().st_mtim = time_to_set;
     }
 
     Ok(())
