diff --git a/src/router/lib.rs b/src/router/lib.rs
index 2d5fc1f685..187540ffe8 100644
--- a/src/router/lib.rs
+++ b/src/router/lib.rs
@@ -1894,7 +1894,7 @@ pub mod pattern {
                         i += 1;
 
                         if matches!(tag, Tag::OptionalCatchAll) {
-                            if input[i as usize] != b']' {
+                            if i > end || input[i as usize] != b']' {
                                 return Err(PatternParseError::PatternMissingClosingBracket);
                             }
                             i += 1;
