diff --git a/src/http/lib.rs b/src/http/lib.rs
index f083ec93da..aea074d95e 100644
--- a/src/http/lib.rs
+++ b/src/http/lib.rs
@@ -4617,17 +4617,16 @@ impl<'a> HTTPClient<'a> {
                                 },
                             ];
                             for to_remove in headers_to_remove.iter() {
-                                let found =
-                                    self.header_entries
-                                        .items_name()
-                                        .iter()
-                                        .position(|name_ptr| {
-                                            let name = self.header_str(*name_ptr);
-                                            name.len() == to_remove.name.len()
-                                                && hash_header_name(name) == to_remove.hash
-                                        });
-                                if let Some(i) = found {
-                                    self.header_entries.ordered_remove(i);
+                                let mut i = 0;
+                                while i < self.header_entries.len() {
+                                    let name = self.header_str(self.header_entries.items_name()[i]);
+                                    if name.len() == to_remove.name.len()
+                                        && hash_header_name(name) == to_remove.hash
+                                    {
+                                        let _ = self.header_entries.ordered_remove(i);
+                                    } else {
+                                        i += 1;
+                                    }
                                 }
                             }
                         }
