diff --git a/src/sql_jsc/mysql/MySQLConnection.rs b/src/sql_jsc/mysql/MySQLConnection.rs
index 77382ca441..633253b7ba 100644
--- a/src/sql_jsc/mysql/MySQLConnection.rs
+++ b/src/sql_jsc/mysql/MySQLConnection.rs
@@ -818,18 +818,7 @@ impl MySQLConnection {
                                     bun_core::scoped_log!(MySQLConnection, "continue auth");
 
                                     if self.ssl_mode == SSLMode::Disable {
-                                        // we are in plain TCP so we need to request the public key
-                                        self.set_status(ConnectionState::AuthenticationAwaitingPk);
-                                        bun_core::scoped_log!(
-                                            MySQLConnection,
-                                            "awaiting public key"
-                                        );
-                                        let mut packet = self.writer().start(self.sequence_id)?;
-
-                                        let request = Auth::caching_sha2_password::PublicKeyRequest;
-                                        request.write(self.writer())?;
-                                        packet.end()?;
-                                        self.flush_data();
+                                        return Err(AnyMySQLError::AuthenticationFailed);
                                     } else {
                                         bun_core::scoped_log!(
                                             MySQLConnection,
