diff --git a/src/bundler/entry_points.rs b/src/bundler/entry_points.rs
index cd933457cd..9d5871922f 100644
--- a/src/bundler/entry_points.rs
+++ b/src/bundler/entry_points.rs
@@ -436,6 +436,20 @@ impl MacroEntryPoint {
         // for the (label, code) slices passed to `init_path_string`.
         let label_len = macro_label_.len();
         entry.code_buffer[..label_len].copy_from_slice(macro_label_);
+        let function_name_single_quoted = strings::format_escapes(
+            function_name,
+            strings::QuoteEscapeFormatFlags {
+                quote_char: b'\'',
+                ..Default::default()
+            },
+        );
+        let function_name_double_quoted = strings::format_escapes(
+            function_name,
+            strings::QuoteEscapeFormatFlags {
+                quote_char: b'"',
+                ..Default::default()
+            },
+        );
 
         let code_len: usize = 'brk: {
             if import_path.base == b"bun" {
@@ -456,8 +470,8 @@ impl MacroEntryPoint {
                      }}\n\
                      \n\
                      Bun.registerMacro({}, macro);",
-                    BStr::new(function_name),
-                    BStr::new(function_name),
+                    function_name_single_quoted,
+                    function_name_double_quoted,
                     macro_id,
                 )
                 .map_err(|_| bun_core::err!("NoSpaceLeft"))?;
@@ -494,8 +508,8 @@ impl MacroEntryPoint {
                         ..Default::default()
                     }
                 ),
-                BStr::new(function_name),
-                BStr::new(function_name),
+                function_name_single_quoted,
+                function_name_double_quoted,
                 bun_fmt::fmt_path_u8(
                     dir_to_use,
                     bun_fmt::PathFormatOptions {
@@ -511,7 +525,7 @@ impl MacroEntryPoint {
                     }
                 ),
                 macro_id,
-                BStr::new(function_name),
+                function_name_single_quoted,
             )
             .map_err(|_| bun_core::err!("NoSpaceLeft"))?;
             cursor.position() as usize
