diff --git a/src/parsers/yaml.rs b/src/parsers/yaml.rs
index 06adb0292b..61337316a8 100644
--- a/src/parsers/yaml.rs
+++ b/src/parsers/yaml.rs
@@ -5484,7 +5484,7 @@ impl<'i, Enc: Encoding> Parser<'i, Enc> {
                 return Err(ParseError::UnexpectedCharacter);
             }
             self.inc(1);
-            while !self.is_char(Enc::ch(b'\n')) && !self.is_char(Enc::ch(b'\r')) {
+            while !self.is_b_char_or_eof() {
                 self.inc(1);
             }
         }
