diff --git a/src/css/properties/custom.rs b/src/css/properties/custom.rs
index 40b7a5b4d3..576c6e3d32 100644
--- a/src/css/properties/custom.rs
+++ b/src/css/properties/custom.rs
@@ -591,7 +591,7 @@ impl TokenList {
                         last_is_delim = false;
                         last_is_whitespace = false;
                     } else if let Ok(color) =
-                        input.try_parse(|i| UnresolvedColor::parse(i, f, options))
+                        input.try_parse(|i| UnresolvedColor::parse(i, f, options, depth))
                     {
                         tokens.push(TokenOrValue::UnresolvedColor(color));
                         last_is_delim = false;
@@ -956,7 +956,12 @@ impl UnresolvedColor {
         }
     }
 
-    pub fn parse(input: &mut Parser, f: &[u8], options: &ParserOptions) -> Result<UnresolvedColor> {
+    pub fn parse(
+        input: &mut Parser,
+        f: &[u8],
+        options: &ParserOptions,
+        depth: usize,
+    ) -> Result<UnresolvedColor> {
         use css_values::color::{
             ComponentParser, HSL, SRGB, parse_hsl_hwb_components, parse_rgb_components,
         };
@@ -970,7 +975,7 @@ impl UnresolvedColor {
                         return Err(i.new_custom_error(ParserError::invalid_value));
                     }
                     i.expect_delim(b'/')?;
-                    let alpha = TokenListFns::parse(i, options, 0)?;
+                    let alpha = TokenListFns::parse(i, options, depth + 1)?;
                     Ok(UnresolvedColor::RGB { r, g, b, alpha })
                 })
             });
@@ -982,7 +987,7 @@ impl UnresolvedColor {
                         return Err(i.new_custom_error(ParserError::invalid_value));
                     }
                     i.expect_delim(b'/')?;
-                    let alpha = TokenListFns::parse(i, options, 0)?;
+                    let alpha = TokenListFns::parse(i, options, depth + 1)?;
                     Ok(UnresolvedColor::HSL { h, s, l, alpha })
                 })
             });
