diff --git a/src/md/blocks.rs b/src/md/blocks.rs
index 918cf26bed..460065e9ec 100644
--- a/src/md/blocks.rs
+++ b/src/md/blocks.rs
@@ -988,16 +988,10 @@ impl Parser<'_> {
             }
 
             // First definition wins
-            let mut already_exists = false;
-            for existing in self.ref_defs.iter() {
-                if existing.label[..] == norm_label[..] {
-                    already_exists = true;
-                    break;
-                }
-            }
-            if !already_exists {
+            let label = norm_label.into_boxed_slice();
+            if self.ref_def_labels.insert(label.clone()) {
                 self.ref_defs.push(crate::ref_defs::RefDef {
-                    label: norm_label.into_boxed_slice(),
+                    label,
                     dest: dest_dupe,
                     title: title_dupe,
                 });
diff --git a/src/md/parser.rs b/src/md/parser.rs
index f0c81b4451..f7d6a294eb 100644
--- a/src/md/parser.rs
+++ b/src/md/parser.rs
@@ -93,6 +93,7 @@ pub struct Parser<'a> {
 
     // Ref defs
     pub ref_defs: Vec<RefDef>,
+    pub ref_def_labels: std::collections::HashSet<Box<[u8]>>,
 
     // State
     pub last_line_has_list_loosening_effect: bool,
@@ -196,6 +197,7 @@ impl<'a> Parser<'a> {
             table_col_count: 0,
             table_alignments: [Align::Default; TABLE_MAXCOLCOUNT as usize],
             ref_defs: Vec::new(),
+            ref_def_labels: std::collections::HashSet::new(),
             last_line_has_list_loosening_effect: false,
             last_list_item_starts_with_two_blank_lines: false,
             max_ref_def_output: (16 * (size as u64)).min(1024 * 1024).min(u32::MAX as u64),
diff --git a/src/md/ref_defs.rs b/src/md/ref_defs.rs
index 348ae5695b..e684410d0a 100644
--- a/src/md/ref_defs.rs
+++ b/src/md/ref_defs.rs
@@ -411,19 +411,13 @@ impl Parser<'_> {
                 if norm_label.is_empty() {
                     break; // whitespace-only labels are invalid
                 }
-                let mut already_exists = false;
-                for existing in self.ref_defs.iter() {
-                    if existing.label[..] == norm_label[..] {
-                        already_exists = true;
-                        break;
-                    }
-                }
-                if !already_exists {
+                let label = norm_label.into_boxed_slice();
+                if self.ref_def_labels.insert(label.clone()) {
                     // Dupe dest and title since they point into self.buffer which gets reused
                     let dest_dupe: Box<[u8]> = Box::from(result.dest);
                     let title_dupe: Box<[u8]> = Box::from(result.title);
                     self.ref_defs.push(RefDef {
-                        label: norm_label.into_boxed_slice(),
+                        label,
                         dest: dest_dupe,
                         title: title_dupe,
                     });
