diff --git a/src/bundler/linker_context/postProcessCSSChunk.rs b/src/bundler/linker_context/postProcessCSSChunk.rs
index e4dd36f59d..c0f0a73ec0 100644
--- a/src/bundler/linker_context/postProcessCSSChunk.rs
+++ b/src/bundler/linker_context/postProcessCSSChunk.rs
@@ -69,16 +69,8 @@ pub fn post_process_css_chunk(
                 line_offset.advance(b"\n");
             }
 
-            let pretty: &[u8] = sources[source_index as usize].path.pretty;
-
-            j.push_static(b"/* ");
-            line_offset.advance(b"/* ");
-
-            j.push_static(pretty);
-            line_offset.advance(pretty);
-
-            j.push_static(b" */\n");
-            line_offset.advance(b" */\n");
+            j.push_static(b"/* */\n");
+            line_offset.advance(b"/* */\n");
         }
 
         if !compile_result.code().is_empty() {
