diff --git a/server/mpm/event/event.c b/server/mpm/event/event.c
index 050d823..dd990ed 100644
--- a/server/mpm/event/event.c
+++ b/server/mpm/event/event.c
@@ -4119,6 +4119,9 @@ static const char *set_worker_factor(cmd_parms * cmd, void *dummy,
     if (val <= 0)
         return "AsyncRequestWorkerFactor argument must be a positive number";
 
+    if (val > (double)UINT_MAX / WORKER_FACTOR_SCALE)
+        return "AsyncRequestWorkerFactor argument is too large";
+
     worker_factor = val * WORKER_FACTOR_SCALE;
     if (worker_factor < WORKER_FACTOR_SCALE) {
         worker_factor = WORKER_FACTOR_SCALE;
