diff --git a/modules/dav/lock/locks.c b/modules/dav/lock/locks.c
index 0f072ec..2ea8006 100644
--- a/modules/dav/lock/locks.c
+++ b/modules/dav/lock/locks.c
@@ -589,6 +589,18 @@ static dav_error * dav_generic_load_lock_record(dav_lockdb *lockdb,
     dav_lock_discovery *dp;
     dav_lock_indirect *ip;
 
+#define dav_corrupt_record()                                            \
+    do {                                                                \
+        apr_dbm_freedatum(lockdb->info->db, val);                       \
+        return dav_new_error(p, HTTP_INTERNAL_SERVER_ERROR,              \
+                             DAV_ERR_LOCK_CORRUPT_DB, 0,                \
+                             "The lock database was found to be corrupt."); \
+    } while (0)
+#define dav_check_read(_len)                                            \
+    if ((_len) > val.dsize - offset) {                                  \
+        dav_corrupt_record();                                           \
+    }
+
     if (add_method != DAV_APPEND_LIST) {
         *direct = NULL;
         *indirect = NULL;
@@ -623,28 +635,42 @@ static dav_error * dav_generic_load_lock_record(dav_lockdb *lockdb,
             dp = apr_pcalloc(p, sizeof(*dp));
 
             /* Copy the dav_lock_discovery_fixed portion */
+            dav_check_read(sizeof(dp->f));
             memcpy(dp, val.dptr + offset, sizeof(dp->f));
             offset += sizeof(dp->f);
 
             /* Copy the lock token. */
+            dav_check_read(sizeof(*dp->locktoken));
             dp->locktoken = apr_pmemdup(p, val.dptr + offset, sizeof(*dp->locktoken));
             offset += sizeof(*dp->locktoken);
 
             /* Do we have an owner field? */
+            dav_check_read(1);
             if (*(val.dptr + offset) == '\0') {
                 ++offset;
             }
             else {
-                apr_size_t len = strlen(val.dptr + offset);
+                apr_size_t len;
+                void *eos = memchr(val.dptr + offset, '\0', val.dsize - offset);
+                if (eos == NULL) {
+                    dav_corrupt_record();
+                }
+                len = (char *)eos - (val.dptr + offset);
                 dp->owner = apr_pstrmemdup(p, val.dptr + offset, len);
                 offset += len + 1;
             }
 
+            dav_check_read(1);
             if (*(val.dptr + offset) == '\0') {
                 ++offset;
             }
             else {
-                apr_size_t len = strlen(val.dptr + offset);
+                apr_size_t len;
+                void *eos = memchr(val.dptr + offset, '\0', val.dsize - offset);
+                if (eos == NULL) {
+                    dav_corrupt_record();
+                }
+                len = (char *)eos - (val.dptr + offset);
                 dp->auth_user = apr_pstrmemdup(p, val.dptr + offset, len);
                 offset += len + 1;
             }
@@ -662,13 +688,17 @@ static dav_error * dav_generic_load_lock_record(dav_lockdb *lockdb,
             /* Create and fill a dav_lock_indirect structure */
 
             ip = apr_pcalloc(p, sizeof(*ip));
+            dav_check_read(sizeof(*ip->locktoken));
             ip->locktoken = apr_pmemdup(p, val.dptr + offset, sizeof(*ip->locktoken));
             offset += sizeof(*ip->locktoken);
+            dav_check_read(sizeof(ip->timeout));
             memcpy(&ip->timeout, val.dptr + offset, sizeof(ip->timeout));
             offset += sizeof(ip->timeout);
             /* length of datum */
+            dav_check_read(sizeof(ip->key.dsize));
             ip->key.dsize = *((int *) (val.dptr + offset));
             offset += sizeof(ip->key.dsize);
+            dav_check_read(ip->key.dsize);
             ip->key.dptr = apr_pmemdup(p, val.dptr + offset, ip->key.dsize);
             offset += ip->key.dsize;
 
@@ -698,6 +728,9 @@ static dav_error * dav_generic_load_lock_record(dav_lockdb *lockdb,
         }
     }
 
+#undef dav_check_read
+#undef dav_corrupt_record
+
     apr_dbm_freedatum(lockdb->info->db, val);
 
     /* Clean up this record if we found expired locks */
