diff --git a/modules/http2/h2_config.c b/modules/http2/h2_config.c
index 94fd8d2..07d10a8 100644
--- a/modules/http2/h2_config.c
+++ b/modules/http2/h2_config.c
@@ -1118,7 +1118,7 @@ const command_rec h2_cmds[] = {
     AP_INIT_TAKE2("H2EarlyHint", h2_conf_add_early_hint, NULL,
                    OR_FILEINFO|OR_AUTHCFG, "add a a 'Link:' header for a 103 Early Hints response."),
     AP_INIT_TAKE1("H2ProxyRequests", h2_conf_set_proxy_requests, NULL,
-                  OR_FILEINFO, "Enables forward proxy requests via HTTP/2"),
+                  RSRC_CONF, "Enables forward proxy requests via HTTP/2"),
     AP_INIT_TAKE1("H2WebSockets", h2_conf_set_websockets, NULL,
                   RSRC_CONF, "off to disable WebSockets over HTTP/2"),
     AP_END_CMD
