diff --git a/modules/http2/h2_session.c b/modules/http2/h2_session.c
index 21ede5c..a76e540 100644
--- a/modules/http2/h2_session.c
+++ b/modules/http2/h2_session.c
@@ -155,26 +155,32 @@ static int spri_cmp(int sid1, nghttp2_stream *s1,
                     int sid2, nghttp2_stream *s2, h2_session *session)
 {
     nghttp2_stream *p1, *p2;
+    int depth = 0;
     
-    p1 = nghttp2_stream_get_parent(s1);
-    p2 = nghttp2_stream_get_parent(s2);
-    
-    if (p1 == p2) {
-        int32_t w1, w2;
+    (void)session;
+    while (depth++ < 1000) {
+        p1 = nghttp2_stream_get_parent(s1);
+        p2 = nghttp2_stream_get_parent(s2);
         
-        w1 = nghttp2_stream_get_weight(s1);
-        w2 = nghttp2_stream_get_weight(s2);
-        return w2 - w1;
+        if (p1 == p2) {
+            int32_t w1, w2;
+            
+            w1 = nghttp2_stream_get_weight(s1);
+            w2 = nghttp2_stream_get_weight(s2);
+            return w2 - w1;
+        }
+        else if (!p1) {
+            /* stream 1 closer to root */
+            return -1;
+        }
+        else if (!p2) {
+            /* stream 2 closer to root */
+            return 1;
+        }
+        s1 = p1;
+        s2 = p2;
     }
-    else if (!p1) {
-        /* stream 1 closer to root */
-        return -1;
-    }
-    else if (!p2) {
-        /* stream 2 closer to root */
-        return 1;
-    }
-    return spri_cmp(sid1, p1, sid2, p2, session);
+    return sid1 - sid2;
 }
 
 static int stream_pri_cmp(int sid1, int sid2, void *ctx)
