diff --git a/modules/mappers/mod_speling.c b/modules/mappers/mod_speling.c
index 2ed65eb..e71eb04 100644
--- a/modules/mappers/mod_speling.c
+++ b/modules/mappers/mod_speling.c
@@ -256,7 +256,7 @@ static int check_speling(request_rec *r)
     pglen = strlen(postgood);
 
     /* Check to see if the URL pieces add up */
-    if (strcmp(postgood, r->uri + (urlen - pglen))) {
+    if (pglen > urlen || strcmp(postgood, r->uri + (urlen - pglen))) {
         return DECLINED;
     }
