diff --git a/modules/md/md_ocsp.c b/modules/md/md_ocsp.c
index d2dfd73..763aa66 100644
--- a/modules/md/md_ocsp.c
+++ b/modules/md/md_ocsp.c
@@ -595,6 +595,7 @@ static apr_status_t ostat_on_resp(const md_http_response_t *resp, void *baton)
     md_data_t der, new_der;
     md_timeperiod_t valid;
     md_ocsp_cert_stat_t nstat;
+    apr_time_t now;
     
     der.data = new_der.data = NULL;
     der.len  = new_der.len = 0;
@@ -692,7 +693,8 @@ static apr_status_t ostat_on_resp(const md_http_response_t *resp, void *baton)
     new_der.len = (apr_size_t)n;
     new_der.free_data = md_openssl_free;
     nstat = (bstatus == V_OCSP_CERTSTATUS_GOOD)? MD_OCSP_CERT_ST_GOOD : MD_OCSP_CERT_ST_REVOKED;
-    valid.start = bup? md_asn1_generalized_time_get(bup) : apr_time_now();
+    now = apr_time_now();
+    valid.start = bup? md_asn1_generalized_time_get(bup) : now;
     if (bnextup) {
         valid.end = md_asn1_generalized_time_get(bnextup);
     }
@@ -701,10 +703,17 @@ static apr_status_t ostat_on_resp(const md_http_response_t *resp, void *baton)
          * Refresh attempts will be started some time earlier. */
         valid.end = valid.start + apr_time_from_sec(MD_SECS_PER_DAY / 2);
     }
+    if (now < valid.start || now > valid.end) {
+        rv = APR_EINVAL;
+        md_result_printf(update->result, rv, "OCSP response outside validity period %s",
+                         md_timeperiod_print(req->pool, &valid));
+        md_result_log(update->result, MD_LOG_DEBUG);
+        goto cleanup;
+    }
     
     /* First, update the instance with a copy */
     apr_thread_mutex_lock(ostat->reg->mutex);
-    ostat_set(ostat, nstat, &new_der, &valid, apr_time_now());
+    ostat_set(ostat, nstat, &new_der, &valid, now);
     apr_thread_mutex_unlock(ostat->reg->mutex);
     
     /* Next, save the original response */
