diff --git a/modules/filters/mod_proxy_html.c b/modules/filters/mod_proxy_html.c
index 4205a61..cf266e1 100644
--- a/modules/filters/mod_proxy_html.c
+++ b/modules/filters/mod_proxy_html.c
@@ -666,6 +666,7 @@ static meta *metafix(request_rec *r, const char *buf, apr_size_t len)
     size_t offs = 0;
     const char *p;
     const char *q;
+    const char *match_end;
     char *header;
     char *content;
     ap_regmatch_t pmatch[2];
@@ -675,9 +676,14 @@ static meta *metafix(request_rec *r, const char *buf, apr_size_t len)
            !ap_regexec_len(seek_meta, buf + offs, len - offs, 2, pmatch, 0)) {
         header = NULL;
         content = NULL;
+        match_end = buf+offs+pmatch[0].rm_eo;
         p = buf+offs+pmatch[1].rm_eo;
-        while (!apr_isalpha(*++p));
-        for (q = p; apr_isalnum(*q) || (*q == '-'); ++q);
+        while (++p < match_end && !apr_isalpha(*p));
+        if (p >= match_end) {
+            offs += pmatch[0].rm_eo;
+            continue;
+        }
+        for (q = p; q < match_end && (apr_isalnum(*q) || (*q == '-')); ++q);
         header = apr_pstrmemdup(r->pool, p, q-p);
         if (ap_cstr_casecmpn(header, "Content-", 8)) {
             /* find content=... string */
@@ -685,22 +691,24 @@ static meta *metafix(request_rec *r, const char *buf, apr_size_t len)
                               pmatch[0].rm_eo - pmatch[0].rm_so);
             /* if it doesn't contain "content", ignore, don't crash! */
             if (p != NULL) {
-                while (*p) {
+                while (p < match_end) {
                     p += 7;
-                    while (apr_isspace(*p))
+                    while (p < match_end && apr_isspace(*p))
                         ++p;
                     /* XXX Should we search for another content= pattern? */
-                    if (*p != '=')
+                    if (p >= match_end || *p != '=')
+                        break;
+                    while (++p < match_end && apr_isspace(*p));
+                    if (p >= match_end)
                         break;
-                    while (*p && apr_isspace(*++p));
                     if ((*p == '\'') || (*p == '"')) {
                         delim = *p++;
-                        for (q = p; *q && *q != delim; ++q);
+                        for (q = p; q < match_end && *q != delim; ++q);
                         /* No terminating delimiter found? Skip the bogus directive */
-                        if (*q != delim)
+                        if (q >= match_end || *q != delim)
                            break;
                     } else {
-                        for (q = p; *q && !apr_isspace(*q) && (*q != '>'); ++q);
+                        for (q = p; q < match_end && !apr_isspace(*q) && (*q != '>'); ++q);
                     }
                     content = apr_pstrmemdup(r->pool, p, q-p);
                     break;
