diff --git a/modules/md/md_crypt.c b/modules/md/md_crypt.c
index eef1268..6df9ef5 100644
--- a/modules/md/md_crypt.c
+++ b/modules/md/md_crypt.c
@@ -1768,6 +1768,9 @@ static const char *alt_names(apr_array_header_t *domains, apr_pool_t *p)
     
     for (i = 0; i < domains->nelts; ++i) {
         domain = APR_ARRAY_IDX(domains, i, const char *);
+        if (strchr(domain, ',')) {
+            return NULL;
+        }
         alts = apr_psprintf(p, "%s%sDNS:%s", alts, sep, domain);
         sep = ",";
     }
@@ -1806,8 +1809,13 @@ static apr_status_t sk_add_alt_names(STACK_OF(X509_EXTENSION) *exts,
 {
     if (domains->nelts > 0) {
         X509_EXTENSION *x;
+        const char *alts;
         
-        x = X509V3_EXT_conf_nid(NULL, NULL, NID_subject_alt_name, (char*)alt_names(domains, p));
+        alts = alt_names(domains, p);
+        if (!alts) {
+            return APR_EINVAL;
+        }
+        x = X509V3_EXT_conf_nid(NULL, NULL, NID_subject_alt_name, (char*)alts);
         if (NULL == x) {
             return APR_EGENERAL;
         }
@@ -2017,6 +2025,7 @@ apr_status_t md_cert_self_sign(md_cert_t **pcert, const char *cn,
 {
     X509 *x;
     md_cert_t *cert = NULL;
+    const char *alts;
     apr_status_t rv;
     
     assert(domains);
@@ -2024,7 +2033,11 @@ apr_status_t md_cert_self_sign(md_cert_t **pcert, const char *cn,
     if (APR_SUCCESS != (rv = mk_x509(&x, pkey, cn, valid_for, p))) goto out;
     
     /* add the domain as alt name */
-    if (APR_SUCCESS != (rv = add_ext(x, NID_subject_alt_name, alt_names(domains, p), p))) {
+    alts = alt_names(domains, p);
+    if (!alts) {
+        rv = APR_EINVAL; goto out;
+    }
+    if (APR_SUCCESS != (rv = add_ext(x, NID_subject_alt_name, alts, p))) {
         md_log_perror(MD_LOG_MARK, MD_LOG_ERR, rv, p, "%s: set alt_name ext", cn);
         goto out;
     }
