diff --git a/modules/md/md_acme.c b/modules/md/md_acme.c
index 099d3a4..50e09ab 100644
--- a/modules/md/md_acme.c
+++ b/modules/md/md_acme.c
@@ -183,7 +183,7 @@ static apr_status_t inspect_problem(md_acme_req_t *req, const md_http_response_t
             req->resp_json = problem;
             ptype = md_json_gets(problem, MD_KEY_TYPE, NULL); 
             pdetail = md_json_gets(problem, MD_KEY_DETAIL, NULL);
-            req->rv = problem_status_get(ptype);
+            req->rv = ptype? problem_status_get(ptype) : APR_EGENERAL;
             md_result_problem_set(req->result, req->rv, ptype, pdetail,
                                   md_json_getj(problem, MD_KEY_SUBPROBLEMS, NULL));
