diff --git a/src/event/quic/ngx_event_quic_openssl_compat.c b/src/event/quic/ngx_event_quic_openssl_compat.c
index c5f4a0f..89e9110 100644
--- a/src/event/quic/ngx_event_quic_openssl_compat.c
+++ b/src/event/quic/ngx_event_quic_openssl_compat.c
@@ -624,6 +624,11 @@ SSL_set_quic_transport_params(SSL *ssl, const uint8_t *params,
 
     c = ngx_ssl_get_connection(ssl);
     qc = ngx_quic_get_connection(c);
+
+    if (qc == NULL || qc->compat == NULL) {
+        return 0;
+    }
+
     com = qc->compat;
 
     com->tp.len = params_len;
@@ -643,6 +648,13 @@ SSL_get_peer_quic_transport_params(const SSL *ssl, const uint8_t **out_params,
 
     c = ngx_ssl_get_connection(ssl);
     qc = ngx_quic_get_connection(c);
+
+    if (qc == NULL || qc->compat == NULL) {
+        *out_params = NULL;
+        *out_params_len = 0;
+        return;
+    }
+
     com = qc->compat;
 
     *out_params = com->ctp.data;
