diff --git a/src/http/modules/ngx_http_gzip_filter_module.c b/src/http/modules/ngx_http_gzip_filter_module.c
index dac21bb..7214a15 100644
--- a/src/http/modules/ngx_http_gzip_filter_module.c
+++ b/src/http/modules/ngx_http_gzip_filter_module.c
@@ -1021,7 +1021,8 @@ static ngx_int_t
 ngx_http_gzip_ratio_variable(ngx_http_request_t *r,
     ngx_http_variable_value_t *v, uintptr_t data)
 {
-    ngx_uint_t            zint, zfrac;
+    size_t                rem, threshold;
+    ngx_uint_t            scale, zint, zfrac;
     ngx_http_gzip_ctx_t  *ctx;
 
     ctx = ngx_http_get_module_ctx(r, ngx_http_gzip_filter_module);
@@ -1041,9 +1042,20 @@ ngx_http_gzip_ratio_variable(ngx_http_request_t *r,
     }
 
     zint = (ngx_uint_t) (ctx->zin / ctx->zout);
-    zfrac = (ngx_uint_t) ((ctx->zin * 100 / ctx->zout) % 100);
+    rem = ctx->zin % ctx->zout;
 
-    if ((ctx->zin * 1000 / ctx->zout) % 10 > 4) {
+    for (scale = 0; scale < 1000; scale++) {
+        threshold = (scale + 1) * (ctx->zout / 1000)
+                    + ((scale + 1) * (ctx->zout % 1000) + 999) / 1000;
+
+        if (rem < threshold) {
+            break;
+        }
+    }
+
+    zfrac = scale / 10;
+
+    if (scale % 10 > 4) {
 
         /* the rounding, e.g., 2.125 to 2.13 */
