diff --git a/src/os/win32/ngx_process_cycle.c b/src/os/win32/ngx_process_cycle.c
index a39335f..70ea4f7 100644
--- a/src/os/win32/ngx_process_cycle.c
+++ b/src/os/win32/ngx_process_cycle.c
@@ -747,8 +747,8 @@ ngx_worker_process_cycle(ngx_cycle_t *cycle, char *mevn)
 
     ngx_close_handle(ngx_cache_manager_event);
     ngx_close_handle(events[0]);
-    ngx_close_handle(events[1]);
-    ngx_close_handle(events[2]);
+    ngx_close_handle(wtid);
+    ngx_close_handle(cmtid);
     ngx_close_handle(mev);
 
     ngx_worker_process_exit(cycle);
