diff --git a/src/http/v3/ngx_http_v3_parse.c b/src/http/v3/ngx_http_v3_parse.c
index bcbf0db..c3d6baa 100644
--- a/src/http/v3/ngx_http_v3_parse.c
+++ b/src/http/v3/ngx_http_v3_parse.c
@@ -900,7 +900,7 @@ ngx_http_v3_parse_field_l(ngx_connection_t *c,
 
             st->literal.length = st->pint.value;
             if (st->literal.length == 0) {
-                return NGX_ERROR;
+                return NGX_HTTP_V3_ERR_DECOMPRESSION_FAILED;
             }
 
             st->state = sw_name;
@@ -1604,7 +1604,7 @@ ngx_http_v3_parse_field_iln(ngx_connection_t *c,
 
             st->literal.length = st->pint.value;
             if (st->literal.length == 0) {
-                return NGX_ERROR;
+                return NGX_HTTP_V3_ERR_ENCODER_STREAM_ERROR;
             }
 
             st->state = sw_name;
