diff --git a/login_yubikey/login_yubikey.c b/login_yubikey/login_yubikey.c
index 9fddc61..19d5b7a 100644
--- a/login_yubikey/login_yubikey.c
+++ b/login_yubikey/login_yubikey.c
@@ -30,6 +30,7 @@
  *
  */
 
+#include <sys/file.h>
 #include <sys/stat.h>
 #include <sys/time.h>
 #include <sys/resource.h>
@@ -72,7 +73,7 @@ main(int argc, char *argv[])
 
 	setpriority(PRIO_PROCESS, 0, 0);
 
-	if (pledge("stdio tty wpath rpath cpath", NULL) == -1) {
+	if (pledge("stdio tty wpath rpath cpath flock", NULL) == -1) {
 		syslog(LOG_AUTH|LOG_ERR, "pledge: %m");
 		exit(EXIT_FAILURE);
 	}
@@ -221,11 +222,6 @@ yubikey_login(const char *username, const char *password)
 	}
 
 	snprintf(fn, sizeof(fn), "%s/%s.ctr", path, username);
-	if ((f = fopen(fn, "r")) != NULL) {
-		if (fscanf(f, "%u", &last_ctr) != 1)
-			last_ctr = 0;
-		fclose(f);
-	}
 
 	yubikey_hex_decode(uid, hexuid, YUBIKEY_UID_SIZE);
 	yubikey_hex_decode(key, hexkey, YUBIKEY_KEY_SIZE);
@@ -281,18 +277,32 @@ yubikey_login(const char *username, const char *password)
 	    "%d crc ok", username, mapok, i, crcok);
 
 	ctr = ((u_int32_t)yubikey_counter(tok.ctr) << 8) | tok.use;
-	if (ctr <= last_ctr) {
-		syslog(LOG_INFO, "user %s: counter <= last (REPLAY ATTACK!)",
-		    username);
-		return (AUTH_FAILED);
-	}
-	syslog(LOG_INFO, "user %s: counter > last [OK]", username);
 	umask(S_IRWXO);
-	if ((f = fopen(fn, "w")) == NULL) {
+	if ((f = fopen(fn, "a+")) == NULL) {
 		syslog(LOG_ERR, "user %s: fopen: %s: %m", username, fn);
 		return (AUTH_FAILED);
 	}
-	fprintf(f, "%u", ctr);
+	if (flock(fileno(f), LOCK_EX) == -1) {
+		syslog(LOG_ERR, "user %s: flock: %s: %m", username, fn);
+		fclose(f);
+		return (AUTH_FAILED);
+	}
+	rewind(f);
+	if (fscanf(f, "%u", &last_ctr) != 1)
+		last_ctr = 0;
+	if (ctr <= last_ctr) {
+		syslog(LOG_INFO, "user %s: counter <= last (REPLAY ATTACK!)",
+		    username);
+		fclose(f);
+		return (AUTH_FAILED);
+	}
+	syslog(LOG_INFO, "user %s: counter > last [OK]", username);
+	rewind(f);
+	if (ftruncate(fileno(f), 0) == -1 || fprintf(f, "%u", ctr) < 0) {
+		syslog(LOG_ERR, "user %s: write: %s: %m", username, fn);
+		fclose(f);
+		return (AUTH_FAILED);
+	}
 	fclose(f);
 
 	return (AUTH_OK);
