diff --git a/spamd/sync.c b/spamd/sync.c
index 5daca75..cc2425c 100644
--- a/spamd/sync.c
+++ b/spamd/sync.c
@@ -17,6 +17,7 @@
  */
 
 #include <sys/socket.h>
+#include <sys/stat.h>
 #include <sys/uio.h>
 #include <sys/ioctl.h>
 #include <sys/queue.h>
@@ -121,6 +122,7 @@ sync_init(const char *iface, const char *baddr, u_short port)
 	struct ifreq ifr;
 	struct ip_mreq mreq;
 	struct sockaddr_in *addr;
+	struct stat st;
 	char ifnam[IFNAMSIZ], *ttlstr;
 	const char *errstr;
 	struct in_addr ina;
@@ -142,15 +144,19 @@ sync_init(const char *iface, const char *baddr, u_short port)
 		}
 	}
 
-	sync_key = SHA1File(SPAM_SYNC_KEY, NULL);
-	if (sync_key == NULL) {
-		if (errno != ENOENT) {
-			fprintf(stderr, "failed to open sync key: %s\n",
-			    strerror(errno));
-			return (-1);
-		}
-		/* Use empty key by default */
-		sync_key = "";
+	if (stat(SPAM_SYNC_KEY, &st) == -1) {
+		fprintf(stderr, "failed to open sync key: %s\n",
+		    strerror(errno));
+		return (-1);
+	}
+	if (st.st_size == 0) {
+		fprintf(stderr, "failed to open sync key: empty key\n");
+		return (-1);
+	}
+	if ((sync_key = SHA1File(SPAM_SYNC_KEY, NULL)) == NULL) {
+		fprintf(stderr, "failed to open sync key: %s\n",
+		    strerror(errno));
+		return (-1);
 	}
 
 	syncfd = socket(AF_INET, SOCK_DGRAM, 0);
