diff --git a/lib/std/crypto/pcurves/p384.zig b/lib/std/crypto/pcurves/p384.zig
index 0dbfdc67f1..b9ff99e321 100644
--- a/lib/std/crypto/pcurves/p384.zig
+++ b/lib/std/crypto/pcurves/p384.zig
@@ -51,7 +51,7 @@ pub const P384 = struct {
         const yy = y.sq();
         const on_curve = @intFromBool(x3AxB.equivalent(yy));
         const is_identity = @intFromBool(x.equivalent(AffineCoordinates.identityElement.x)) & @intFromBool(y.equivalent(AffineCoordinates.identityElement.y));
-        if ((on_curve | is_identity) == 0) {
+        if (on_curve == 0) {
             return error.InvalidEncoding;
         }
         var ret = P384{ .x = x, .y = y, .z = Fe.one };
