diff --git a/net/if_spppsubr.c b/net/if_spppsubr.c
index 1ddc1ec..e089ec7 100644
--- a/net/if_spppsubr.c
+++ b/net/if_spppsubr.c
@@ -3815,6 +3815,8 @@ sppp_pap_input(struct sppp *sp, struct mbuf *m)
 		}
 		if (name_len > AUTHMAXLEN ||
 		    passwd_len > AUTHMAXLEN ||
+		    name_len != strlen(sp->hisauth.name) ||
+		    passwd_len != strlen(sp->hisauth.secret) ||
 		    bcmp(name, sp->hisauth.name, name_len) != 0 ||
 		    bcmp(passwd, sp->hisauth.secret, passwd_len) != 0) {
 			/* action scn, tld */
