diff --git a/net/if_tun.c b/net/if_tun.c
index 2dcdda1..8d4df34 100644
--- a/net/if_tun.c
+++ b/net/if_tun.c
@@ -1009,7 +1009,7 @@ tun_dev_write(dev_t dev, struct uio *uio, int ioflag, int align)
 			m_clget(m, M_DONTWAIT, alen);
 			if (!ISSET(m->m_flags, M_EXT)) {
 				error = ENOMEM;
-				goto put;
+				goto drop;
 			}
 		}
 
@@ -1031,7 +1031,7 @@ tun_dev_write(dev_t dev, struct uio *uio, int ioflag, int align)
 		n = m_get(M_DONTWAIT, MT_DATA);
 		if (n == NULL) {
 			error = ENOMEM;
-			goto put;
+			goto drop;
 		}
 
 		align = 0;
