diff --git a/net/slcompress.c b/net/slcompress.c
index 12907da..ca565ed 100644
--- a/net/slcompress.c
+++ b/net/slcompress.c
@@ -136,7 +136,11 @@ sl_compress_setup(struct slcompress *comp, int max_state)
 }
 
 #define DECODEL(f) { \
+	if (cp >= ep) \
+		goto bad; \
 	if (*cp == 0) {\
+		if (ep - cp < 3) \
+			goto bad; \
 		(f) = htonl(ntohl(f) + ((cp[1] << 8) | cp[2])); \
 		cp += 3; \
 	} else { \
@@ -145,7 +149,11 @@ sl_compress_setup(struct slcompress *comp, int max_state)
 }
 
 #define DECODES(f) { \
+	if (cp >= ep) \
+		goto bad; \
 	if (*cp == 0) {\
+		if (ep - cp < 3) \
+			goto bad; \
 		(f) = htons(ntohs(f) + ((cp[1] << 8) | cp[2])); \
 		cp += 3; \
 	} else { \
@@ -154,7 +162,11 @@ sl_compress_setup(struct slcompress *comp, int max_state)
 }
 
 #define DECODEU(f) { \
+	if (cp >= ep) \
+		goto bad; \
 	if (*cp == 0) {\
+		if (ep - cp < 3) \
+			goto bad; \
 		(f) = htons((cp[1] << 8) | cp[2]); \
 		cp += 3; \
 	} else { \
@@ -463,7 +475,7 @@ int
 sl_uncompress_tcp_core(u_char *buf, int buflen, int total_len, u_int type,
     struct slcompress *comp, u_char **hdrp, u_int *hlenp)
 {
-	u_char *cp;
+	u_char *cp, *ep;
 	u_int hlen, changes;
 	struct tcphdr *th;
 	struct cstate *cs;
@@ -506,11 +518,14 @@ sl_uncompress_tcp_core(u_char *buf, int buflen, int total_len, u_int type,
 	/* We've got a compressed packet. */
 	INCR(sls_compressedin)
 	cp = buf;
+	ep = buf + buflen;
+	if (cp >= ep)
+		goto bad;
 	changes = *cp++;
 	if (changes & NEW_C) {
 		/* Make sure the state index is in range, then grab the state.
 		 * If we have a good state index, clear the 'discard' flag. */
-		if (*cp >= MAX_STATES)
+		if (cp >= ep || *cp >= MAX_STATES)
 			goto bad;
 
 		comp->flags &=~ SLF_TOSS;
@@ -527,6 +542,8 @@ sl_uncompress_tcp_core(u_char *buf, int buflen, int total_len, u_int type,
 	cs = &comp->rstate[comp->last_recv];
 	hlen = cs->cs_ip.ip_hl << 2;
 	th = (struct tcphdr *)&((u_char *)&cs->cs_ip)[hlen];
+	if (ep - cp < 2)
+		goto bad;
 	th->th_sum = htons((*cp << 8) | cp[1]);
 	cp += 2;
 	if (changes & TCP_PUSH_BIT)
