diff --git a/netinet/if_ether.c b/netinet/if_ether.c
index 653499a..55f4a49 100644
--- a/netinet/if_ether.c
+++ b/netinet/if_ether.c
@@ -616,7 +616,7 @@ in_arpinput(struct ifnet *ifp, struct mbuf *m)
 	sin.sin_len = sizeof(sin);
 	sin.sin_family = AF_INET;
 
-	if (ETHER_IS_MULTICAST(ea->arp_sha) &&
+	if (ETHER_IS_MULTICAST(ea->arp_sha) ||
 	    ETHER_IS_BROADCAST(ea->arp_sha)) {
 		inet_ntop(AF_INET, &isaddr, addr, sizeof(addr));
 		log(LOG_ERR, "arp: ether address is broadcast for IP address "
