diff --git a/library/proc_macro/src/lib.rs b/library/proc_macro/src/lib.rs
index a01bf38a62d..fcc14a452e0 100644
--- a/library/proc_macro/src/lib.rs
+++ b/library/proc_macro/src/lib.rs
@@ -1461,7 +1461,7 @@ fn get_hashes_str(num: u8) -> &'static str {
     #[unstable(feature = "proc_macro_value", issue = "136652")]
     pub fn byte_character_value(&self) -> Result<u8, ConversionErrorKind> {
         self.0.symbol.with(|symbol| match self.0.kind {
-            bridge::LitKind::Char => {
+            bridge::LitKind::Byte => {
                 unescape_byte(symbol).map_err(ConversionErrorKind::FailedToUnescape)
             }
             _ => Err(ConversionErrorKind::InvalidLiteralKind),
