diff --git a/library/std/src/os/unix/fs.rs b/library/std/src/os/unix/fs.rs
index 219b340b924..cbee20910a9 100644
--- a/library/std/src/os/unix/fs.rs
+++ b/library/std/src/os/unix/fs.rs
@@ -208,7 +208,9 @@ fn read_buf_exact_at(&self, mut buf: BorrowedCursor<'_>, mut offset: u64) -> io:
                 Err(e) => return Err(e),
             }
             let n = buf.written() - prev_written;
-            offset += n as u64;
+            offset = offset.checked_add(n as u64).ok_or_else(|| {
+                io::Error::new(io::ErrorKind::InvalidInput, "offset overflow")
+            })?;
             if n == 0 {
                 return Err(io::Error::READ_EXACT_EOF);
             }
