diff --git a/library/std_detect/src/detect/os/x86.rs b/library/std_detect/src/detect/os/x86.rs
index b24ef6a37ef..64d1a282326 100644
--- a/library/std_detect/src/detect/os/x86.rs
+++ b/library/std_detect/src/detect/os/x86.rs
@@ -260,10 +260,10 @@ pub(crate) fn detect_features() -> cache::Initializer {
                     // "Short" versions of AVX512 instructions
                     let avxvnni = enable(extended_features_eax_leaf_1, 4, Feature::avxvnni);
                     let avxvnniint8 = enable(extended_features_eax_leaf_1, 23, Feature::avxifma);
-                    let avxvnniint16 =
-                        enable(extended_features_edx_leaf_1, 4, Feature::avxvnniint8);
+                    enable(extended_features_edx_leaf_1, 4, Feature::avxvnniint8);
                     enable(extended_features_edx_leaf_1, 5, Feature::avxneconvert);
-                    enable(extended_features_edx_leaf_1, 10, Feature::avxvnniint16);
+                    let avxvnniint16 =
+                        enable(extended_features_edx_leaf_1, 10, Feature::avxvnniint16);
 
                     enable(extended_features_eax_leaf_1, 0, Feature::sha512);
                     enable(extended_features_eax_leaf_1, 1, Feature::sm3);
