diff --git a/library/std/src/sys/net/connection/xous/tcplistener.rs b/library/std/src/sys/net/connection/xous/tcplistener.rs
index 8818ef2ca9a..55b0f269391 100644
--- a/library/std/src/sys/net/connection/xous/tcplistener.rs
+++ b/library/std/src/sys/net/connection/xous/tcplistener.rs
@@ -121,9 +121,10 @@ pub fn accept(&self) -> io::Result<(TcpStream, SocketAddr)> {
             receive_request.raw[0] = 1;
         }
 
+        let fd = self.fd.load(Ordering::Relaxed);
         if let Ok((_offset, _valid)) = crate::os::xous::ffi::lend_mut(
             services::net_server(),
-            services::NetLendMut::StdTcpAccept(self.fd.load(Ordering::Relaxed)).into(),
+            services::NetLendMut::StdTcpAccept(fd).into(),
             &mut receive_request.raw,
             0,
             0,
@@ -167,7 +168,17 @@ pub fn accept(&self) -> io::Result<(TcpStream, SocketAddr)> {
                 // replenish the listener
                 let mut local_copy = self.local.clone(); // port is non-0 by this time, but the method signature needs a mut
                 let new_fd = TcpListener::bind_inner(&mut local_copy)?;
-                self.fd.store(new_fd, Ordering::Relaxed);
+                if self
+                    .fd
+                    .compare_exchange(fd, new_fd, Ordering::Relaxed, Ordering::Relaxed)
+                    .is_err()
+                {
+                    crate::os::xous::ffi::blocking_scalar(
+                        services::net_server(),
+                        services::NetBlockingScalar::StdTcpClose(new_fd).into(),
+                    )
+                    .unwrap();
+                }
 
                 // now return a stream converted from the old stream's fd
                 Ok((TcpStream::from_listener(stream_fd, self.local.port(), port, addr), addr))
