diff --git a/library/core/src/time.rs b/library/core/src/time.rs
index 9073131b35b..f5f9df0439a 100644
--- a/library/core/src/time.rs
+++ b/library/core/src/time.rs
@@ -1243,6 +1243,9 @@ pub const fn div_duration_floor(self, rhs: Duration) -> u128 {
                   without modifying the original"]
     #[inline]
     pub const fn div_duration_ceil(self, rhs: Duration) -> u128 {
+        if rhs.is_zero() {
+            panic!("divide by zero error when dividing durations");
+        }
         self.as_nanos().div_ceil(rhs.as_nanos())
     }
 }
