diff --git a/library/std/src/sys/net/connection/xous/udp.rs b/library/std/src/sys/net/connection/xous/udp.rs
index ce54ea3b79e..4d89243477b 100644
--- a/library/std/src/sys/net/connection/xous/udp.rs
+++ b/library/std/src/sys/net/connection/xous/udp.rs
@@ -240,6 +240,9 @@ pub fn send_to(&self, buf: &[u8], addr: &SocketAddr) -> io::Result<usize> {
                 }
             }
         }
+        if buf.len() > tx_req.raw[21..].len() {
+            return Err(io::const_error!(io::ErrorKind::InvalidInput, "payload too large"));
+        }
         let len = buf.len() as u16;
         let len_bytes = len.to_le_bytes();
         tx_req.raw[19] = len_bytes[0];
