diff --git a/library/std_detect/src/detect/os/aarch64.rs b/library/std_detect/src/detect/os/aarch64.rs
index 3232e435d52..08ec4efa04f 100644
--- a/library/std_detect/src/detect/os/aarch64.rs
+++ b/library/std_detect/src/detect/os/aarch64.rs
@@ -93,19 +93,20 @@ pub(crate) fn parse_system_registers(
         let fphp = bits_shift(aa64pfr0, 19, 16) >= 1;
         let asimd = bits_shift(aa64pfr0, 23, 20) < 0xF;
         let asimdhp = bits_shift(aa64pfr0, 23, 20) >= 1;
+        let simd = fp && asimd && (!fphp | asimdhp);
         enable_feature(Feature::fp, fp);
         enable_feature(Feature::fp16, fphp);
         // SIMD support requires float support - if half-floats are
         // supported, it also requires half-float support:
-        enable_feature(Feature::asimd, fp && asimd && (!fphp | asimdhp));
+        enable_feature(Feature::asimd, simd);
         // SIMD extensions require SIMD support:
-        enable_feature(Feature::aes, asimd && bits_shift(aa64isar0, 7, 4) >= 2);
+        enable_feature(Feature::aes, simd && bits_shift(aa64isar0, 7, 4) >= 2);
         let sha1 = bits_shift(aa64isar0, 11, 8) >= 1;
         let sha2 = bits_shift(aa64isar0, 15, 12) >= 1;
-        enable_feature(Feature::sha2, asimd && sha1 && sha2);
-        enable_feature(Feature::rdm, asimd && bits_shift(aa64isar0, 31, 28) >= 1);
-        enable_feature(Feature::dotprod, asimd && bits_shift(aa64isar0, 47, 44) >= 1);
-        enable_feature(Feature::sve, asimd && bits_shift(aa64pfr0, 35, 32) >= 1);
+        enable_feature(Feature::sha2, simd && sha1 && sha2);
+        enable_feature(Feature::rdm, simd && bits_shift(aa64isar0, 31, 28) >= 1);
+        enable_feature(Feature::dotprod, simd && bits_shift(aa64isar0, 47, 44) >= 1);
+        enable_feature(Feature::sve, simd && bits_shift(aa64pfr0, 35, 32) >= 1);
     }
 
     // ID_AA64ISAR1_EL1 - Instruction Set Attribute Register 1
