diff --git a/library/compiler-builtins/builtins-test-intrinsics/Cargo.toml b/library/compiler-builtins/builtins-test-intrinsics/Cargo.toml
index fed2ac39fb3..633d2537c5f 100644
--- a/library/compiler-builtins/builtins-test-intrinsics/Cargo.toml
+++ b/library/compiler-builtins/builtins-test-intrinsics/Cargo.toml
@@ -11,9 +11,9 @@ compiler_builtins = { path = "../builtins-shim", features = ["compiler-builtins"
 panic-handler = { path = "../crates/panic-handler" }
 
 [target.'cfg(all(target_arch = "arm", not(any(target_env = "gnu", target_env = "musl")), target_os = "linux"))'.dev-dependencies]
-test = { git = "https://github.com/japaric/utest" }
-utest-cortex-m-qemu = { default-features = false, git = "https://github.com/japaric/utest" }
-utest-macros = { git = "https://github.com/japaric/utest" }
+test = { git = "https://github.com/japaric/utest", rev = "e32073e2b078e3bee46001c13ae4c1acf368d762" }
+utest-cortex-m-qemu = { default-features = false, git = "https://github.com/japaric/utest", rev = "e32073e2b078e3bee46001c13ae4c1acf368d762" }
+utest-macros = { git = "https://github.com/japaric/utest", rev = "e32073e2b078e3bee46001c13ae4c1acf368d762" }
 
 [features]
 c = ["compiler_builtins/c"]
