diff --git a/library/stdarch/crates/stdarch-gen-loongarch/src/main.rs b/library/stdarch/crates/stdarch-gen-loongarch/src/main.rs
index 3a946a12d66..e3d4f628c86 100644
--- a/library/stdarch/crates/stdarch-gen-loongarch/src/main.rs
+++ b/library/stdarch/crates/stdarch-gen-loongarch/src/main.rs
@@ -761,7 +761,7 @@ union v4sf
 {{
     __m128 v;
     int64_t i64[2];
-    uint32_t i32[2];
+    uint32_t i32[4];
     float f32[4];
 }};
