diff --git a/library/stdarch/crates/stdarch-gen-loongarch/src/main.rs b/library/stdarch/crates/stdarch-gen-loongarch/src/main.rs
index 3a946a12d66..e7310c53a1a 100644
--- a/library/stdarch/crates/stdarch-gen-loongarch/src/main.rs
+++ b/library/stdarch/crates/stdarch-gen-loongarch/src/main.rs
@@ -769,7 +769,7 @@ union v8sf
 {{
     __m256 v;
     int64_t i64[4];
-    uint32_t i32[4];
+    uint32_t i32[8];
     float f32[8];
 }};
