diff --git a/library/stdarch/crates/core_arch/src/aarch64/mte.rs b/library/stdarch/crates/core_arch/src/aarch64/mte.rs
index a5031a45c1a..bde88c2125d 100644
--- a/library/stdarch/crates/core_arch/src/aarch64/mte.rs
+++ b/library/stdarch/crates/core_arch/src/aarch64/mte.rs
@@ -51,6 +51,7 @@ pub unsafe fn __arm_mte_create_random_tag<T>(src: *const T, mask: u64) -> *const
 #[target_feature(enable = "mte")]
 #[unstable(feature = "stdarch_aarch64_mte", issue = "129010")]
 pub unsafe fn __arm_mte_increment_tag<const OFFSET: i64, T>(src: *const T) -> *const T {
+    static_assert_range!(OFFSET, 0..=15);
     addg_(src as *const (), OFFSET) as *const T
 }
