diff --git a/library/stdarch/crates/core_arch/src/s390x/vector.rs b/library/stdarch/crates/core_arch/src/s390x/vector.rs
index fc5af1b14d0..8b24547ebbb 100644
--- a/library/stdarch/crates/core_arch/src/s390x/vector.rs
+++ b/library/stdarch/crates/core_arch/src/s390x/vector.rs
@@ -5888,7 +5888,7 @@ pub unsafe fn vec_all_nge<T: sealed::VectorCompare>(a: T, b: T) -> i32 {
 #[target_feature(enable = "vector")]
 #[unstable(feature = "stdarch_s390x", issue = "135681")]
 pub unsafe fn vec_any_lt<T: sealed::VectorCompare>(a: T, b: T) -> i32 {
-    !vec_all_ge(a, b)
+    i32::from(vec_all_ge(a, b) == 0)
 }
 
 /// Any Elements Less Than or Equal
@@ -5896,7 +5896,7 @@ pub unsafe fn vec_any_lt<T: sealed::VectorCompare>(a: T, b: T) -> i32 {
 #[target_feature(enable = "vector")]
 #[unstable(feature = "stdarch_s390x", issue = "135681")]
 pub unsafe fn vec_any_le<T: sealed::VectorCompare>(a: T, b: T) -> i32 {
-    !vec_all_gt(a, b)
+    i32::from(vec_all_gt(a, b) == 0)
 }
 
 /// Any Elements Greater Than
@@ -5904,7 +5904,7 @@ pub unsafe fn vec_any_le<T: sealed::VectorCompare>(a: T, b: T) -> i32 {
 #[target_feature(enable = "vector")]
 #[unstable(feature = "stdarch_s390x", issue = "135681")]
 pub unsafe fn vec_any_gt<T: sealed::VectorCompare>(a: T, b: T) -> i32 {
-    !vec_all_le(a, b)
+    i32::from(vec_all_le(a, b) == 0)
 }
 
 /// Any Elements Greater Than or Equal
@@ -5912,7 +5912,7 @@ pub unsafe fn vec_any_gt<T: sealed::VectorCompare>(a: T, b: T) -> i32 {
 #[target_feature(enable = "vector")]
 #[unstable(feature = "stdarch_s390x", issue = "135681")]
 pub unsafe fn vec_any_ge<T: sealed::VectorCompare>(a: T, b: T) -> i32 {
-    !vec_all_lt(a, b)
+    i32::from(vec_all_lt(a, b) == 0)
 }
 
 /// Any Elements Not Less Than
