diff --git a/library/std/src/sys/pal/hermit/mod.rs b/library/std/src/sys/pal/hermit/mod.rs
index 53f6ddd7065..51ecd06ffaa 100644
--- a/library/std/src/sys/pal/hermit/mod.rs
+++ b/library/std/src/sys/pal/hermit/mod.rs
@@ -88,7 +88,7 @@ fn is_negative(&self) -> bool {
         }
 
         fn negate(&self) -> i32 {
-            i32::try_from(-(*self)).unwrap()
+            self.checked_neg().and_then(|n| i32::try_from(n).ok()).unwrap_or(i32::MAX)
         }
     })*)
 }
@@ -99,7 +99,7 @@ fn is_negative(&self) -> bool {
     }
 
     fn negate(&self) -> i32 {
-        -(*self)
+        self.checked_neg().unwrap_or(i32::MAX)
     }
 }
 impl_is_negative! { i8 i16 i64 isize }
