diff --git a/library/std/src/sys/process/unix/fuchsia.rs b/library/std/src/sys/process/unix/fuchsia.rs
index 3fae5ec1468..ca53a530f46 100644
--- a/library/std/src/sys/process/unix/fuchsia.rs
+++ b/library/std/src/sys/process/unix/fuchsia.rs
@@ -85,7 +85,11 @@ unsafe fn do_exec(
                 zx_cvt(status)?;
 
                 let mut cloned_fd = 0;
-                zx_cvt(fdio_fd_create(handle, &mut cloned_fd))?;
+                let status = fdio_fd_create(handle, &mut cloned_fd);
+                if status < 0 {
+                    zx_handle_close(handle);
+                }
+                zx_cvt(status)?;
 
                 Ok(fdio_spawn_action_t {
                     action: FDIO_SPAWN_ACTION_TRANSFER_FD,
