diff --git a/library/test/src/bench.rs b/library/test/src/bench.rs
index 62e51026b81..a0a62960a46 100644
--- a/library/test/src/bench.rs
+++ b/library/test/src/bench.rs
@@ -230,7 +230,7 @@ pub fn benchmark<F>(
         Ok(Err(_)) => TestResult::TrFailed,
     };
 
-    let stdout = data.lock().unwrap().to_vec();
+    let stdout = data.lock().unwrap_or_else(|err| err.into_inner()).to_vec();
     let message = CompletedTest::new(id, desc, test_result, None, stdout);
     monitor_ch.send(message).unwrap();
 }
