diff --git a/library/std/src/sys/stdio/xous.rs b/library/std/src/sys/stdio/xous.rs
index a92167642b7..b713930590f 100644
--- a/library/std/src/sys/stdio/xous.rs
+++ b/library/std/src/sys/stdio/xous.rs
@@ -96,7 +96,7 @@ fn write(&mut self, s: &[u8]) -> core::result::Result<usize, io::Error> {
             for (&s, d) in s.iter().zip(request.0.iter_mut()) {
                 *d = s;
             }
-            try_lend(gfx, 0 /* AppendPanicText */, &request.0, 0, s.len()).ok();
+            try_lend(gfx, 0 /* AppendPanicText */, &request.0, 0, s.len().min(request.0.len())).ok();
         }
         Ok(s.len())
     }
