diff --git a/library/std_detect/src/detect/os/linux/mod.rs b/library/std_detect/src/detect/os/linux/mod.rs
index aec94f963f5..25e791ce3de 100644
--- a/library/std_detect/src/detect/os/linux/mod.rs
+++ b/library/std_detect/src/detect/os/linux/mod.rs
@@ -21,6 +21,7 @@ fn read_file(orig_path: &str) -> Result<Vec<u8>, alloc::string::String> {
             data.reserve(4096);
             let spare = data.spare_capacity_mut();
             match libc::read(file, spare.as_mut_ptr() as *mut _, spare.len()) {
+                -1 if *libc::__errno_location() == libc::EINTR => continue,
                 -1 => {
                     libc::close(file);
                     return Err(format!("Error while reading from file at {orig_path}"));
