diff --git a/library/compiler-builtins/ci/download-compiler-rt.sh b/library/compiler-builtins/ci/download-compiler-rt.sh
index 87b337d8255..5d83096d1aa 100755
--- a/library/compiler-builtins/ci/download-compiler-rt.sh
+++ b/library/compiler-builtins/ci/download-compiler-rt.sh
@@ -5,6 +5,8 @@
 set -eux
 
 rust_llvm_version=20.1-2025-02-13
+rust_llvm_sha256=9145e8a2904cc2f9bff911a798497edf34bd4a39089882b116ea042cbb9de9ca
 
 curl -L --retry 3 -o code.tar.gz "https://github.com/rust-lang/llvm-project/archive/rustc/${rust_llvm_version}.tar.gz"
+echo "${rust_llvm_sha256}  code.tar.gz" | sha256sum --check --status
 tar xzf code.tar.gz --strip-components 1 llvm-project-rustc-${rust_llvm_version}/compiler-rt
