diff --git a/library/core/src/slice/sort/shared/smallsort.rs b/library/core/src/slice/sort/shared/smallsort.rs
index e555fce4408..ee951ecb1c3 100644
--- a/library/core/src/slice/sort/shared/smallsort.rs
+++ b/library/core/src/slice/sort/shared/smallsort.rs
@@ -608,7 +608,8 @@ pub fn insertion_sort_shift_left<T, F: FnMut(&T, &T) -> bool>(
 }
 
 /// SAFETY: The caller MUST guarantee that `v_base` is valid for 4 reads and
-/// `dst` is valid for 4 writes. The result will be stored in `dst[0..4]`.
+/// `dst` is valid for 4 writes. The regions `v_base[0..4]` and `dst[0..4]`
+/// MUST NOT overlap. The result will be stored in `dst[0..4]`.
 pub unsafe fn sort4_stable<T, F: FnMut(&T, &T) -> bool>(
     v_base: *const T,
     dst: *mut T,
