diff --git a/library/std/src/sys/pal/uefi/system_time.rs b/library/std/src/sys/pal/uefi/system_time.rs
index 557a49b27c2..9dfb9f75050 100644
--- a/library/std/src/sys/pal/uefi/system_time.rs
+++ b/library/std/src/sys/pal/uefi/system_time.rs
@@ -36,7 +36,7 @@ pub(crate) const fn from_uefi(t: &Time) -> Option<Duration> {
         && t.day <= 31
         && t.day != 0
         && t.second < 60
-        && t.minute <= 60
+        && t.minute < 60
         && t.hour < 24
         && t.nanosecond < 1_000_000_000
         && ((t.timezone <= 1440 && t.timezone >= -1440)
