diff --git a/library/stdarch/crates/stdarch-gen-arm/src/input.rs b/library/stdarch/crates/stdarch-gen-arm/src/input.rs
index adefbf3215b..034098e4c0f 100644
--- a/library/stdarch/crates/stdarch-gen-arm/src/input.rs
+++ b/library/stdarch/crates/stdarch-gen-arm/src/input.rs
@@ -70,9 +70,7 @@ fn cmp(&self, other: &Self) -> std::cmp::Ordering {
             (InputType::PredicateForm(pf1), InputType::PredicateForm(pf2)) => pf1.cmp(pf2),
             (InputType::Type(ty1), InputType::Type(ty2)) => ty1.cmp(ty2),
 
-            (InputType::NVariantOp(None), InputType::NVariantOp(Some(..))) => Less,
-            (InputType::NVariantOp(Some(..)), InputType::NVariantOp(None)) => Greater,
-            (InputType::NVariantOp(_), InputType::NVariantOp(_)) => Equal,
+            (InputType::NVariantOp(op1), InputType::NVariantOp(op2)) => op1.cmp(op2),
 
             (InputType::Type(..), InputType::PredicateForm(..)) => Less,
             (InputType::PredicateForm(..), InputType::Type(..)) => Greater,
