diff --git a/library/stdarch/crates/stdarch-gen-arm/src/wildcards.rs b/library/stdarch/crates/stdarch-gen-arm/src/wildcards.rs
index 6c40d88df45..fc88f6b09b1 100644
--- a/library/stdarch/crates/stdarch-gen-arm/src/wildcards.rs
+++ b/library/stdarch/crates/stdarch-gen-arm/src/wildcards.rs
@@ -89,8 +89,9 @@ fn from_str(s: &str) -> Result<Self, Self::Err> {
                 ("type", index, None, None) => Ok(Wildcard::Type(index)),
                 ("neon_type", index, tuple, modifier) => {
                     if let Some(str_suffix) = modifier {
-                        let suffix_kind = SuffixKind::from_str(str_suffix);
-                        return Ok(Wildcard::NEONType(index, tuple, Some(suffix_kind.unwrap())));
+                        let suffix_kind = SuffixKind::from_str(str_suffix)
+                            .map_err(|_| format!("{str_suffix:#?} is not a valid suffix"))?;
+                        return Ok(Wildcard::NEONType(index, tuple, Some(suffix_kind)));
                     } else {
                         Ok(Wildcard::NEONType(index, tuple, None))
                     }
