diff --git a/library/compiler-builtins/libm-test/src/generate/random.rs b/library/compiler-builtins/libm-test/src/generate/random.rs
index 32bd2f24ee0..cfeb53bbea8 100644
--- a/library/compiler-builtins/libm-test/src/generate/random.rs
+++ b/library/compiler-builtins/libm-test/src/generate/random.rs
@@ -160,7 +160,7 @@ fn get_cases(ctx: &CheckCtx) -> (impl Iterator<Item = Self>, u64) {
                 let iter0 = random_ints(count0, range0);
                 let iter1 = random_ints(count1, range1.clone());
                 let iter = product2(iter0, iter1);
-                (iter, count0 * count1)
+                (iter, count0.strict_mul(count1))
             }
         }
     };
