diff --git a/library/compiler-builtins/libm/src/math/generic/fma.rs b/library/compiler-builtins/libm/src/math/generic/fma.rs
index aaf459d1b61..73817ac6ae5 100644
--- a/library/compiler-builtins/libm/src/math/generic/fma.rs
+++ b/library/compiler-builtins/libm/src/math/generic/fma.rs
@@ -71,9 +71,10 @@ pub fn fma_round<F>(x: F, y: F, z: F, _round: Round) -> FpResult<F>
                 // Exactly `sbits`, nothing to do
             } else if d < sbits {
                 // Remaining shift fits within `sbits`. Leave `z` in place, shift `x * y`
+                let rlo_orig = rlo;
                 rlo = (rhi << (sbits - d)) | (rlo >> d);
                 // Set the sticky bit
-                rlo |= IntTy::<F>::from((rlo << (sbits - d)) != zero);
+                rlo |= IntTy::<F>::from((rlo_orig << (sbits - d)) != zero);
                 rhi = rhi >> d;
             } else {
                 // `z`'s magnitude is enough that `x * y` is irrelevant. It was nonzero, so set
