diff --git a/library/core/src/iter/adapters/filter_map.rs b/library/core/src/iter/adapters/filter_map.rs
index 3fd7e16d1c1..27ef60405bd 100644
--- a/library/core/src/iter/adapters/filter_map.rs
+++ b/library/core/src/iter/adapters/filter_map.rs
@@ -70,6 +70,10 @@ fn next_chunk<const N: usize>(
     ) -> Result<[Self::Item; N], array::IntoIter<Self::Item, N>> {
         let mut array: [MaybeUninit<Self::Item>; N] = [const { MaybeUninit::uninit() }; N];
 
+        if N == 0 {
+            return Ok(unsafe { MaybeUninit::array_assume_init(array) });
+        }
+
         struct Guard<'a, T> {
             array: &'a mut [MaybeUninit<T>],
             initialized: usize,
