diff --git a/library/core/src/pat.rs b/library/core/src/pat.rs
index 2670c261419..be3bd1f0eb1 100644
--- a/library/core/src/pat.rs
+++ b/library/core/src/pat.rs
@@ -70,6 +70,9 @@ impl const RangePattern for char {
     const MAX: Self = char::MAX;
 
     fn sub_one(self) -> Self {
+        if self == char::MIN {
+            panic!("exclusive range to start of valid chars");
+        }
         match char::from_u32(self as u32 - 1) {
             None => panic!("exclusive range to start of valid chars"),
             Some(val) => val,
