diff --git a/library/std/src/sys/args/zkvm.rs b/library/std/src/sys/args/zkvm.rs
index d26bf1eaff9..a1054fc2887 100644
--- a/library/std/src/sys/args/zkvm.rs
+++ b/library/std/src/sys/args/zkvm.rs
@@ -16,7 +16,7 @@ fn get_args() -> Vec<&'static OsStr> {
         // Get the size of the argument then the data.
         let arg_len = unsafe { abi::sys_argv(ptr::null_mut(), 0, i) };
 
-        let arg_len_words = (arg_len + WORD_SIZE - 1) / WORD_SIZE;
+        let arg_len_words = arg_len.checked_add(WORD_SIZE - 1).expect("argument length overflow") / WORD_SIZE;
         let words = unsafe { abi::sys_alloc_words(arg_len_words) };
 
         let arg_len2 = unsafe { abi::sys_argv(words, arg_len_words, i) };
