diff --git a/usr.sbin/syslogd/syslogd.c b/usr.sbin/syslogd/syslogd.c
index dfa741c..67bca2b 100644
--- a/usr.sbin/syslogd/syslogd.c
+++ b/usr.sbin/syslogd/syslogd.c
@@ -1143,7 +1143,8 @@ acceptcb(int lfd, short event, void *arg, int usetls)
 	int			 fd, error;
 
 	sslen = sizeof(ss);
-	if ((fd = reserve_accept4(lfd, event, ev, tcp_acceptcb,
+	if ((fd = reserve_accept4(lfd, event, ev,
+	    usetls ? tls_acceptcb : tcp_acceptcb,
 	    (struct sockaddr *)&ss, &sslen, SOCK_NONBLOCK)) == -1) {
 		if (errno != ENFILE && errno != EMFILE &&
 		    errno != EINTR && errno != EWOULDBLOCK &&
