diff --git a/usr.sbin/rpki-client/main.c b/usr.sbin/rpki-client/main.c
index 26bb511..4a2a98b 100644
--- a/usr.sbin/rpki-client/main.c
+++ b/usr.sbin/rpki-client/main.c
@@ -517,22 +517,25 @@ queue_add_from_cert(const struct cert *cert, struct nca_tree *ncas)
 	struct fqdnlistentry	*le;
 	char			*nfile, *npath, *host;
 	const char		*uri, *repouri, *file;
-	size_t			 repourisz;
+	size_t			 repourisz, hostsz;
 	int			 shortlisted = 0;
 
 	if (strncmp(cert->repo, RSYNC_PROTO, RSYNC_PROTO_LEN) != 0)
 		errx(1, "unexpected protocol");
 	host = cert->repo + RSYNC_PROTO_LEN;
+	hostsz = strcspn(host, "/");
 
 	LIST_FOREACH(le, &skiplist, entry) {
-		if (strncasecmp(host, le->fqdn, strcspn(host, "/")) == 0) {
+		if (strlen(le->fqdn) == hostsz &&
+		    strncasecmp(host, le->fqdn, hostsz) == 0) {
 			warnx("skipping %s (listed in skiplist)", cert->repo);
 			return;
 		}
 	}
 
 	LIST_FOREACH(le, &shortlist, entry) {
-		if (strncasecmp(host, le->fqdn, strcspn(host, "/")) == 0) {
+		if (strlen(le->fqdn) == hostsz &&
+		    strncasecmp(host, le->fqdn, hostsz) == 0) {
 			shortlisted = 1;
 			break;
 		}
