diff --git a/bin/ksh/main.c b/bin/ksh/main.c
index b011b22..e55e227 100644
--- a/bin/ksh/main.c
+++ b/bin/ksh/main.c
@@ -393,7 +393,8 @@ main(int argc, char *argv[])
 	getopts_reset(1);
 
 	/* Disable during .profile/ENV reading */
-	restricted = Flag(FRESTRICTED);
+	restricted = Flag(FRESTRICTED) || is_restricted(argv[0]) ||
+	    is_restricted(str_val(global("SHELL")));
 	Flag(FRESTRICTED) = 0;
 	errexit = Flag(FERREXIT);
 	Flag(FERREXIT) = 0;
@@ -412,7 +413,7 @@ main(int argc, char *argv[])
 
 	if (Flag(FPRIVILEGED))
 		include("/etc/suid_profile", 0, NULL, 1);
-	else if (Flag(FTALKING)) {
+	else if (Flag(FTALKING) && !restricted) {
 		char *env_file;
 
 		/* include $ENV */
