diff --git a/sbin/iked/ocsp.c b/sbin/iked/ocsp.c
index 91c44fc..b1e2c71 100644
--- a/sbin/iked/ocsp.c
+++ b/sbin/iked/ocsp.c
@@ -570,8 +570,7 @@ ocsp_parse_response(struct iked_ocsp *ocsp, OCSP_RESPONSE *resp)
 		errstr = "no status found";
 		goto done;
 	}
-	if (env->sc_ocsp_tolerate &&
-	    !OCSP_check_validity(thisupd, nextupd, env->sc_ocsp_tolerate,
+	if (!OCSP_check_validity(thisupd, nextupd, env->sc_ocsp_tolerate,
 	    env->sc_ocsp_maxage)) {
 		ca_sslerror(SPI_SH(&ocsp->ocsp_sh, __func__));
 		errstr = "status times invalid";
