diff --git a/usr.bin/ssh/sftp-server.c b/usr.bin/ssh/sftp-server.c
index 33062da..5c9c54c 100644
--- a/usr.bin/ssh/sftp-server.c
+++ b/usr.bin/ssh/sftp-server.c
@@ -731,6 +731,8 @@ process_init(void)
 static void
 process_open(uint32_t id)
 {
+	static const struct sftp_handler write_handler =
+	    { "write", NULL, SSH2_FXP_WRITE, process_write, 1 };
 	uint32_t pflags;
 	Attrib a;
 	char *name;
@@ -746,7 +748,11 @@ process_open(uint32_t id)
 	mode = (a.flags & SSH2_FILEXFER_ATTR_PERMISSIONS) ? a.perm : 0666;
 	logit("open \"%s\" flags %s mode 0%o",
 	    name, string_from_portable(pflags), mode);
-	if (readonly &&
+	if (((flags & O_ACCMODE) != O_RDONLY ||
+	    (flags & (O_CREAT|O_TRUNC)) != 0) &&
+	    !request_permitted(&write_handler)) {
+		status = SSH2_FX_PERMISSION_DENIED;
+	} else if (readonly &&
 	    ((flags & O_ACCMODE) != O_RDONLY ||
 	    (flags & (O_CREAT|O_TRUNC)) != 0)) {
 		verbose("Refusing open request in read-only mode");
