diff --git a/usr.sbin/vmd/virtio.c b/usr.sbin/vmd/virtio.c
index 54c1c1d..1ce49f0 100644
--- a/usr.sbin/vmd/virtio.c
+++ b/usr.sbin/vmd/virtio.c
@@ -244,8 +244,13 @@ virtio_update_qa(struct virtio_dev *dev)
 	/* Update any host va mappings. */
 	if (vq_info->q_gpa > 0) {
 		hva = hvaddr_mem(vq_info->q_gpa, vring_size(vq_info->qs));
-		if (hva == NULL)
-			fatalx("%s: failed to translate gpa to hva", __func__);
+		if (hva == NULL) {
+			log_warnx("%s: failed to translate gpa to hva", __func__);
+			vq_info->vq_enabled = 0;
+			vq_info->q_hva = NULL;
+			dev->status |= DEVICE_NEEDS_RESET;
+			return;
+		}
 		vq_info->q_hva = hva;
 	} else {
 		vq_info->q_hva = NULL;
