diff --git a/usr.sbin/vmd/virtio.c b/usr.sbin/vmd/virtio.c
index 54c1c1d..dbcf9cb 100644
--- a/usr.sbin/vmd/virtio.c
+++ b/usr.sbin/vmd/virtio.c
@@ -212,7 +212,15 @@ virtio_update_qa(struct virtio_dev *dev)
 		vq_info->qs = dev->pci_cfg.queue_size;
 		vq_info->mask = vq_info->qs - 1;
 
-		if (vq_info->qs > 0 && vq_info->qs % 2 == 0) {
+		if (vq_info->qs > 0 && vq_info->qs % 2 == 0 &&
+		    dev->pci_cfg.queue_avail >= dev->pci_cfg.queue_desc &&
+		    dev->pci_cfg.queue_used >= dev->pci_cfg.queue_desc &&
+		    dev->pci_cfg.queue_avail - dev->pci_cfg.queue_desc <=
+		    vring_size(vq_info->qs) - sizeof(uint16_t) *
+		    (2 + vq_info->qs) &&
+		    dev->pci_cfg.queue_used - dev->pci_cfg.queue_desc <=
+		    vring_size(vq_info->qs) - (sizeof(uint16_t) * 2 +
+		    sizeof(struct vring_used_elem) * vq_info->qs)) {
 			vq_info->vq_availoffset = dev->pci_cfg.queue_avail -
 			    dev->pci_cfg.queue_desc;
 			vq_info->vq_usedoffset = dev->pci_cfg.queue_used -
