diff --git a/usr.bin/ssh/ssh-keygen.c b/usr.bin/ssh/ssh-keygen.c
index 25ee558..8c1af54 100644
--- a/usr.bin/ssh/ssh-keygen.c
+++ b/usr.bin/ssh/ssh-keygen.c
@@ -3061,6 +3061,12 @@ sk_suffix(const char *application, const uint8_t *user, size_t userlen)
 		ret =  xstrdup(p);
 	else
 		ret = xstrdup(application);
+	if (strchr(ret, '/') != NULL || strstr(ret, "..") != NULL ||
+	    strchr(ret, '\\') != NULL) {
+		cp = tohex(ret, strlen(ret));
+		free(ret);
+		ret = cp;
+	}
 
 	/* Count trailing zeros in user */
 	for (i = 0; i < userlen; i++) {
