diff --git a/usr.sbin/bgpd/session_bgp.c b/usr.sbin/bgpd/session_bgp.c
index 65bd9a9..8ba7325 100644
--- a/usr.sbin/bgpd/session_bgp.c
+++ b/usr.sbin/bgpd/session_bgp.c
@@ -602,7 +602,7 @@ parse_header(struct ibuf *msg, void *arg, int *fd)
 		return (NULL);
 	}
 
-	if (peer->capa.ann.ext_msg)
+	if (peer->capa.neg.ext_msg)
 		maxlen = MAX_EXT_PKTSIZE;
 
 	if (len < MSGSIZE_HEADER || len > maxlen) {
