diff --git a/usr.sbin/pcidump/pcidump.c b/usr.sbin/pcidump/pcidump.c
index b19e467..127861f 100644
--- a/usr.sbin/pcidump/pcidump.c
+++ b/usr.sbin/pcidump/pcidump.c
@@ -703,7 +703,7 @@ void
 dump_caplist(int bus, int dev, int func, u_int8_t ptr)
 {
 	u_int32_t reg;
-	u_int8_t cap;
+	u_int8_t cap, seen[256] = { 0 };
 
 	if (pci_read(bus, dev, func, PCI_COMMAND_STATUS_REG, &reg) != 0)
 		return;
@@ -714,6 +714,9 @@ dump_caplist(int bus, int dev, int func, u_int8_t ptr)
 		return;
 	ptr = PCI_CAPLIST_PTR(reg);
 	while (ptr != 0) {
+		if (ptr > 0xfc || (ptr & 0x03) || seen[ptr])
+			return;
+		seen[ptr] = 1;
 		if (pci_read(bus, dev, func, ptr, &reg) != 0)
 			return;
 		cap = PCI_CAPLIST_CAP(reg);
