diff --git a/usr.sbin/pcidump/pcidump.c b/usr.sbin/pcidump/pcidump.c
index b19e467..b91e589 100644
--- a/usr.sbin/pcidump/pcidump.c
+++ b/usr.sbin/pcidump/pcidump.c
@@ -668,10 +668,15 @@ dump_pcie_enhanced_caplist(int bus, int dev, int func)
 	u_int32_t capidx;
 	u_int16_t ptr;
 	u_int16_t ecap;
+	u_int8_t seen[0x1000] = { 0 };
 
 	ptr = PCI_PCIE_ECAP;
 
 	do {
+		if (ptr >= nitems(seen) || seen[ptr])
+			return;
+		seen[ptr] = 1;
+
 		if (pci_read(bus, dev, func, ptr, &reg) != 0)
 			return;
