diff --git a/usr.bin/ftp/ftp.c b/usr.bin/ftp/ftp.c
index c6832cc..905e2d4 100644
--- a/usr.bin/ftp/ftp.c
+++ b/usr.bin/ftp/ftp.c
@@ -1687,6 +1687,17 @@ dataconn(const char *lmode)
 		(void)close(data), data = -1;
 		return (NULL);
 	}
+	if (from.sa.sa_family != hisctladdr.sa.sa_family ||
+	    (from.sa.sa_family == AF_INET &&
+	    from.sin.sin_addr.s_addr != hisctladdr.sin.sin_addr.s_addr) ||
+	    (from.sa.sa_family == AF_INET6 &&
+	    memcmp(&from.sin6.sin6_addr, &hisctladdr.sin6.sin6_addr,
+	    sizeof(from.sin6.sin6_addr)) != 0)) {
+		warnx("data connection from wrong host");
+		(void)close(s);
+		(void)close(data), data = -1;
+		return (NULL);
+	}
 	(void)close(data);
 	data = s;
 	if (from.sa.sa_family == AF_INET) {
