diff --git a/usr.sbin/installboot/i386_installboot.c b/usr.sbin/installboot/i386_installboot.c
index d1830fb..a930967 100644
--- a/usr.sbin/installboot/i386_installboot.c
+++ b/usr.sbin/installboot/i386_installboot.c
@@ -665,6 +665,8 @@ findgptefisys(int devfd, struct disklabel *dl, int *gpartp,
 
 	off = letoh64(gh.gh_part_lba) * dl->d_secsize;
 	ghpartsize = letoh32(gh.gh_part_size);
+	if (ghpartsize < sizeof(struct gpt_partition))
+		return (-1);
 	ghpartspersec = dl->d_secsize / ghpartsize;
 	ghpartnum = letoh32(gh.gh_part_num);
 	if ((secbuf = malloc(dl->d_secsize)) == NULL)
