diff --git a/usr.sbin/iscsid/vscsi.c b/usr.sbin/iscsid/vscsi.c
index c19b50c..9f79510 100644
--- a/usr.sbin/iscsid/vscsi.c
+++ b/usr.sbin/iscsid/vscsi.c
@@ -241,10 +241,13 @@ vscsi_callback(struct connection *c, void *arg, struct pdu *p)
 			status = VSCSI_STAT_SENSE;
 			/* stupid encoding of sense data in the data segment */
 			buf = pdu_getbuf(p, &n, PDU_DATA);
-			if (buf) {
-				size = buf[0] << 8 | buf[1];
-				buf += 2;
+			if (buf == NULL || n < 2) {
+				log_debug("vscsi_callback: bad scsi response");
+				conn_fail(c);
+				goto done;
 			}
+			size = buf[0] << 8 | buf[1];
+			buf += 2;
 			break;
 		default:
 			status = VSCSI_STAT_ERR;
@@ -279,6 +282,7 @@ send_status:
 		    t->target, t->lun);
 		log_pdu(p, 1);
 	}
+done:
 	pdu_free(p);
 }
