diff --git a/sbin/iked/ca.c b/sbin/iked/ca.c
index b6175ea..d7da572 100644
--- a/sbin/iked/ca.c
+++ b/sbin/iked/ca.c
@@ -280,6 +280,11 @@ ca_decode_cert_bundle(struct iked *env, struct iked_sahdr *sh,
 		len -= sizeof(datalen);
 
 		/* Value */
+		if (datalen == 0) {
+			log_info("%s: zero-length certificate",
+			    SPI_SH(sh, __func__));
+			goto done;
+		}
 		if (len < datalen) {
 			log_info("%s: short len %zu < datalen %zu",
 			    SPI_SH(sh, __func__), len, datalen);
