diff --git a/usr.sbin/acme-client/json.c b/usr.sbin/acme-client/json.c
index 4081269..297256b 100644
--- a/usr.sbin/acme-client/json.c
+++ b/usr.sbin/acme-client/json.c
@@ -401,7 +401,8 @@ json_parse_challenge(struct jsmnn *n, struct chng *p)
 		p->status = json_parse_response(obj);
 		if (p->status == CHNG_INVALID) {
 			error = json_getobj(obj, "error");
-			p->error = json_getstr(error, "detail");
+			if (error != NULL)
+				p->error = json_getstr(error, "detail");
 		}
 		return p->uri != NULL && p->token != NULL;
 	}
