diff --git a/usr.sbin/npppd/npppd/chap.c b/usr.sbin/npppd/npppd/chap.c
index 8716eb2..9c37ef5 100644
--- a/usr.sbin/npppd/npppd/chap.c
+++ b/usr.sbin/npppd/npppd/chap.c
@@ -428,6 +428,7 @@ static void
 chap_response(chap *_this, int authok, u_char *pktp, int lpktp)
 {
 	const char *realm_name;
+	char username[4 * MAX_USERNAME_LENGTH + 1];
 
 	CHAP_ASSERT(_this != NULL);
 	CHAP_ASSERT(pktp != NULL);
@@ -437,10 +438,11 @@ chap_response(chap *_this, int authok, u_char *pktp, int lpktp)
 	ppp_output(_this->ppp, PPP_PROTO_CHAP, (authok)? 3 : 4, _this->challid,
 	    pktp, lpktp);
 
+	strvis(username, _this->name, VIS_DQ | VIS_WHITE);
 	realm_name = npppd_ppp_get_realm_name(_this->ppp->pppd, _this->ppp);
 	if (!authok) {
 		chap_log(_this, LOG_ALERT,
-		    "logtype=Failure username=\"%s\" realm=%s", _this->name,
+		    "logtype=Failure username=\"%s\" realm=%s", username,
 		    realm_name);
 		chap_stop(_this);
 		/* Stop the PPP if the authentication is failed. */
@@ -452,7 +454,7 @@ chap_response(chap *_this, int authok, u_char *pktp, int lpktp)
 		    sizeof(_this->ppp->username));
 		chap_log(_this, LOG_INFO,
 		    "logtype=Success username=\"%s\" "
-		    "realm=%s", _this->name, realm_name);
+		    "realm=%s", username, realm_name);
 		chap_stop(_this);
 		/* We change our state to prepare to resend requests. */
 		_this->state = CHAP_STATE_SENT_RESPONSE;
