diff --git a/usr.sbin/ypldap/ypldap.c b/usr.sbin/ypldap/ypldap.c
index ccabd97..3a17bc3 100644
--- a/usr.sbin/ypldap/ypldap.c
+++ b/usr.sbin/ypldap/ypldap.c
@@ -418,7 +418,9 @@ main_dispatch_client(int fd, short events, void *p)
 			if (env->update_trashed)
 				break;
 
-			(void)memcpy(&ir, imsg.data, n - IMSG_HEADER_SIZE);
+			if (n - IMSG_HEADER_SIZE != sizeof(ir))
+				break;
+			(void)memcpy(&ir, imsg.data, sizeof(ir));
 			if ((ge = calloc(1, sizeof(*ge))) == NULL ||
 			    (ge->ge_line = strdup(ir.ir_line)) == NULL) {
 				/*
