diff --git a/usr.sbin/dhcrelay6/bpf.c b/usr.sbin/dhcrelay6/bpf.c
index ea3a133..c1f4b15 100644
--- a/usr.sbin/dhcrelay6/bpf.c
+++ b/usr.sbin/dhcrelay6/bpf.c
@@ -355,8 +355,9 @@ receive_packet(struct interface_info *interface, unsigned char *buf,
 		 * do is drop it.
 		 */
 		if (hdr.bh_caplen != hdr.bh_datalen) {
-			interface->rbuf_offset += hdr.bh_hdrlen =
-			    hdr.bh_caplen;
+			interface->rbuf_offset = BPF_WORDALIGN(
+			    interface->rbuf_offset + hdr.bh_hdrlen +
+			    hdr.bh_caplen);
 			continue;
 		}
