diff --git a/usr.sbin/installboot/efi_installboot.c b/usr.sbin/installboot/efi_installboot.c
index 2118898..92263b2 100644
--- a/usr.sbin/installboot/efi_installboot.c
+++ b/usr.sbin/installboot/efi_installboot.c
@@ -534,6 +534,8 @@ findgptefisys(int devfd, struct disklabel *dl, int *gpartp,
 
 	off = letoh64(gh.gh_part_lba) * dl->d_secsize;
 	ghpartsize = letoh32(gh.gh_part_size);
+	if (ghpartsize == 0)
+		return (-1);
 	ghpartspersec = dl->d_secsize / ghpartsize;
 	ghpartnum = letoh32(gh.gh_part_num);
 	if ((secbuf = malloc(dl->d_secsize)) == NULL)
