diff --git a/usr.bin/indent/indent.c b/usr.bin/indent/indent.c
index 8679c11..7ae97a7 100644
--- a/usr.bin/indent/indent.c
+++ b/usr.bin/indent/indent.c
@@ -1308,7 +1308,7 @@ bakcopy(void)
 	    errc(1, ENAMETOOLONG, "%s.BAK", p);
 
     /* copy in_name to backup file */
-    bakchn = open(bakfile, O_CREAT | O_TRUNC | O_WRONLY, 0600);
+    bakchn = open(bakfile, O_CREAT | O_EXCL | O_NOFOLLOW | O_WRONLY, 0600);
     if (bakchn == -1)
 	err(1, "%s", bakfile);
     while ((n = read(fileno(input), buff, sizeof buff)) > 0)
