diff --git a/usr.sbin/lpd/engine_lpr.c b/usr.sbin/lpd/engine_lpr.c
index 2d019d8..f1d4cd8 100644
--- a/usr.sbin/lpd/engine_lpr.c
+++ b/usr.sbin/lpd/engine_lpr.c
@@ -355,7 +355,7 @@ cmpsockaddr(const struct sockaddr *a, const struct sockaddr *b)
 	case AF_INET6:
 		aa = &(((const struct sockaddr_in6*)a)->sin6_addr);
 		ab = &(((const struct sockaddr_in6*)b)->sin6_addr);
-		l = sizeof(((const struct sockaddr_in*)a)->sin_addr);
+		l = sizeof(((const struct sockaddr_in6*)a)->sin6_addr);
 		return memcmp(aa, ab, l);
 
 	}
