diff --git a/sbin/unwind/resolver.c b/sbin/unwind/resolver.c
index d026bb0..bea5800 100644
--- a/sbin/unwind/resolver.c
+++ b/sbin/unwind/resolver.c
@@ -1027,9 +1027,10 @@ resolve_done(struct uw_resolver *res, void *arg, int rcode,
 		force_acceptbogus = 0;
 
 	timespecsub(&tp, &last_network_change, &elapsed);
-	if (sec != SECURE && elapsed.tv_sec < DOUBT_NXDOMAIN_SEC &&
+	if (sec != SECURE && sec != BOGUS &&
+	    elapsed.tv_sec < DOUBT_NXDOMAIN_SEC &&
 	    !force_acceptbogus && res->type != UW_RES_ASR &&
-	    (result->rcode == LDNS_RCODE_NXDOMAIN || sec == BOGUS)) {
+	    result->rcode == LDNS_RCODE_NXDOMAIN) {
 		/*
 		 * Doubt NXDOMAIN or BOGUS if we just switched networks, we
 		 * might be behind a captive portal.
