diff --git a/usr.bin/patch/inp.c b/usr.bin/patch/inp.c
index 3583814..7dcc94a 100644
--- a/usr.bin/patch/inp.c
+++ b/usr.bin/patch/inp.c
@@ -57,7 +57,7 @@ static size_t	lines_per_buf;	/* how many lines per buffer */
 static size_t	tibuflen;	/* plan b buffer length */
 static size_t	tireclen;	/* length of records in tmp file */
 
-static bool	rev_in_string(const char *);
+static bool	rev_in_string(const char *, size_t);
 static bool	reallocate_lines(size_t *);
 
 /* returns false if insufficient memory */
@@ -247,7 +247,7 @@ plan_a(const char *filename)
 	/* now check for revision, if any */
 
 	if (revision != NULL) {
-		if (i_womp == NULL || !rev_in_string(i_womp)) {
+		if (i_womp == NULL || !rev_in_string(i_womp, (size_t)i_size)) {
 			if (force) {
 				if (verbose)
 					say("Warning: this file doesn't appear "
@@ -301,7 +301,7 @@ plan_b(const char *filename)
 			last_line_missing_eol = true;
 			len++;
 		}
-		if (revision != NULL && !found_revision && rev_in_string(p))
+		if (revision != NULL && !found_revision && rev_in_string(p, len))
 			found_revision = true;
 		if (len > maxlen)
 			maxlen = len;	/* find longest line */
@@ -409,20 +409,22 @@ ifetch(LINENUM line, int whichbuf)
  * True if the string argument contains the revision number we want.
  */
 static bool
-rev_in_string(const char *string)
+rev_in_string(const char *string, size_t len)
 {
-	const char	*s;
-	size_t		patlen;
+	size_t		i, patlen;
 
 	if (revision == NULL)
 		return true;
 	patlen = strlen(revision);
-	if (strnEQ(string, revision, patlen) &&
+	if (patlen < len && strnEQ(string, revision, patlen) &&
 	    isspace((unsigned char)string[patlen]))
 		return true;
-	for (s = string; *s; s++) {
-		if (isspace((unsigned char)*s) && strnEQ(s + 1, revision, patlen) &&
-		    isspace((unsigned char)s[patlen + 1])) {
+	if (len == 0 || patlen >= len - 1)
+		return false;
+	for (i = 0; i < len - patlen - 1 && string[i] != '\0'; i++) {
+		if (isspace((unsigned char)string[i]) &&
+		    strnEQ(string + i + 1, revision, patlen) &&
+		    isspace((unsigned char)string[i + patlen + 1])) {
 			return true;
 		}
 	}
