diff --git a/usr.bin/pkgconf/libpkgconf/pkg.c b/usr.bin/pkgconf/libpkgconf/pkg.c
index a8c1f26..085ef79 100644
--- a/usr.bin/pkgconf/libpkgconf/pkg.c
+++ b/usr.bin/pkgconf/libpkgconf/pkg.c
@@ -1666,7 +1666,8 @@ pkgconf_pkg_walk_conflicts_list(pkgconf_client_t *client,
 				return PKGCONF_PKG_ERRF_PACKAGE_CONFLICT;
 			}
 
-			pkgconf_pkg_unref(client, pkgdep);
+			if (pkgdep != NULL)
+				pkgconf_pkg_unref(client, pkgdep);
 		}
 	}
