diff --git a/usr.sbin/dvmrpd/report.c b/usr.sbin/dvmrpd/report.c
index b961a52..80a5590 100644
--- a/usr.sbin/dvmrpd/report.c
+++ b/usr.sbin/dvmrpd/report.c
@@ -97,6 +97,8 @@ recv_report(struct nbr *nbr, char *buf, u_int16_t len)
 		 */
 
 		/* read four bytes */
+		if (len < sizeof(netmask))
+			return;
 		memcpy(&netmask, buf, sizeof(netmask));
 		/* ditch one byte, since we only need three */
 		netmask = ntohl(netmask) >> 8;
@@ -119,6 +121,8 @@ recv_report(struct nbr *nbr, char *buf, u_int16_t len)
 			 * Read 4 bytes and use the netmask from above to
 			 * determine the netid.
 			 */
+			if (len < sizeof(netid))
+				return;
 			memcpy(&netid, buf, sizeof(netid));
 			netid &= netmask;
 
@@ -126,6 +130,8 @@ recv_report(struct nbr *nbr, char *buf, u_int16_t len)
 			len -= netid_len;
 
 			/* get metric */
+			if (len < sizeof(metric))
+				return;
 			memcpy(&metric, buf, sizeof(metric));
 			buf += sizeof(metric);
 			len -= sizeof(metric);
