diff --git a/usr.sbin/dvmrpd/report.c b/usr.sbin/dvmrpd/report.c
index b961a52..95f8f00 100644
--- a/usr.sbin/dvmrpd/report.c
+++ b/usr.sbin/dvmrpd/report.c
@@ -96,6 +96,8 @@ recv_report(struct nbr *nbr, char *buf, u_int16_t len)
 		 * The most significant part of the mask is always 255.
 		 */
 
+		if (len < sizeof(netmask))
+			return;
 		/* read four bytes */
 		memcpy(&netmask, buf, sizeof(netmask));
 		/* ditch one byte, since we only need three */
@@ -116,16 +118,19 @@ recv_report(struct nbr *nbr, char *buf, u_int16_t len)
 			 *
 			 * The length of the netid is depending on the above
 			 * netmask.
-			 * Read 4 bytes and use the netmask from above to
-			 * determine the netid.
 			 */
-			memcpy(&netid, buf, sizeof(netid));
+			if (len < netid_len)
+				return;
+			netid = 0;
+			memcpy(&netid, buf, netid_len);
 			netid &= netmask;
 
 			buf += netid_len;
 			len -= netid_len;
 
 			/* get metric */
+			if (len < sizeof(metric))
+				return;
 			memcpy(&metric, buf, sizeof(metric));
 			buf += sizeof(metric);
 			len -= sizeof(metric);
