diff --git a/usr.sbin/ocspcheck/http.c b/usr.sbin/ocspcheck/http.c
index 46b01f1..914d315 100644
--- a/usr.sbin/ocspcheck/http.c
+++ b/usr.sbin/ocspcheck/http.c
@@ -34,6 +34,8 @@
 
 #include "http.h"
 
+#define HTTP_MAX_BODY_SIZE	(20 * 1024)
+
 /*
  * A buffer for transferring HTTP/S data.
  */
@@ -416,6 +418,11 @@ http_body_read(const struct http *http, struct httpxfer *trans, size_t *sz)
 		else if (ssz == 0)
 			break;
 
+		if (trans->bbufsz > HTTP_MAX_BODY_SIZE ||
+		    (size_t)ssz > HTTP_MAX_BODY_SIZE - trans->bbufsz) {
+			warnx("%s: body too large", http->src.ip);
+			return NULL;
+		}
 		pp = recallocarray(trans->bbuf,
 		    trans->bbufsz, trans->bbufsz + ssz, 1);
 		if (pp == NULL) {
