diff --git a/usr.sbin/route6d/route6d.c b/usr.sbin/route6d/route6d.c
index 5a49e8c..22163cc 100644
--- a/usr.sbin/route6d/route6d.c
+++ b/usr.sbin/route6d/route6d.c
@@ -1331,6 +1331,9 @@ riprequest(struct ifc *ifcp, struct netinfo6 *np, int nn,
 	if (!(nn == 1 && IN6_IS_ADDR_UNSPECIFIED(&np->rip6_dest) &&
 	      np->rip6_plen == 0 && np->rip6_metric == HOPCNT_INFINITY6)) {
 		/* Specific response, don't split-horizon */
+		if (RIPSIZE(nn) > RIP6_MAXMTU)
+			nn = (RIP6_MAXMTU - sizeof(struct rip6) +
+			    sizeof(struct netinfo6)) / sizeof(struct netinfo6);
 		log_debug("\tRIP Request");
 		for (i = 0; i < nn; i++, np++) {
 			rrt = rtsearch(np, NULL);
