diff --git a/usr.sbin/unwindctl/unwindctl.c b/usr.sbin/unwindctl/unwindctl.c
index 2735c33..b21f11b 100644
--- a/usr.sbin/unwindctl/unwindctl.c
+++ b/usr.sbin/unwindctl/unwindctl.c
@@ -280,10 +280,16 @@ show_autoconf_msg(struct imsg *imsg)
 	struct ctl_forwarder_info	*cfi;
 	char				 ifnamebuf[IFNAMSIZ];
 	char				*if_name;
+	size_t				 ip_len;
 
 	switch (imsg->hdr.type) {
 	case IMSG_CTL_AUTOCONF_RESOLVER_INFO:
+		if (IMSG_DATA_SIZE(*imsg) != sizeof(*cfi))
+			break;
 		cfi = imsg->data;
+		ip_len = strnlen(cfi->ip, sizeof(cfi->ip));
+		if (ip_len == sizeof(cfi->ip))
+			break;
 		if (!autoconf_forwarders++)
 			printf("autoconfiguration forwarders:\n");
 		if (cfi->if_index != last_if_index || cfi->src != last_src) {
@@ -300,7 +306,7 @@ show_autoconf_msg(struct imsg *imsg)
 			last_src = cfi->src;
 		}
 
-		if (line_len + 1 + strlen(cfi->ip) > sizeof(fwd_line)) {
+		if (line_len + 1 + ip_len > sizeof(fwd_line)) {
 			printf("%s\n", fwd_line);
 			snprintf(fwd_line, sizeof(fwd_line), "%*s", label_len,
 			    " ");
