diff --git a/sbin/dhcp6leased/engine.c b/sbin/dhcp6leased/engine.c
index b818b32..1f2202c 100644
--- a/sbin/dhcp6leased/engine.c
+++ b/sbin/dhcp6leased/engine.c
@@ -1017,6 +1017,12 @@ parse_ia_pd_options(uint8_t *p, size_t len, struct prefix *prefix)
 				break;
 			}
 
+			if (iaprefix.prefix_len > 128) {
+				log_warnx("%s: invalid prefix length, ignoring IA_PD",
+				    __func__);
+				break;
+			}
+
 			prefix->prefix = iaprefix.prefix;
 			prefix->prefix_len = iaprefix.prefix_len;
 			prefix->vltime = ntohl(iaprefix.vltime);
