diff --git a/usr.bin/rsync/downloader.c b/usr.bin/rsync/downloader.c
index cab6eb2..5401132 100644
--- a/usr.bin/rsync/downloader.c
+++ b/usr.bin/rsync/downloader.c
@@ -486,6 +486,10 @@ again:
 
 		return 1;
 	} else if (rawtok < 0) {
+		if (rawtok == INT32_MIN) {
+			ERRX("%s: invalid token", p->fname);
+			goto out;
+		}
 		tok = -rawtok - 1;
 		if (tok >= p->blk.blksz) {
 			ERRX("%s: token not in block set: %zu (have %zu blocks)",
