diff --git a/usr.bin/showmount/showmount.c b/usr.bin/showmount/showmount.c
index c264f68..3ae93dc 100644
--- a/usr.bin/showmount/showmount.c
+++ b/usr.bin/showmount/showmount.c
@@ -55,6 +55,9 @@
 #define	DODUMP		0x1
 #define	DOEXPORTS	0x2
 
+#define	MAXEXPORTS	1024
+#define	MAXGROUPS	1024
+
 struct mountlist {
 	struct mountlist *ml_left;
 	struct mountlist *ml_right;
@@ -310,12 +313,15 @@ xdr_exports(XDR *xdrsp, struct exportslist **exp)
 	struct exportslist *ep;
 	struct grouplist *gp;
 	int bool, grpbool;
+	int exports_count = 0, groups_count = 0;
 	char *strp;
 
 	*exp = NULL;
 	if (!xdr_bool(xdrsp, &bool))
 		return (0);
 	while (bool) {
+		if (exports_count++ >= MAXEXPORTS)
+			return (0);
 		ep = malloc(sizeof(struct exportslist));
 		if (ep == NULL)
 			return (0);
@@ -326,6 +332,8 @@ xdr_exports(XDR *xdrsp, struct exportslist **exp)
 		if (!xdr_bool(xdrsp, &grpbool))
 			return (0);
 		while (grpbool) {
+			if (groups_count++ >= MAXGROUPS)
+				return (0);
 			gp = malloc(sizeof(struct grouplist));
 			if (gp == NULL)
 				return (0);
