diff --git a/sbin/slaacd/engine.c b/sbin/slaacd/engine.c
index f2320c0..f8946bd 100644
--- a/sbin/slaacd/engine.c
+++ b/sbin/slaacd/engine.c
@@ -1493,10 +1493,16 @@ parse_ra(struct slaacd_iface *iface, struct imsg_ra *ra)
 				goto err;
 			}
 
+			prf = (struct nd_opt_prefix_info*) nd_opt_hdr;
+			if (prf->nd_opt_pi_prefix_len > 128) {
+				log_warnx("invalid ND_OPT_PREFIX_INFORMATION: "
+				    "prefix length > 128");
+				goto err;
+			}
+
 			if ((prefix = calloc(1, sizeof(*prefix))) == NULL)
 				fatal("calloc");
 
-			prf = (struct nd_opt_prefix_info*) nd_opt_hdr;
 			prefix->prefix = prf->nd_opt_pi_prefix;
 			prefix->prefix_len = prf->nd_opt_pi_prefix_len;
 			prefix->onlink = prf->nd_opt_pi_flags_reserved &
