diff --git a/usr.bin/gprof/elf.c b/usr.bin/gprof/elf.c
index 3233608..1c594e8 100644
--- a/usr.bin/gprof/elf.c
+++ b/usr.bin/gprof/elf.c
@@ -39,7 +39,7 @@
 
 #include "gprof.h"
 
-static bool wantsym(const Elf_Sym *, const char *);
+static bool wantsym(const Elf_Sym *, const char *, size_t);
 
 /* Things which get -E excluded by default. */
 static char	*excludes[] = { ".mcount", "_mcleanup", NULL };
@@ -102,7 +102,7 @@ getnfile(const char *filename, char ***defaultEs)
     /* Count the symbols that we're interested in. */
     nname = 0;
     for (i = 1;  i < symtabct;  i++)
-	if (wantsym(&symtab[i], strtab))
+	if (wantsym(&symtab[i], strtab, sh_strtab->sh_size))
 	    nname++;
 
 #ifdef DEBUG
@@ -121,7 +121,7 @@ getnfile(const char *filename, char ***defaultEs)
     for (i = 1;  i < symtabct;  i++) {
 	const Elf_Sym *sym = &symtab[i];
 
-	if (wantsym(sym, strtab)) {
+	if (wantsym(sym, strtab, sh_strtab->sh_size)) {
 	    npe->value = sym->st_value;
 	    npe->name = strtab + sym->st_name;
 #ifdef DEBUG
@@ -139,7 +139,7 @@ getnfile(const char *filename, char ***defaultEs)
 }
 
 static bool
-wantsym(const Elf_Sym *sym, const char *strtab)
+wantsym(const Elf_Sym *sym, const char *strtab, size_t strtabsz)
 {
     int type;
     int bind;
@@ -147,7 +147,8 @@ wantsym(const Elf_Sym *sym, const char *strtab)
     type = ELF_ST_TYPE(sym->st_info);
     bind = ELF_ST_BIND(sym->st_info);
 
-    if (type != STT_FUNC || (aflag && bind == STB_LOCAL))
+    if (type != STT_FUNC || (aflag && bind == STB_LOCAL) ||
+      sym->st_name >= strtabsz)
 #if 0
  ||
       (uflag && strchr(strtab + sym->st_name, '.') != NULL))
