diff --git a/usr.bin/talk/ctl_transact.c b/usr.bin/talk/ctl_transact.c
index 8852b06..d8789db 100644
--- a/usr.bin/talk/ctl_transact.c
+++ b/usr.bin/talk/ctl_transact.c
@@ -50,6 +50,8 @@ void
 ctl_transact(struct in_addr target, CTL_MSG msg, int type, CTL_RESPONSE *rp)
 {
 	struct pollfd pfd[1];
+	struct sockaddr_in from;
+	socklen_t fromlen;
 	int nready, cc;
 
 	msg.type = type;
@@ -86,7 +88,9 @@ ctl_transact(struct in_addr target, CTL_MSG msg, int type, CTL_RESPONSE *rp)
 		 * request/acknowledgements being sent)
 		 */
 		do {
-			cc = recv(ctl_sockt, (char *)rp, sizeof (*rp), 0);
+			fromlen = sizeof (from);
+			cc = recvfrom(ctl_sockt, (char *)rp, sizeof (*rp), 0,
+			    (struct sockaddr *)&from, &fromlen);
 			if (cc < 0) {
 				if (errno == EINTR)
 					continue;
@@ -94,7 +98,11 @@ ctl_transact(struct in_addr target, CTL_MSG msg, int type, CTL_RESPONSE *rp)
 			}
 			nready = poll(pfd, 1, 0);
 		} while (nready > 0 && (rp->vers != TALK_VERSION ||
-		    rp->type != type));
-	} while (rp->vers != TALK_VERSION || rp->type != type);
+		    rp->type != type ||
+		    from.sin_addr.s_addr != daemon_addr.sin_addr.s_addr ||
+		    from.sin_port != daemon_addr.sin_port));
+	} while (rp->vers != TALK_VERSION || rp->type != type ||
+	    from.sin_addr.s_addr != daemon_addr.sin_addr.s_addr ||
+	    from.sin_port != daemon_addr.sin_port);
 	rp->id_num = ntohl(rp->id_num);
 }
