diff --git a/usr.bin/tftp/tftp.c b/usr.bin/tftp/tftp.c
index 265f51e..56f6a0e 100644
--- a/usr.bin/tftp/tftp.c
+++ b/usr.bin/tftp/tftp.c
@@ -583,7 +583,7 @@ static void
 oack(struct tftphdr *tp, int size, int trace)
 {
 	int	 i, len, off;
-	char	*opt, *val;
+	char	*end, *opt, *val;
 
 	u_short op = ntohs(tp->th_opcode);
 
@@ -611,12 +611,14 @@ oack(struct tftphdr *tp, int size, int trace)
 		}
 		/* got option and value */
 		val++;
+		if ((end = memchr(val, '\0', size - i - 1)) == NULL)
+			break;
+		len = end - val + 1;
 		if (trace)
 			printf("%s=%s", opt, val);
 		else
 			if (oack_set(opt, val) == -1)
 				break;
-		len = strlen(val) + 1;
 		val += len;
 		opt = val;
 		i += len;
