diff --git a/usr.sbin/dhcpd/dhcp.c b/usr.sbin/dhcpd/dhcp.c
index 2380980..090bd0e 100644
--- a/usr.sbin/dhcpd/dhcp.c
+++ b/usr.sbin/dhcpd/dhcp.c
@@ -510,8 +510,17 @@ dhcpdecline(struct packet *packet)
 		return;
 	}
 
-	/* If we found a lease, mark it as unusable and complain. */
-	if (lease)
+	/* If we found a lease owned by this client, mark it as unusable. */
+	if (lease && ((lease->uid_len &&
+	    packet->options[DHO_DHCP_CLIENT_IDENTIFIER].data &&
+	    lease->uid_len ==
+	    packet->options[DHO_DHCP_CLIENT_IDENTIFIER].len &&
+	    !memcmp(packet->options[DHO_DHCP_CLIENT_IDENTIFIER].data,
+	    lease->uid, lease->uid_len)) || (!lease->uid_len &&
+	    lease->hardware_addr.htype == packet->raw->htype &&
+	    lease->hardware_addr.hlen == packet->raw->hlen &&
+	    !memcmp(lease->hardware_addr.haddr, packet->raw->chaddr,
+	    lease->hardware_addr.hlen))))
 		abandon_lease(lease, "declined.");
 }
