diff --git a/usr.sbin/mrinfo/mrinfo.c b/usr.sbin/mrinfo/mrinfo.c
index 41e3a7e..87f990e 100644
--- a/usr.sbin/mrinfo/mrinfo.c
+++ b/usr.sbin/mrinfo/mrinfo.c
@@ -191,6 +191,8 @@ accept_neighbors(u_int32_t src, u_int32_t dst, u_char *p, int datalen,
 		u_int32_t laddr;
 		int ncount;
 
+		if (ep - p < 7)
+			return;
 		GET_ADDR(laddr);
 		laddr = htonl(laddr);
 		metric = *p++;
@@ -199,6 +201,8 @@ accept_neighbors(u_int32_t src, u_int32_t dst, u_char *p, int datalen,
 		while (--ncount >= 0) {
 			u_int32_t neighbor;
 
+			if (ep - p < 4)
+				return;
 			GET_ADDR(neighbor);
 			neighbor = htonl(neighbor);
 			printf("  %s -> ", inet_fmt(laddr, s1));
