diff --git a/usr.sbin/mrouted/route.c b/usr.sbin/mrouted/route.c
index 9271eb2..cb4fe38 100644
--- a/usr.sbin/mrouted/route.c
+++ b/usr.sbin/mrouted/route.c
@@ -799,7 +799,7 @@ accept_report(u_int32_t src, u_int32_t dst, char *p, int datalen,
     /*
      * If the last entry is default, change mask from 0xff000000 to 0
      */
-    if (rt[nrt-1].origin == 0)
+    if (nrt > 0 && rt[nrt-1].origin == 0)
 	rt[nrt-1].mask = 0;
 
     logit(LOG_DEBUG, 0, "Updating %d routes from %s to %s", nrt,
