diff --git a/usr.sbin/npppd/pppoe/pppoe_session.c b/usr.sbin/npppd/pppoe/pppoe_session.c
index 104c3a7..cd24afa 100644
--- a/usr.sbin/npppd/pppoe/pppoe_session.c
+++ b/usr.sbin/npppd/pppoe/pppoe_session.c
@@ -377,6 +377,9 @@ pppoe_session_recv_PADR(pppoe_session *_this, slist *tag_list)
 	}
 
 	if (ac_cookie) {
+		if (ac_cookie->length != sizeof(uint32_t))
+			goto fail;
+
 		/* avoid a session which has already has cookie. */
 		if (hash_lookup(pppoed0->acookie_hash,
 		    (void *)ac_cookie->value) != NULL)
