diff --git a/usr.sbin/ypserv/ypserv/acl.c b/usr.sbin/ypserv/ypserv/acl.c
index 2cba27a..dd39584 100644
--- a/usr.sbin/ypserv/ypserv/acl.c
+++ b/usr.sbin/ypserv/ypserv/acl.c
@@ -327,8 +327,8 @@ acl_init(char *file)
 
 			if (state == ACLS_ALLOW_NET_MASK ||
 			    state == ACLS_DENY_NET_MASK) {
-				if (*k >= '0' && *k <= '9') {
-					(void)inet_aton(k, &mask);
+				if (*k >= '0' && *k <= '9' &&
+				    inet_aton(k, &mask) != 0) {
 					state = state + ACLD_NET_EOL;
 				} else
 					state = ACLE_NONET;
