diff --git a/usr.sbin/ypserv/ypserv/acl.c b/usr.sbin/ypserv/ypserv/acl.c
index 2cba27a..e89c5a9 100644
--- a/usr.sbin/ypserv/ypserv/acl.c
+++ b/usr.sbin/ypserv/ypserv/acl.c
@@ -450,8 +450,10 @@ acl_securenet(char *file)
 
 		if (state == ACLS_ALLOW_NET_MASK) {
 			if (*k >= '0' && *k <= '9') {
-				(void)inet_aton(k, &mask);
-				state = ACLS_ALLOW_NET;
+				if (inet_aton(k, &mask) == 1)
+					state = ACLS_ALLOW_NET;
+				else
+					state = ACLE_NONET;
 			} else
 				state = ACLE_NONET;
 
@@ -472,8 +474,10 @@ acl_securenet(char *file)
 
 		if (state == ACLS_ALLOW_NET) {
 			if (*k >= '0' && *k <= '9') {
-				(void)inet_aton(k, &addr);
-				state = ACLS_ALLOW_NET_EOL;
+				if (inet_aton(k, &addr) == 1)
+					state = ACLS_ALLOW_NET_EOL;
+				else
+					state = ACLE_NONET;
 			} else
 				state = ACLE_NONET;
 		}
