diff --git a/usr.bin/cdio/cddb.c b/usr.bin/cdio/cddb.c
index babce4a..7c54832 100644
--- a/usr.bin/cdio/cddb.c
+++ b/usr.bin/cdio/cddb.c
@@ -365,6 +365,8 @@ cddb(const char *host_port, int n, struct cd_toc_entry *e, char *arg)
 		k = strtonum(line, 0, n - 1, &errstr);
 		if (errstr != NULL)
 			continue;
+		if (result[k] != NULL)
+			continue;
 		safe_copy(&result[k], end);
 	}
 	fprintf(cout, "QUIT\r\n");
