diff --git a/usr.sbin/rpki-client/ip.c b/usr.sbin/rpki-client/ip.c
index d04f626..90b9fbe 100644
--- a/usr.sbin/rpki-client/ip.c
+++ b/usr.sbin/rpki-client/ip.c
@@ -150,8 +150,8 @@ ip_addr_check_overlap(const struct cert_ip *ip, const char *fn,
 	for (i = 0; i < num_ips; i++) {
 		if (ips[i].afi != ip->afi)
 			continue;
-		if (memcmp(ips[i].max, ip->min, sz) <= 0 ||
-		    memcmp(ips[i].min, ip->max, sz) >= 0)
+		if (memcmp(ips[i].max, ip->min, sz) < 0 ||
+		    memcmp(ips[i].min, ip->max, sz) > 0)
 			continue;
 		if (!quiet) {
 			warnx("%s: RFC 3779 section 2.2.3.5: "
