diff --git a/usr.sbin/vmd/vioblk.c b/usr.sbin/vmd/vioblk.c
index e8c193c..c039aff 100644
--- a/usr.sbin/vmd/vioblk.c
+++ b/usr.sbin/vmd/vioblk.c
@@ -267,7 +267,7 @@ vioblk_notifyq(struct virtio_dev *dev, uint16_t vq_idx)
 	struct vioblk_dev *vioblk = &dev->vioblk;
 
 	/* Invalid queue? */
-	if (vq_idx > dev->num_queues)
+	if (vq_idx >= dev->num_queues)
 		return (0);
 
 	vq_info = &dev->vq[vq_idx];
