diff --git a/usr.sbin/vmd/vioscsi.c b/usr.sbin/vmd/vioscsi.c
index 6f3aa17..24bd3fd 100644
--- a/usr.sbin/vmd/vioscsi.c
+++ b/usr.sbin/vmd/vioscsi.c
@@ -2189,6 +2189,10 @@ vioscsi_notifyq(struct virtio_dev *dev, uint16_t vq_idx)
 	struct virtio_vq_acct acct;
 	struct virtio_vq_info *vq_info;
 
+	if (vq_idx >= dev->num_queues) {
+		log_warnx("%s: invalid virtqueue index %u", __func__, vq_idx);
+		return (0);
+	}
 	vq_info = &dev->vq[vq_idx];
 	vr = vq_info->q_hva;
 	if (vr == NULL)
