diff --git a/lib/libossaudio/ossaudio.c b/lib/libossaudio/ossaudio.c
index bad73ac..da566b3 100644
--- a/lib/libossaudio/ossaudio.c
+++ b/lib/libossaudio/ossaudio.c
@@ -113,7 +113,7 @@ mixer_ondesc(void *unused, struct sioctl_desc *d, int val)
 	 * we support only numeric "level" controls, first 2 channels
 	 */
 	if (d->type != SIOCTL_NUM || d->node0.unit >= 2 ||
-	    strcmp(d->func, "level") != 0)
+	    d->maxval == 0 || strcmp(d->func, "level") != 0)
 		return;
 
 	/*
