diff --git a/lib/libusbhid/usage.c b/lib/libusbhid/usage.c
index 5885994..4674f1c 100644
--- a/lib/libusbhid/usage.c
+++ b/lib/libusbhid/usage.c
@@ -234,8 +234,8 @@ hid_usage_in_page(unsigned int u)
 	for (j = 0; j < pages[k].pagesize; j++) {
 		us = pages[k].page_contents[j].usage;
 		if (us == -1) {
-			snprintf(b, sizeof b,
-			    pages[k].page_contents[j].name, i);
+			snprintf(b, sizeof b, "%s",
+			    pages[k].page_contents[j].name);
 			return b;
 		}
 		if (us == i)
