diff --git a/usr.bin/gencat/gencat.c b/usr.bin/gencat/gencat.c
index bb66db1..253a225 100644
--- a/usr.bin/gencat/gencat.c
+++ b/usr.bin/gencat/gencat.c
@@ -686,8 +686,11 @@ MCDelSet(int setId)
 
 		msg = LIST_FIRST(&set->msghead);
 		while (msg) {
+			struct _msgT *nextmsg = LIST_NEXT(msg, entries);
+
 			free(msg->str);
 			LIST_REMOVE(msg, entries);
+			msg = nextmsg;
 		}
 
 		LIST_REMOVE(set, entries);
