diff --git a/usr.sbin/ldomctl/mdesc.c b/usr.sbin/ldomctl/mdesc.c
index 3cae1f7..50b90a5 100644
--- a/usr.sbin/ldomctl/mdesc.c
+++ b/usr.sbin/ldomctl/mdesc.c
@@ -430,6 +430,7 @@ md_ingest(void *buf, size_t size)
 	uint8_t *node_blk;
 	uint8_t *name_blk;
 	uint8_t *data_blk;
+	uint32_t offset;
 	uint64_t index;
 
 	if (size < sizeof(struct md_header))
@@ -470,7 +471,11 @@ md_ingest(void *buf, size_t size)
 			if (node == NULL)
 				errx(1, "Corrupt MD");
 			str = name_blk + betoh32(mde->name_offset);
-			data = data_blk + betoh32(mde->d.y.data_offset);
+			offset = betoh32(mde->d.y.data_offset);
+			if (offset >= data_blk_size ||
+			    memchr(data_blk + offset, '\0', data_blk_size - offset) == NULL)
+				errx(1, "Corrupt MD");
+			data = data_blk + offset;
 			md_add_prop_str(md, node, str, data);
 			break;
 		case MD_PROP_DATA:
