diff --git a/usr.sbin/ldomctl/mdstore.c b/usr.sbin/ldomctl/mdstore.c
index 3592fca..9e009cb 100644
--- a/usr.sbin/ldomctl/mdstore.c
+++ b/usr.sbin/ldomctl/mdstore.c
@@ -210,6 +210,7 @@ mdstore_rx_data(struct ldc_conn *lc, uint64_t svc_handle, void *data,
 {
 	struct mdstore_list_resp *mr = data;
 	struct mdstore_set *set;
+	char *end;
 	int idx;
 
 	if (mr->result != MDST_SUCCESS) {
@@ -228,13 +229,17 @@ mdstore_rx_data(struct ldc_conn *lc, uint64_t svc_handle, void *data,
 
 	switch (mdstore_command) {
 	case MDSET_LIST_REQUEST:
-		for (idx = 0, len = 0; len < mr->payload_len - 24; idx++) {
+		for (idx = 0, len = 0; len + 24 < mr->payload_len; idx++) {
+			end = memchr(&mr->sets[len], '\0',
+			    mr->payload_len - 24 - len);
+			if (end == NULL)
+				errx(1, "Unterminated set name");
 			set = xmalloc(sizeof(*set));
 			set->name = xstrdup(&mr->sets[len]);
 			set->booted_set = (idx == mr->booted_set);
 			set->boot_set = (idx == mr->boot_set);
 			TAILQ_INSERT_TAIL(&mdstore_sets, set, link);
-			len += strlen(&mr->sets[len]) + 1;
+			len += end - &mr->sets[len] + 1;
 			if (mdstore_major >= 2)
 				len += sizeof(uint64_t); /* skip timestamp */
 			if (mdstore_major >= 3)
