diff --git a/usr.sbin/ospf6d/rde_lsdb.c b/usr.sbin/ospf6d/rde_lsdb.c
index 33b6b69..3572765 100644
--- a/usr.sbin/ospf6d/rde_lsdb.c
+++ b/usr.sbin/ospf6d/rde_lsdb.c
@@ -249,7 +249,7 @@ lsa_check(struct rde_nbr *nbr, struct lsa *lsa, u_int16_t len)
 		}
 		if (lsa_get_prefix(((char *)lsa) + sizeof(lsa->hdr) +
 		    sizeof(lsa->data.pref_sum),
-		    len - sizeof(lsa->hdr) + sizeof(lsa->data.pref_sum),
+		    len - sizeof(lsa->hdr) - sizeof(lsa->data.pref_sum),
 		    NULL) == -1) {
 			log_warnx("lsa_check: "
 			    "invalid LSA prefix summary packet");
