diff --git a/usr.sbin/vmd/i8253.c b/usr.sbin/vmd/i8253.c
index 00b0945..e9a29d3 100644
--- a/usr.sbin/vmd/i8253.c
+++ b/usr.sbin/vmd/i8253.c
@@ -348,7 +348,7 @@ i8253_reset(uint8_t chn)
 
 	i8253_channel[chn].in_use = 1;
 	i8253_channel[chn].state = 0;
-	tv.tv_usec = (i8253_channel[chn].start * NS_PER_TICK) / 1000;
+	tv.tv_usec = (i8253_channel[chn].start * NS_PER_TICK + 999) / 1000;
 	clock_gettime(CLOCK_MONOTONIC, &i8253_channel[chn].ts);
 	evtimer_add(&i8253_channel[chn].timer, &tv);
 }
@@ -374,7 +374,7 @@ i8253_fire(int fd, short type, void *arg)
 
 	if (ctr->mode != TIMER_INTTC) {
 		timerclear(&tv);
-		tv.tv_usec = (ctr->start * NS_PER_TICK) / 1000;
+		tv.tv_usec = (ctr->start * NS_PER_TICK + 999) / 1000;
 		evtimer_add(&ctr->timer, &tv);
 	} else
 		ctr->state = 1;
