diff --git a/usr.bin/pkgconf/libpkgconf/tuple.c b/usr.bin/pkgconf/libpkgconf/tuple.c
index 83f6a47..035fa01 100644
--- a/usr.bin/pkgconf/libpkgconf/tuple.c
+++ b/usr.bin/pkgconf/libpkgconf/tuple.c
@@ -322,7 +322,7 @@ pkgconf_tuple_parse(const pkgconf_client_t *client, pkgconf_list_t *vars, const
 			bptr += pkgconf_strlcpy(buf, client->sysroot_dir, sizeof buf);
 	}
 
-	for (ptr = value; *ptr != '\0' && bptr - buf < PKGCONF_BUFSIZE; ptr++)
+	for (ptr = value; *ptr != '\0' && bptr - buf < PKGCONF_BUFSIZE - 1; ptr++)
 	{
 		if (*ptr != '$' || (*ptr == '$' && *(ptr + 1) != '{'))
 			*bptr++ = *ptr;
