diff --git a/usr.bin/snmp/smi.c b/usr.bin/snmp/smi.c
index 5a29914..5028062 100644
--- a/usr.bin/snmp/smi.c
+++ b/usr.bin/snmp/smi.c
@@ -374,6 +374,8 @@ smi_print_element(struct ber_oid *oid, struct ber_element *root, int print_hint,
 			goto fail;
 		if (root->be_class == BER_CLASS_APPLICATION &&
 		    root->be_type == SNMP_T_IPADDR) {
+			if (root->be_len != sizeof(struct in_addr))
+				goto fail;
 			if (asprintf(&str, "%s%s",
 			    print_hint ? "IpAddress: " : "",
 			    inet_ntoa(*(struct in_addr *)buf)) == -1)
