diff --git a/usr.sbin/ypserv/common/yplib_host.c b/usr.sbin/ypserv/common/yplib_host.c
index 4a42107..bad28ad 100644
--- a/usr.sbin/ypserv/common/yplib_host.c
+++ b/usr.sbin/ypserv/common/yplib_host.c
@@ -219,13 +219,24 @@ yp_next_host(CLIENT *client, char *indomain, char *inmap, char *inkey,
 	if ( !(r = ypprot_err(yprkv.stat)) ) {
 		*outkeylen = yprkv.key.keydat_len;
 		*outkey = malloc(*outkeylen+1);
+		if (*outkey == NULL) {
+			r = YP_YPERR;
+			goto out;
+		}
 		memcpy(*outkey, yprkv.key.keydat_val, *outkeylen);
 		(*outkey)[*outkeylen] = '\0';
 		*outvallen = yprkv.val.valdat_len;
 		*outval = malloc(*outvallen+1);
+		if (*outval == NULL) {
+			free(*outkey);
+			*outkey = NULL;
+			r = YP_YPERR;
+			goto out;
+		}
 		memcpy(*outval, yprkv.val.valdat_val, *outvallen);
 		(*outval)[*outvallen] = '\0';
 	}
+out:
 	xdr_free(xdr_ypresp_key_val, (char *)&yprkv);
 	return r;
 }
