diff --git a/usr.sbin/ypserv/yppush/yppush_svc.c b/usr.sbin/ypserv/yppush/yppush_svc.c
index b0ed5cb..a212d56 100644
--- a/usr.sbin/ypserv/yppush/yppush_svc.c
+++ b/usr.sbin/ypserv/yppush/yppush_svc.c
@@ -96,7 +96,6 @@ yppush_xfrrespprog_1(struct svc_req *rqstp, SVCXPRT *transp)
 	if (!svc_getargs(transp, xdr_argument, (caddr_t)&argument)) {
 		svcerr_decode(transp);
 		_rpcsvcdirty = 0;
-		exit(1);
 		return;
 	}
 	result = (*local)(&argument, rqstp);
