diff --git a/usr.sbin/bgpd/rde_filter.c b/usr.sbin/bgpd/rde_filter.c
index 6520629..97052ff 100644
--- a/usr.sbin/bgpd/rde_filter.c
+++ b/usr.sbin/bgpd/rde_filter.c
@@ -1172,9 +1172,6 @@ rde_filter_out(struct rde_filter *rf, struct rde_peer *peer,
 		 */
 		return (ACTION_DENY);
 
-	if (prefix->aid == AID_FLOWSPECv4 || prefix->aid == AID_FLOWSPECv6)
-		return (ACTION_ALLOW);
-
 	for (i = 0; i < rf->len; i++) {
 		f = &rf->rules[i];
 		if (rde_filter_match(&f->match, peer, from, state,
