diff --git a/usr.sbin/lpd/lp_banner.c b/usr.sbin/lpd/lp_banner.c
index 6d363df..a5f93ab 100644
--- a/usr.sbin/lpd/lp_banner.c
+++ b/usr.sbin/lpd/lp_banner.c
@@ -1130,6 +1130,8 @@ lp_banner(int scfd, char *scsp, int pw)
 		sp = scsp;
 		for (nchrs = 0; ; ) {
 			d = dropit(c = TRC(cc = *sp++));
+			if (c >= sizeof(scnkey) / sizeof(scnkey[0]))
+				c = TRC(cc = ' ');
 			if ((!d && scnhgt > HEIGHT) || (scnhgt <= DROP && d))
 				for (j = WIDTH; --j;)
 					*strp++ = BACKGND;
