diff --git a/usr.sbin/traceroute/worker.c b/usr.sbin/traceroute/worker.c
index 0865f4d..e6328c7 100644
--- a/usr.sbin/traceroute/worker.c
+++ b/usr.sbin/traceroute/worker.c
@@ -521,15 +521,17 @@ packet_ok4(struct tr_conf *conf, struct msghdr *mhdr, int cc, int *seq)
 		struct icmp *icmpp;
 
 		hip = &icp->icmp_ip;
+		if (type == ICMP_ECHOREPLY && conf->proto == IPPROTO_ICMP &&
+		    icp->icmp_id == htons(conf->ident)) {
+			*seq = ntohs(icp->icmp_seq);
+			return (-2); /* we got there */
+		}
+		if (cc < ICMP_MINLEN + sizeof(*hip))
+			return (0);
 		hlen = hip->ip_hl << 2;
 
 		switch (conf->proto) {
 		case IPPROTO_ICMP:
-			if (type == ICMP_ECHOREPLY &&
-			    icp->icmp_id == htons(conf->ident)) {
-				*seq = ntohs(icp->icmp_seq);
-				return (-2); /* we got there */
-			}
 			icmpp = (struct icmp *)((u_char *)hip + hlen);
 			if (hlen + 8 <= cc && hip->ip_p == IPPROTO_ICMP &&
 			    icmpp->icmp_id == htons(conf->ident)) {
