diff --git a/crates/wasi/src/p1.rs b/crates/wasi/src/p1.rs
index 95c688e2db..32da0f0489 100644
--- a/crates/wasi/src/p1.rs
+++ b/crates/wasi/src/p1.rs
@@ -2541,7 +2541,7 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiP1Ctx {
                                             } else {
                                                 types::Eventrwflags::empty()
                                             },
-                                            nbytes: 1,
+                                            nbytes,
                                         },
                                     }
                                 }
