diff --git a/crates/wasi/src/p2/pipe.rs b/crates/wasi/src/p2/pipe.rs
index 05dbff1a56..c0a1c44dc5 100644
--- a/crates/wasi/src/p2/pipe.rs
+++ b/crates/wasi/src/p2/pipe.rs
@@ -199,7 +199,11 @@ impl AsyncReadStream {
                 Poll::Ready(())
             }
             Poll::Ready(None) => {
-                panic!("no more sender for an open AsyncReadStream - should be impossible")
+                self.closed = true;
+                self.buffer = Some(Err(StreamError::Trap(format_err!(
+                    "AsyncReadStream sender died - should be impossible"
+                ))));
+                Poll::Ready(())
             }
             Poll::Pending => Poll::Pending,
         }
