diff --git a/src/crypto/internal/fips140test/acvp_test.go b/src/crypto/internal/fips140test/acvp_test.go
index 6a0b46af2b..a89ccbf3a8 100644
--- a/src/crypto/internal/fips140test/acvp_test.go
+++ b/src/crypto/internal/fips140test/acvp_test.go
@@ -1486,6 +1486,9 @@ func cmdAesGcmOpen(randNonce bool) command {
 
 			if randNonce {
 				// for randNonce tests acvptool appends the nonce to the end of the ciphertext.
+				if len(ciphertext) < 12 {
+					return [][]byte{{0}, nil}, nil
+				}
 				nonce = ciphertext[len(ciphertext)-12:]
 				ciphertext = ciphertext[:len(ciphertext)-12]
 			}
