diff --git a/src/crypto/internal/fips140test/acvp_test.go b/src/crypto/internal/fips140test/acvp_test.go
index 6a0b46af2b..5acc40b0b5 100644
--- a/src/crypto/internal/fips140test/acvp_test.go
+++ b/src/crypto/internal/fips140test/acvp_test.go
@@ -1948,6 +1948,9 @@ func cmdRsaSigVerAft(hashFunc func() hash.Hash, hashName string, pss bool) comma
 			msg := args[2]
 			sig := args[3]
 
+			if len(eBytes) > 4 {
+				return [][]byte{{0}}, nil
+			}
 			paddedE := make([]byte, 4)
 			copy(paddedE[4-len(eBytes):], eBytes)
 			e := int(binary.BigEndian.Uint32(paddedE))
