diff --git a/src/crypto/internal/fips140/aes/_asm/standard/aes_amd64.go b/src/crypto/internal/fips140/aes/_asm/standard/aes_amd64.go
index 44e0a79289..9e2fad28be 100644
--- a/src/crypto/internal/fips140/aes/_asm/standard/aes_amd64.go
+++ b/src/crypto/internal/fips140/aes/_asm/standard/aes_amd64.go
@@ -36,7 +36,7 @@ func main() {
 		"·_expand_key_256a<>",
 		"·_expand_key_256b<>",
 	}
-	removePeskyUnicodeDot(internalFunctions, "../../asm_amd64.s")
+	removePeskyUnicodeDot(internalFunctions, "../../aes_amd64.s")
 }
 
 func encryptBlockAsm() {
