diff --git a/src/crypto/internal/fips140test/entropy_test.go b/src/crypto/internal/fips140test/entropy_test.go
index 82d921ffb7..c9b3b5b10c 100644
--- a/src/crypto/internal/fips140test/entropy_test.go
+++ b/src/crypto/internal/fips140test/entropy_test.go
@@ -63,6 +63,7 @@ func TestEntropySamples(t *testing.T) {
 		if err != nil {
 			t.Fatalf("failed to create %q: %v", restartSamplesName, err)
 		}
+		defer f.Close()
 		for i := range restartSamples {
 			if _, err := f.Write(restartSamples[i][:]); err != nil {
 				t.Fatalf("failed to write samples to %q: %v", restartSamplesName, err)
