diff --git a/src/crypto/tls/quic.go b/src/crypto/tls/quic.go
index c7d8508acb..4596a19fc9 100644
--- a/src/crypto/tls/quic.go
+++ b/src/crypto/tls/quic.go
@@ -214,10 +214,10 @@ func (q *QUICConn) Start(ctx context.Context) error {
 	if q.conn.quic.started {
 		return quicError(errors.New("tls: Start called more than once"))
 	}
-	q.conn.quic.started = true
 	if q.conn.config.MinVersion < VersionTLS13 {
 		return quicError(errors.New("tls: Config MinVersion must be at least TLS 1.3"))
 	}
+	q.conn.quic.started = true
 	go q.conn.HandshakeContext(ctx)
 	if _, ok := <-q.conn.quic.blockedc; !ok {
 		return q.conn.handshakeErr
