diff --git a/src/crypto/x509/internal/macos/security.go b/src/crypto/x509/internal/macos/security.go
index 6af19bafe5..7aacefbdb9 100644
--- a/src/crypto/x509/internal/macos/security.go
+++ b/src/crypto/x509/internal/macos/security.go
@@ -76,7 +76,7 @@ func SecPolicyCreateSSL(name string) (CFRef, error) {
 	}
 	ret := syscall(abi.FuncPCABI0(x509_SecPolicyCreateSSL_trampoline), 1 /* true */, uintptr(hostname), 0, 0, 0, 0)
 	if ret == 0 {
-		return 0, OSStatus{"SecPolicyCreateSSL", int32(ret)}
+		return 0, errors.New("SecPolicyCreateSSL: failed to create policy")
 	}
 	return CFRef(ret), nil
 }
