diff --git a/src/crypto/x509/internal/macos/security.go b/src/crypto/x509/internal/macos/security.go
index 6af19bafe5..12c0ab9bfe 100644
--- a/src/crypto/x509/internal/macos/security.go
+++ b/src/crypto/x509/internal/macos/security.go
@@ -140,7 +140,7 @@ func x509_SecCertificateCopyData_trampoline()
 func SecTrustCopyCertificateChain(trustObj CFRef) (CFRef, error) {
 	ret := syscall(abi.FuncPCABI0(x509_SecTrustCopyCertificateChain_trampoline), uintptr(trustObj), 0, 0, 0, 0, 0)
 	if ret == 0 {
-		return 0, OSStatus{"SecTrustCopyCertificateChain", int32(ret)}
+		return 0, errors.New("SecTrustCopyCertificateChain: returned NULL certificate chain")
 	}
 	return CFRef(ret), nil
 }
