diff --git a/src/crypto/internal/fips140/ecdsa/ecdsa.go b/src/crypto/internal/fips140/ecdsa/ecdsa.go
index 1c00f55810..fd6978cfee 100644
--- a/src/crypto/internal/fips140/ecdsa/ecdsa.go
+++ b/src/crypto/internal/fips140/ecdsa/ecdsa.go
@@ -39,7 +39,7 @@ type PublicKey struct {
 }
 
 func (pub *PublicKey) Bytes() []byte {
-	return pub.q
+	return bytes.Clone(pub.q)
 }
 
 type curveID string
@@ -192,7 +192,7 @@ func NewPublicKey[P Point[P]](c *Curve[P], Q []byte) (*PublicKey, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &PublicKey{curve: c.curve, q: Q}, nil
+	return &PublicKey{curve: c.curve, q: bytes.Clone(Q)}, nil
 }
 
 // GenerateKey generates a new ECDSA private key pair for the specified curve.
