diff --git a/src/crypto/internal/fips140/aes/gcm/_asm/gcm/gcm_amd64_asm.go b/src/crypto/internal/fips140/aes/gcm/_asm/gcm/gcm_amd64_asm.go
index ed5f14b938..343e1b6369 100644
--- a/src/crypto/internal/fips140/aes/gcm/_asm/gcm/gcm_amd64_asm.go
+++ b/src/crypto/internal/fips140/aes/gcm/_asm/gcm/gcm_amd64_asm.go
@@ -1412,8 +1412,20 @@ func gcmAesDecTail(pTbl, ctx, ks, ptxLen, aluCTR, aluTMP, aluK, NR GPPhysical) {
 	Instruction(&ir.Instruction{Opcode: "LEAQ", Operands: []Op{andMask, aluCTR}})
 	MOVOU(Mem{Base: aluCTR, Index: aluTMP, Scale: 1}.Offset(-16), T1)
 
-	MOVOU(Mem{Base: ctx}, B0)
-	PAND(T1, B0)
+	LEAQ(Mem{Base: ctx, Index: ptxLen, Scale: 1}.Offset(-1), ctx)
+	PXOR(B0, B0)
+	// Hack to get Avo to emit:
+	// 	MOVQ ptxLen, aluTMP
+	Instruction(&ir.Instruction{Opcode: "MOVQ", Operands: []Op{ptxLen, aluTMP}})
+
+	Label("ctxLoadLoop")
+	PSLLDQ(Imm(1), B0)
+	PINSRB(Imm(0), Mem{Base: ctx}, B0)
+	LEAQ(Mem{Base: ctx}.Offset(-1), ctx)
+	// Hack to get Avo to emit:
+	// 	DECQ aluTMP
+	Instruction(&ir.Instruction{Opcode: "DECQ", Operands: []Op{aluTMP}})
+	JNE(LabelRef("ctxLoadLoop"))
 
 	MOVOU(B0, T1)
 	PSHUFB(BSWAP, B0)
