diff --git a/src/crypto/internal/fips140/tls13/tls13.go b/src/crypto/internal/fips140/tls13/tls13.go
index f2c2423a25..321720102a 100644
--- a/src/crypto/internal/fips140/tls13/tls13.go
+++ b/src/crypto/internal/fips140/tls13/tls13.go
@@ -18,7 +18,7 @@ import (
 
 // ExpandLabel implements HKDF-Expand-Label from RFC 8446, Section 7.1.
 func ExpandLabel[H hash.Hash](hash func() H, secret []byte, label string, context []byte, length int) []byte {
-	if len("tls13 ")+len(label) > 255 || len(context) > 255 {
+	if len("tls13 ")+len(label) > 255 || len(context) > 255 || length < 0 || length > 65535 {
 		// It should be impossible for this to panic: labels are fixed strings,
 		// and context is either a fixed-length computed hash, or parsed from a
 		// field which has the same length limitation.
