diff --git a/src/crypto/internal/boring/ecdh.go b/src/crypto/internal/boring/ecdh.go
index 4a887fdfd6..ffc1cf5a39 100644
--- a/src/crypto/internal/boring/ecdh.go
+++ b/src/crypto/internal/boring/ecdh.go
@@ -34,15 +34,15 @@ func (k *PrivateKeyECDH) finalize() {
 }
 
 func NewPublicKeyECDH(curve string, bytes []byte) (*PublicKeyECDH, error) {
-	if len(bytes) != 1+2*curveSize(curve) {
-		return nil, errors.New("NewPublicKeyECDH: wrong key length")
-	}
-
 	nid, err := curveNID(curve)
 	if err != nil {
 		return nil, err
 	}
 
+	if len(bytes) != 1+2*curveSize(curve) {
+		return nil, errors.New("NewPublicKeyECDH: wrong key length")
+	}
+
 	group := C._goboringcrypto_EC_GROUP_new_by_curve_name(nid)
 	if group == nil {
 		return nil, fail("EC_GROUP_new_by_curve_name")
@@ -70,14 +70,15 @@ func NewPublicKeyECDH(curve string, bytes []byte) (*PublicKeyECDH, error) {
 func (k *PublicKeyECDH) Bytes() []byte { return k.bytes }
 
 func NewPrivateKeyECDH(curve string, bytes []byte) (*PrivateKeyECDH, error) {
-	if len(bytes) != curveSize(curve) {
-		return nil, errors.New("NewPrivateKeyECDH: wrong key length")
-	}
-
 	nid, err := curveNID(curve)
 	if err != nil {
 		return nil, err
 	}
+
+	if len(bytes) != curveSize(curve) {
+		return nil, errors.New("NewPrivateKeyECDH: wrong key length")
+	}
+
 	key := C._goboringcrypto_EC_KEY_new_by_curve_name(nid)
 	if key == nil {
 		return nil, fail("EC_KEY_new_by_curve_name")
