diff --git a/src/crypto/rsa/fips.go b/src/crypto/rsa/fips.go
index fb2395886b..2524485a72 100644
--- a/src/crypto/rsa/fips.go
+++ b/src/crypto/rsa/fips.go
@@ -231,11 +231,12 @@ func encryptOAEP(hash hash.Hash, mgfHash hash.Hash, random io.Reader, pub *Publi
 	boring.UnreachableExceptTests()
 
 	hash = fips140hash.Unwrap(hash)
+	mgfHash = fips140hash.Unwrap(mgfHash)
 
 	if err := checkFIPS140OnlyPublicKey(pub); err != nil {
 		return nil, err
 	}
-	if fips140only.Enforced() && !fips140only.ApprovedHash(hash) {
+	if fips140only.Enforced() && (!fips140only.ApprovedHash(hash) || !fips140only.ApprovedHash(mgfHash)) {
 		return nil, errors.New("crypto/rsa: use of hash functions other than SHA-2 or SHA-3 is not allowed in FIPS 140-only mode")
 	}
 	if fips140only.Enforced() && !fips140only.ApprovedRandomReader(random) {
