diff --git a/src/crypto/ecdsa/ecdsa_legacy.go b/src/crypto/ecdsa/ecdsa_legacy.go
index 2fb1b21a60..a9e6c3b538 100644
--- a/src/crypto/ecdsa/ecdsa_legacy.go
+++ b/src/crypto/ecdsa/ecdsa_legacy.go
@@ -119,6 +119,9 @@ func signLegacy(priv *PrivateKey, csprng io.Reader, hash []byte) (sig []byte, er
 			}
 
 			kInv = new(big.Int).ModInverse(k, N)
+			if kInv == nil {
+				continue
+			}
 
 			r, _ = c.ScalarBaseMult(k.Bytes())
 			r.Mod(r, N)
