diff --git a/src/body.rs b/src/body.rs
index 9df3130..e86b8b4 100644
--- a/src/body.rs
+++ b/src/body.rs
@@ -311,6 +311,9 @@ impl HttpBody for Body {
                             Err(e) => return Poll::Ready(Some(Err(e.into()))),
                             Ok(()) => {
                                 let chunk = decoder_state.get_mut().get_mut().split().freeze();
+                                if chunk.is_empty() {
+                                    continue;
+                                }
                                 return Poll::Ready(Some(Ok(chunk)));
                             }
                         },
