diff --git a/src/config/secret_store.rs b/src/config/secret_store.rs
index 35707cc..107849e 100644
--- a/src/config/secret_store.rs
+++ b/src/config/secret_store.rs
@@ -157,7 +157,10 @@ impl TryFrom<Table> for SecretStoreConfig {
                                 }
                             })?;
                             std::env::var(var)
-                                .unwrap_or_else(|_| String::new())
+                                .map_err(|e| FastlyConfigError::InvalidSecretStoreDefinition {
+                                    name: store_name.to_string(),
+                                    err: SecretStoreConfigError::EnvVarError(e),
+                                })?
                                 .into_bytes()
                         } else {
                             unreachable!()
diff --git a/src/error.rs b/src/error.rs
index 3f8a373..05c4201 100644
--- a/src/error.rs
+++ b/src/error.rs
@@ -794,6 +794,8 @@ pub enum SecretStoreConfigError {
     FileNotAString(String),
     #[error("The `env` value for the object `{0}` is not a string.")]
     EnvNotAString(String),
+    #[error(transparent)]
+    EnvVarError(std::env::VarError),
 
     #[error("The `key` key for an object is not set. It must be used.")]
     NoKey,
