diff --git a/src/wiggle_abi/shielding.rs b/src/wiggle_abi/shielding.rs
index e959fc3..c200797 100644
--- a/src/wiggle_abi/shielding.rs
+++ b/src/wiggle_abi/shielding.rs
@@ -82,6 +82,10 @@ impl fastly_shielding::FastlyShielding for Session {
             }
         }
 
+        if !shield_backend_options.is_empty() {
+            return Err(Error::InvalidArgument);
+        }
+
         let Ok(uri) = Uri::from_str(&shield_uri) else {
             return Err(Error::InvalidArgument);
         };
