diff --git a/src/component/erl.rs b/src/component/erl.rs
index 840a0a0..08cabb1 100644
--- a/src/component/erl.rs
+++ b/src/component/erl.rs
@@ -11,7 +11,7 @@ pub(crate) fn check_rate(
     _pb: &str,
     _ttl: u32,
 ) -> Result<bool, types::Error> {
-    Ok(false)
+    Err(types::Error::Unsupported)
 }
 
 pub(crate) fn ratecounter_increment(
@@ -20,7 +20,7 @@ pub(crate) fn ratecounter_increment(
     _entry: String,
     _delta: u32,
 ) -> Result<(), types::Error> {
-    Ok(())
+    Err(types::Error::Unsupported)
 }
 
 pub(crate) fn ratecounter_lookup_rate(
@@ -29,7 +29,7 @@ pub(crate) fn ratecounter_lookup_rate(
     _entry: String,
     _window: u32,
 ) -> Result<u32, types::Error> {
-    Ok(0)
+    Err(types::Error::Unsupported)
 }
 
 pub(crate) fn ratecounter_lookup_count(
@@ -38,7 +38,7 @@ pub(crate) fn ratecounter_lookup_count(
     _entry: String,
     _duration: u32,
 ) -> Result<u32, types::Error> {
-    Ok(0)
+    Err(types::Error::Unsupported)
 }
 
 pub(crate) fn penaltybox_add(
@@ -47,7 +47,7 @@ pub(crate) fn penaltybox_add(
     _entry: String,
     _ttl: u32,
 ) -> Result<(), types::Error> {
-    Ok(())
+    Err(types::Error::Unsupported)
 }
 
 pub(crate) fn penaltybox_has(
@@ -55,5 +55,5 @@ pub(crate) fn penaltybox_has(
     _pb: &str,
     _entry: String,
 ) -> Result<bool, types::Error> {
-    Ok(false)
+    Err(types::Error::Unsupported)
 }
