diff --git a/src/wiggle_abi/erl_impl.rs b/src/wiggle_abi/erl_impl.rs
index dba0cd7..ba0b6d5 100644
--- a/src/wiggle_abi/erl_impl.rs
+++ b/src/wiggle_abi/erl_impl.rs
@@ -27,7 +27,7 @@ impl FastlyErl for Session {
         _entry: GuestPtr<str>,
         _delta: u32,
     ) -> std::result::Result<(), Error> {
-        Ok(())
+        Err(Error::NotAvailable("Edge rate limiting"))
     }
 
     fn ratecounter_lookup_rate(
@@ -57,7 +57,7 @@ impl FastlyErl for Session {
         _entry: GuestPtr<str>,
         _ttl: u32,
     ) -> std::result::Result<(), Error> {
-        Ok(())
+        Err(Error::NotAvailable("Edge rate limiting"))
     }
 
     fn penaltybox_has(
