diff --git a/src/core/cors.cpp b/src/core/cors.cpp
index 79b8ad8c..adf2ed6a 100644
--- a/src/core/cors.cpp
+++ b/src/core/cors.cpp
@@ -88,18 +88,17 @@ void applyCorsHeaders(const HttpHeaders &requestHeaders, HttpHeaders *responseHe
                 HttpHeader("Access-Control-Expose-Headers", HttpHeaders::join(exposeHeaders));
     }
 
-    if (!responseHeaders->contains("Access-Control-Allow-Credentials"))
-        *responseHeaders += HttpHeader("Access-Control-Allow-Credentials", "true");
-
     if (!responseHeaders->contains("Access-Control-Allow-Origin")) {
         CowByteArray origin = requestHeaders.get("Origin");
 
         if (origin.isEmpty())
-            origin = "*";
-
-        *responseHeaders += HttpHeader("Access-Control-Allow-Origin", origin);
+            *responseHeaders += HttpHeader("Access-Control-Allow-Origin", "*");
     }
 
+    if (!responseHeaders->contains("Access-Control-Allow-Credentials") &&
+        responseHeaders->contains("Access-Control-Allow-Origin"))
+        *responseHeaders += HttpHeader("Access-Control-Allow-Credentials", "true");
+
     if (!responseHeaders->contains("Access-Control-Max-Age"))
         *responseHeaders += HttpHeader("Access-Control-Max-Age", "3600");
 }
