diff --git a/lib/ldap.c b/lib/ldap.c
index 3705754476..9247f4111a 100644
--- a/lib/ldap.c
+++ b/lib/ldap.c
@@ -194,9 +194,8 @@ static ULONG ldap_win_bind_auth(LDAP *server, const char *user,
       rc = LDAP_NO_MEMORY;
     }
   }
-  else {
+  else if(method) {
     /* proceed with current user credentials */
-    method = LDAP_AUTH_NEGOTIATE;
     rc = ldap_bind_s(server, NULL, NULL, method);
   }
   return rc;
