diff --git a/lib/telnet.c b/lib/telnet.c
index c5ce9c2c97..f395753c88 100644
--- a/lib/telnet.c
+++ b/lib/telnet.c
@@ -49,6 +49,7 @@

 #include "url.h"
 #include "transfer.h"
+#include "connect.h"
 #include "sendf.h"
 #include "curl_trc.h"
 #include "progress.h"
@@ -643,15 +644,22 @@ static CURLcode send_telnet_data(struct Curl_easy *data,
     outbuf = (const unsigned char *)buffer;
   }
   while(!result && total_written < outlen) {
+    timediff_t timeout_ms = Curl_timeleft_ms(data);
     /* Make sure socket is writable to avoid EWOULDBLOCK condition */
     struct pollfd pfd[1];
     pfd[0].fd = conn->sock[FIRSTSOCKET];
     pfd[0].events = POLLOUT;
-    switch(Curl_poll(pfd, 1, -1)) {
+    if(timeout_ms < 0) {
+      result = CURLE_OPERATION_TIMEDOUT;
+      break;
+    }
+    switch(Curl_poll(pfd, 1, timeout_ms ? timeout_ms : -1)) {
     case -1:                    /* error, abort writing */
-    case 0:                     /* timeout (will never happen) */
       result = CURLE_SEND_ERROR;
       break;
+    case 0:                     /* timeout */
+      result = CURLE_OPERATION_TIMEDOUT;
+      break;
     default:                    /* write! */
       bytes_written = 0;
       result = Curl_xfer_send(data, outbuf + total_written,
