diff --git a/lib/tftp.c b/lib/tftp.c
index 6cc672d447..5825b79df0 100644
--- a/lib/tftp.c
+++ b/lib/tftp.c
@@ -1054,6 +1054,42 @@ static CURLcode tftp_receive_packet(struct Curl_easy *data,
       }
     }
     else {
+      const struct Curl_sockaddr_ex *peer_addr =
+        Curl_conn_get_remote_addr(data, FIRSTSOCKET);
+      const struct sockaddr *from = (const struct sockaddr *)&remote_addr;
+      const struct sockaddr *peer;
+      bool same_addr = FALSE;
+      if(!peer_addr)
+        return CURLE_FAILED_INIT;
+      peer = &peer_addr->curl_sa_addr;
+      if(peer->sa_family == from->sa_family) {
+        switch(from->sa_family) {
+        case AF_INET: {
+          const struct sockaddr_in *from4 = (const void *)from;
+          const struct sockaddr_in *peer4 = (const void *)peer;
+          same_addr = !memcmp(&from4->sin_addr, &peer4->sin_addr,
+                              sizeof(from4->sin_addr));
+          break;
+        }
+#ifdef USE_IPV6
+        case AF_INET6: {
+          const struct sockaddr_in6 *from6 = (const void *)from;
+          const struct sockaddr_in6 *peer6 = (const void *)peer;
+          same_addr = !memcmp(&from6->sin6_addr, &peer6->sin6_addr,
+                              sizeof(from6->sin6_addr));
+#ifdef HAVE_SOCKADDR_IN6_SIN6_SCOPE_ID
+          same_addr = same_addr &&
+            (from6->sin6_scope_id == peer6->sin6_scope_id);
+#endif
+          break;
+        }
+#endif
+        }
+      }
+      if(!same_addr) {
+        failf(data, "Data received from another address");
+        return CURLE_RECV_ERROR;
+      }
       /* pin address on first use */
       state->remote_pinned = TRUE;
       state->remote_addrlen = fromlen;
