diff --git a/contrib/minizip/unzip.c b/contrib/minizip/unzip.c
index 827c23a..88f417e 100644
--- a/contrib/minizip/unzip.c
+++ b/contrib/minizip/unzip.c
@@ -1921,6 +1921,9 @@ extern int ZEXPORT unzGetGlobalComment(unzFile file, char * szComment, uLong uSi
     if (uReadThis>s->gi.size_comment)
         uReadThis = s->gi.size_comment;
 
+    if ((uReadThis > 0) && (szComment == NULL))
+        return (int)UNZ_PARAMERROR;
+
     if (ZSEEK64(s->z_filefunc,s->filestream,s->central_pos+22,ZLIB_FILEFUNC_SEEK_SET)!=0)
         return UNZ_ERRNO;
