diff --git a/contrib/minizip/zip.c b/contrib/minizip/zip.c
index 0bbdf77..0fe1734 100644
--- a/contrib/minizip/zip.c
+++ b/contrib/minizip/zip.c
@@ -1399,6 +1399,8 @@ extern int ZEXPORT zipOpenNewFileInZip4_64(zipFile file, const char* filename, c
     zi->ci.size_centralExtraFree = 32; /* Extra space we have reserved in case we need to add ZIP64 extra info data */
 
     zi->ci.central_header = (char*)ALLOC((uInt)zi->ci.size_centralheader + zi->ci.size_centralExtraFree);
+    if (zi->ci.central_header == NULL)
+        return ZIP_INTERNALERROR;
 
     zi->ci.size_centralExtra = size_extrafield_global;
     zip64local_putValue_inmemory(zi->ci.central_header,(uLong)CENTRALHEADERMAGIC,4);
