diff --git a/contrib/dotzlib/DotZLib/Inflater.cs b/contrib/dotzlib/DotZLib/Inflater.cs
index 38a8e3a..4df4f83 100644
--- a/contrib/dotzlib/DotZLib/Inflater.cs
+++ b/contrib/dotzlib/DotZLib/Inflater.cs
@@ -90,6 +90,10 @@ namespace DotZLib
                 OnDataAvailable();
             }
             while (err == 0);
+
+            if (err != 1)
+                throw new ZLibException(err);
+
             setChecksum( _ztream.adler );
             inflateReset(ref _ztream);
             resetOutput();
