diff --git a/contrib/minizip/zip.c b/contrib/minizip/zip.c
index 0bbdf77..1b715e7 100644
--- a/contrib/minizip/zip.c
+++ b/contrib/minizip/zip.c
@@ -1318,10 +1318,8 @@ extern int ZEXPORT zipOpenNewFileInZip4_64(zipFile file, const char* filename, c
     uInt i;
     int err = ZIP_OK;
 
-#    ifdef NOCRYPT
     if (password != NULL)
         return ZIP_PARAMERROR;
-#    endif
 
     if (file == NULL)
         return ZIP_PARAMERROR;
