diff --git a/examples/gzappend.c b/examples/gzappend.c
index 23e93cf..f98a4dc 100644
--- a/examples/gzappend.c
+++ b/examples/gzappend.c
@@ -396,8 +396,7 @@ local void gztack(char *name, int gd, z_stream *strm, int last)
     if (name != NULL) {
         fd = open(name, O_RDONLY, 0);
         if (fd == -1)
-            fprintf(stderr, "gzappend warning: %s not found, skipping ...\n",
-                    name);
+            bye("cannot open ", name);
     }
 
     /* allocate buffers */
