diff --git a/contrib/puff/bin-writer.c b/contrib/puff/bin-writer.c
index cfd8306..5508b5e 100644
--- a/contrib/puff/bin-writer.c
+++ b/contrib/puff/bin-writer.c
@@ -12,8 +12,12 @@ int main(void) {
 
     // Read characters until EOF
     while((ch = getchar()) != EOF) {
+        int next;
         hexStr[0] = (char)ch;
-        hexStr[1] = (char)getchar();
+        next = getchar();
+        if(next == EOF)
+          return 1;
+        hexStr[1] = (char)next;
         hexStr[2] = '\0'; // Null-terminate string
         char *endptr;
         unsigned char byte = (unsigned char)strtol(hexStr, &endptr, 16);
