diff --git a/contrib/puff/bin-writer.c b/contrib/puff/bin-writer.c
index cfd8306..64e31cd 100644
--- a/contrib/puff/bin-writer.c
+++ b/contrib/puff/bin-writer.c
@@ -17,6 +17,8 @@ int main(void) {
         hexStr[2] = '\0'; // Null-terminate string
         char *endptr;
         unsigned char byte = (unsigned char)strtol(hexStr, &endptr, 16);
+        if(endptr != hexStr + 2)
+          return 1;
         fwrite(&byte, 1, 1, stdout);
         if((ch = getchar()) == EOF) // Read seaparating space
           break;
